{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curate perturbation dataset with `PerturbationCurator`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we use pertpy's `PerturbationCurator` to ensure that a perturbatin dataset conforms to both, `CELLxGENE` (schema 5.1.0) and pertpy's defined criteria.\n",
    "More specifically, the `PerturbationCurator` builds upon [cellxgene-lamin](https://github.com/laminlabs/cellxgene-lamin) and extends it by further requiring `cell_line` and `X_perturbations` columns for the perturbations.\n",
    "\n",
    "This guide demonstrates how to curate a complex, real world perturbation dataset [McFarland et al. 2020](https://www.nature.com/articles/s41467-020-17440-w) using `PerturbationCurator`. Please have a look at [lamindb's perturbation guide](https://docs.lamin.ai/perturbation) for more details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: sunnyosun/test-perturbation\n"
     ]
    }
   ],
   "source": [
    "#! pip install pertpy-datasets\n",
    "\n",
    "# using a local instance here but in practice, we use `laminlabs/pertpy-datasets`\n",
    "!lamin init --storage ./test-perturbation --schema bionty,wetlab,ourprojects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: sunnyosun/test-perturbation\n",
      "\u001b[92m→\u001b[0m loaded Transform('HIRTYxL3'), re-started Run('ZIxUL0yY') at 2024-12-12 17:26:45 UTC\n",
      "\u001b[92m→\u001b[0m notebook imports: bionty==0.53.2 lamindb==0.77.2 ourprojects==0.0.1 pertpy_datasets==0.1.0 scanpy==1.10.1 wetlab==0.36.0\n"
     ]
    }
   ],
   "source": [
    "import lamindb as ln\n",
    "import bionty as bt\n",
    "import wetlab as wl\n",
    "import ourprojects as ops\n",
    "import pertpy_datasets as pts\n",
    "import scanpy as sc\n",
    "from datetime import timedelta\n",
    "\n",
    "ln.track(\"HIRTYxL3aZc70000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": [
     "hide-cell"
    ]
   },
   "outputs": [],
   "source": [
    "# only for a new instance, add a source to wetlab.Compound\n",
    "chebi_source = bt.Source.filter(entity=\"wetlab.Compound\", name=\"chebi\").first()\n",
    "if not chebi_source:\n",
    "    wl.Compound.add_source(\n",
    "        bt.Source.filter(entity=\"Drug\", name=\"chebi\").first()\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m completing transfer to track Artifact('Xk7Qaik9') as input\n",
      "\u001b[92m→\u001b[0m mapped records: Artifact(uid='Xk7Qaik9vBLV4PKf0001')\n",
      "\u001b[92m→\u001b[0m transferred records: \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cell_det_rate</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_quality</th>\n",
       "      <th>channel</th>\n",
       "      <th>disease</th>\n",
       "      <th>dose_unit</th>\n",
       "      <th>dose_value</th>\n",
       "      <th>doublet_CL1</th>\n",
       "      <th>doublet_CL2</th>\n",
       "      <th>doublet_GMM_prob</th>\n",
       "      <th>doublet_dev_imp</th>\n",
       "      <th>doublet_z_margin</th>\n",
       "      <th>hash_assignment</th>\n",
       "      <th>hash_tag</th>\n",
       "      <th>num_SNPs</th>\n",
       "      <th>organism</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>singlet_ID</th>\n",
       "      <th>singlet_dev</th>\n",
       "      <th>singlet_dev_z</th>\n",
       "      <th>singlet_margin</th>\n",
       "      <th>singlet_z_margin</th>\n",
       "      <th>time</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>tot_reads</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>chembl-ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACTGGTGTCTCTCTG</th>\n",
       "      <td>ACH-000390</td>\n",
       "      <td>True</td>\n",
       "      <td>0.093159</td>\n",
       "      <td>LUDLU-1</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>TE14_OESOPHAGUS</td>\n",
       "      <td>2.269468e-10</td>\n",
       "      <td>0.009426</td>\n",
       "      <td>0.403316</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>481</td>\n",
       "      <td>human</td>\n",
       "      <td>trametinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>0.655877</td>\n",
       "      <td>14.860933</td>\n",
       "      <td>0.462273</td>\n",
       "      <td>12.351139</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>787</td>\n",
       "      <td>1</td>\n",
       "      <td>3045</td>\n",
       "      <td>12895.0</td>\n",
       "      <td>3.202792</td>\n",
       "      <td>24.955409</td>\n",
       "      <td>CHEMBL2103875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAGGCTCAGATTTCG</th>\n",
       "      <td>ACH-000444</td>\n",
       "      <td>True</td>\n",
       "      <td>0.145728</td>\n",
       "      <td>LU99</td>\n",
       "      <td>normal</td>\n",
       "      <td>2</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.5</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>MCAS_OVARY</td>\n",
       "      <td>8.562908e-04</td>\n",
       "      <td>0.010173</td>\n",
       "      <td>0.188284</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>1003</td>\n",
       "      <td>human</td>\n",
       "      <td>afatinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>0.762847</td>\n",
       "      <td>10.648094</td>\n",
       "      <td>0.474590</td>\n",
       "      <td>8.164565</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1597</td>\n",
       "      <td>1</td>\n",
       "      <td>4763</td>\n",
       "      <td>23161.0</td>\n",
       "      <td>7.473771</td>\n",
       "      <td>18.051898</td>\n",
       "      <td>CHEMBL1173655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCCAAATCAAGCCGTC</th>\n",
       "      <td>ACH-000396</td>\n",
       "      <td>True</td>\n",
       "      <td>0.117330</td>\n",
       "      <td>J82</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>urinary bladder carcinoma</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>IGR1_SKIN</td>\n",
       "      <td>6.490367e-08</td>\n",
       "      <td>0.009686</td>\n",
       "      <td>1.185862</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>647</td>\n",
       "      <td>human</td>\n",
       "      <td>dabrafenib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>0.651059</td>\n",
       "      <td>14.740111</td>\n",
       "      <td>0.404508</td>\n",
       "      <td>11.188513</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1159</td>\n",
       "      <td>1</td>\n",
       "      <td>3834</td>\n",
       "      <td>18062.0</td>\n",
       "      <td>2.762706</td>\n",
       "      <td>22.085040</td>\n",
       "      <td>CHEMBL2028663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGGAGAAGTCGCGTCA</th>\n",
       "      <td>ACH-000997</td>\n",
       "      <td>True</td>\n",
       "      <td>0.005422</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>low_quality</td>\n",
       "      <td>7</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>HCT15_LARGE_INTESTINE</td>\n",
       "      <td>NCIH322_LUNG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029753</td>\n",
       "      <td>0.000794</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>30</td>\n",
       "      <td>human</td>\n",
       "      <td>gemcitabine</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>HCT15_LARGE_INTESTINE</td>\n",
       "      <td>0.970247</td>\n",
       "      <td>2.852338</td>\n",
       "      <td>0.168971</td>\n",
       "      <td>0.833455</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>76</td>\n",
       "      <td>1</td>\n",
       "      <td>178</td>\n",
       "      <td>726.0</td>\n",
       "      <td>70.247934</td>\n",
       "      <td>5.785124</td>\n",
       "      <td>CHEMBL888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTTGGAGATCGATA</th>\n",
       "      <td>ACH-000723</td>\n",
       "      <td>True</td>\n",
       "      <td>0.132708</td>\n",
       "      <td>YD-10B</td>\n",
       "      <td>low_quality</td>\n",
       "      <td>nan</td>\n",
       "      <td>head and neck cancer</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YD10B_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>647V_URINARY_TRACT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.156492</td>\n",
       "      <td>1.556214</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>874</td>\n",
       "      <td>human</td>\n",
       "      <td>sggpx4-2</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>Male</td>\n",
       "      <td>YD10B_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>0.292802</td>\n",
       "      <td>3.272682</td>\n",
       "      <td>0.016459</td>\n",
       "      <td>0.330120</td>\n",
       "      <td>72, 96</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>2105</td>\n",
       "      <td>1</td>\n",
       "      <td>4341</td>\n",
       "      <td>20693.0</td>\n",
       "      <td>0.695887</td>\n",
       "      <td>16.242208</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   depmap_id  cancer  cell_det_rate cell_line cell_quality  \\\n",
       "AACTGGTGTCTCTCTG  ACH-000390    True       0.093159   LUDLU-1       normal   \n",
       "ATAGGCTCAGATTTCG  ACH-000444    True       0.145728      LU99       normal   \n",
       "GCCAAATCAAGCCGTC  ACH-000396    True       0.117330       J82       normal   \n",
       "CGGAGAAGTCGCGTCA  ACH-000997    True       0.005422    HCT-15  low_quality   \n",
       "TAGTTGGAGATCGATA  ACH-000723    True       0.132708    YD-10B  low_quality   \n",
       "\n",
       "                 channel                    disease dose_unit  dose_value  \\\n",
       "AACTGGTGTCTCTCTG     nan                lung cancer        µM         0.1   \n",
       "ATAGGCTCAGATTTCG       2                lung cancer        µM         0.5   \n",
       "GCCAAATCAAGCCGTC     nan  urinary bladder carcinoma        µM         0.1   \n",
       "CGGAGAAGTCGCGTCA       7          colorectal cancer        µM         0.1   \n",
       "TAGTTGGAGATCGATA     nan       head and neck cancer       nan         NaN   \n",
       "\n",
       "                                      doublet_CL1         doublet_CL2  \\\n",
       "AACTGGTGTCTCTCTG                      LUDLU1_LUNG     TE14_OESOPHAGUS   \n",
       "ATAGGCTCAGATTTCG                        LU99_LUNG          MCAS_OVARY   \n",
       "GCCAAATCAAGCCGTC                J82_URINARY_TRACT           IGR1_SKIN   \n",
       "CGGAGAAGTCGCGTCA            HCT15_LARGE_INTESTINE        NCIH322_LUNG   \n",
       "TAGTTGGAGATCGATA  YD10B_UPPER_AERODIGESTIVE_TRACT  647V_URINARY_TRACT   \n",
       "\n",
       "                  doublet_GMM_prob  doublet_dev_imp  doublet_z_margin  \\\n",
       "AACTGGTGTCTCTCTG      2.269468e-10         0.009426          0.403316   \n",
       "ATAGGCTCAGATTTCG      8.562908e-04         0.010173          0.188284   \n",
       "GCCAAATCAAGCCGTC      6.490367e-08         0.009686          1.185862   \n",
       "CGGAGAAGTCGCGTCA               NaN         0.029753          0.000794   \n",
       "TAGTTGGAGATCGATA               NaN         0.156492          1.556214   \n",
       "\n",
       "                 hash_assignment hash_tag  num_SNPs organism perturbation  \\\n",
       "AACTGGTGTCTCTCTG             nan      nan       481    human   trametinib   \n",
       "ATAGGCTCAGATTTCG             nan      nan      1003    human     afatinib   \n",
       "GCCAAATCAAGCCGTC             nan      nan       647    human   dabrafenib   \n",
       "CGGAGAAGTCGCGTCA             nan      nan        30    human  gemcitabine   \n",
       "TAGTTGGAGATCGATA             nan      nan       874    human     sggpx4-2   \n",
       "\n",
       "                 perturbation_type   sex                       singlet_ID  \\\n",
       "AACTGGTGTCTCTCTG              drug  Male                      LUDLU1_LUNG   \n",
       "ATAGGCTCAGATTTCG              drug  Male                        LU99_LUNG   \n",
       "GCCAAATCAAGCCGTC              drug  Male                J82_URINARY_TRACT   \n",
       "CGGAGAAGTCGCGTCA              drug  Male            HCT15_LARGE_INTESTINE   \n",
       "TAGTTGGAGATCGATA            CRISPR  Male  YD10B_UPPER_AERODIGESTIVE_TRACT   \n",
       "\n",
       "                  singlet_dev  singlet_dev_z  singlet_margin  \\\n",
       "AACTGGTGTCTCTCTG     0.655877      14.860933        0.462273   \n",
       "ATAGGCTCAGATTTCG     0.762847      10.648094        0.474590   \n",
       "GCCAAATCAAGCCGTC     0.651059      14.740111        0.404508   \n",
       "CGGAGAAGTCGCGTCA     0.970247       2.852338        0.168971   \n",
       "TAGTTGGAGATCGATA     0.292802       3.272682        0.016459   \n",
       "\n",
       "                  singlet_z_margin    time tissue_type  tot_reads  nperts  \\\n",
       "AACTGGTGTCTCTCTG         12.351139      24   cell_line        787       1   \n",
       "ATAGGCTCAGATTTCG          8.164565      24   cell_line       1597       1   \n",
       "GCCAAATCAAGCCGTC         11.188513      24   cell_line       1159       1   \n",
       "CGGAGAAGTCGCGTCA          0.833455      24   cell_line         76       1   \n",
       "TAGTTGGAGATCGATA          0.330120  72, 96   cell_line       2105       1   \n",
       "\n",
       "                  ngenes  ncounts  percent_mito  percent_ribo      chembl-ID  \n",
       "AACTGGTGTCTCTCTG    3045  12895.0      3.202792     24.955409  CHEMBL2103875  \n",
       "ATAGGCTCAGATTTCG    4763  23161.0      7.473771     18.051898  CHEMBL1173655  \n",
       "GCCAAATCAAGCCGTC    3834  18062.0      2.762706     22.085040  CHEMBL2028663  \n",
       "CGGAGAAGTCGCGTCA     178    726.0     70.247934      5.785124      CHEMBL888  \n",
       "TAGTTGGAGATCGATA    4341  20693.0      0.695887     16.242208            NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = ln.Artifact.using(\"laminlabs/lamindata\").get(uid=\"Xk7Qaik9vBLV4PKf0001\").load()\n",
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate an embedding because CELLxGENE requires one\n",
    "sc.tl.pca(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curate and register perturbations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genetic perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, 'sggpx4-2', 'sglacz', 'sggpx4-1', 'sgor2j2']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[\"genetic_perturbation\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"CRISPR\", None\n",
    ")\n",
    "\n",
    "list(adata.obs[\"genetic_perturbation\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>system</th>\n",
       "      <th>description</th>\n",
       "      <th>sequence</th>\n",
       "      <th>on_target_score</th>\n",
       "      <th>off_target_score</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ddg2RoCmNNAi</td>\n",
       "      <td>sggpx4-1</td>\n",
       "      <td>CRISPR-Cas9</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:26:56.849537+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid      name       system description sequence on_target_score  \\\n",
       "id                                                                             \n",
       "1   ddg2RoCmNNAi  sggpx4-1  CRISPR-Cas9        None     None            None   \n",
       "\n",
       "   off_target_score  run_id                       created_at  created_by_id  \n",
       "id                                                                           \n",
       "1              None       1 2024-12-12 17:26:56.849537+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning existing PerturbationTarget record with same name: 'GPX4'\n",
      "\u001b[93m!\u001b[0m ambiguous validation in Bionty for 1 record: 'OR2J2'\n"
     ]
    }
   ],
   "source": [
    "# register genetic perturbations with their target genes\n",
    "perturbations = {\n",
    "    \"sggpx4-1\": \"GPX4\",\n",
    "    \"sggpx4-2\": \"GPX4\",\n",
    "    \"sgor2j2\": \"OR2J2\", # cutting control\n",
    "}\n",
    "\n",
    "for sg_name, gene_symbol in perturbations.items():\n",
    "    perturbation = wl.GeneticPerturbation(system=\"CRISPR-Cas9\", name=sg_name, description = \"cutting control\" if sg_name == \"sgor2j2\" else None).save()\n",
    "    target = wl.PerturbationTarget(name=gene_symbol).save()\n",
    "    perturbation.targets.add(target)\n",
    "    gene = bt.Gene.from_source(symbol=gene_symbol, organism=\"human\").save()\n",
    "    target.genes.set([gene] if isinstance(gene, bt.Gene) else gene)\n",
    "\n",
    "# register the negative control without targets: Non-cutting control, plasmid backbone\n",
    "wl.GeneticPerturbation(system=\"CRISPR-Cas9\", name=\"sglacz\").save();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m \u001b[1;91mdid not create\u001b[0m Compound records for \u001b[1;93m6 non-validated\u001b[0m \u001b[3mnames\u001b[0m: \u001b[1;93m'azd5591', 'brd3379', 'control', 'idasanutlin', 'prexasertib', 'taselisib'\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "# register compounds from CHEBI\n",
    "adata.obs[\"compound\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"drug\", None\n",
    ")\n",
    "\n",
    "wl.Compound.from_values(adata.obs[\"compound\"]).save();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the remaining compounds are not in CHEBI and we create records for them\n",
    "compounds = [wl.Compound(name=cpd) for cpd in {\"azd5591\", \"brd3379\", \"control\", \"idasanutlin\", \"prexasertib\",\"taselisib\"}]\n",
    "ln.save(compounds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>concentration</th>\n",
       "      <th>concentration_unit</th>\n",
       "      <th>duration</th>\n",
       "      <th>compound_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>uZddC5g2dKOb</td>\n",
       "      <td>trametinib_0.1µM_24h</td>\n",
       "      <td>None</td>\n",
       "      <td>0.1</td>\n",
       "      <td>µM</td>\n",
       "      <td>1 days</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:27:09.902673+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid                  name description  concentration  \\\n",
       "id                                                                  \n",
       "1   uZddC5g2dKOb  trametinib_0.1µM_24h        None            0.1   \n",
       "\n",
       "   concentration_unit duration  compound_id  run_id  \\\n",
       "id                                                    \n",
       "1                  µM   1 days            1       1   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "1  2024-12-12 17:27:09.902673+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning existing CompoundPerturbation record with same name: 'control'\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>concentration</th>\n",
       "      <th>concentration_unit</th>\n",
       "      <th>duration</th>\n",
       "      <th>compound_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P7hRCQCDY7PY</td>\n",
       "      <td>bortezomib_2.5µM_6h</td>\n",
       "      <td>None</td>\n",
       "      <td>2.5</td>\n",
       "      <td>µM</td>\n",
       "      <td>0 days 06:00:00</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:27:10.003650+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid                 name description  concentration  \\\n",
       "id                                                                 \n",
       "7   P7hRCQCDY7PY  bortezomib_2.5µM_6h        None            2.5   \n",
       "\n",
       "   concentration_unit        duration  compound_id  run_id  \\\n",
       "id                                                           \n",
       "7                  µM 0 days 06:00:00            6       1   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "7  2024-12-12 17:27:10.003650+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>concentration</th>\n",
       "      <th>concentration_unit</th>\n",
       "      <th>duration</th>\n",
       "      <th>compound_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SeHMdmnKhIX5</td>\n",
       "      <td>brd3379_0.0µM_6h</td>\n",
       "      <td>None</td>\n",
       "      <td>0.0</td>\n",
       "      <td>µM</td>\n",
       "      <td>0 days 06:00:00</td>\n",
       "      <td>153</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:27:10.029267+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid              name description  concentration  \\\n",
       "id                                                              \n",
       "8   SeHMdmnKhIX5  brd3379_0.0µM_6h        None            0.0   \n",
       "\n",
       "   concentration_unit        duration  compound_id  run_id  \\\n",
       "id                                                           \n",
       "8                  µM 0 days 06:00:00          153       1   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "8  2024-12-12 17:27:10.029267+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>concentration</th>\n",
       "      <th>concentration_unit</th>\n",
       "      <th>duration</th>\n",
       "      <th>compound_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>frhpjw4KIBuX</td>\n",
       "      <td>idasanutlin_2.5µM_6h</td>\n",
       "      <td>None</td>\n",
       "      <td>2.5</td>\n",
       "      <td>µM</td>\n",
       "      <td>0 days 06:00:00</td>\n",
       "      <td>157</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:27:10.118204+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid                  name description  concentration  \\\n",
       "id                                                                  \n",
       "16  frhpjw4KIBuX  idasanutlin_2.5µM_6h        None            2.5   \n",
       "\n",
       "   concentration_unit        duration  compound_id  run_id  \\\n",
       "id                                                           \n",
       "16                 µM 0 days 06:00:00          157       1   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "16 2024-12-12 17:27:10.118204+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>concentration</th>\n",
       "      <th>concentration_unit</th>\n",
       "      <th>duration</th>\n",
       "      <th>compound_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>uZddC5g2dKOb</td>\n",
       "      <td>trametinib_0.1µM_24h</td>\n",
       "      <td>None</td>\n",
       "      <td>0.1</td>\n",
       "      <td>µM</td>\n",
       "      <td>1 days</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:27:09.902673+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>dIYVGDHIAl4G</td>\n",
       "      <td>trametinib_0.1µM_48h</td>\n",
       "      <td>None</td>\n",
       "      <td>0.1</td>\n",
       "      <td>µM</td>\n",
       "      <td>2 days</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:27:09.989896+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid                  name description  concentration  \\\n",
       "id                                                                  \n",
       "1   uZddC5g2dKOb  trametinib_0.1µM_24h        None            0.1   \n",
       "6   dIYVGDHIAl4G  trametinib_0.1µM_48h        None            0.1   \n",
       "\n",
       "   concentration_unit duration  compound_id  run_id  \\\n",
       "id                                                    \n",
       "1                  µM   1 days            1       1   \n",
       "6                  µM   2 days            1       1   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "1  2024-12-12 17:27:09.902673+00:00              1  \n",
       "6  2024-12-12 17:27:09.989896+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# register compound perturbations: compound + dose + duration\n",
    "compound_meta = adata.obs[adata.obs[\"compound\"].notna()][\n",
    "    [\"perturbation\", \"dose_unit\", \"dose_value\", \"time\"]\n",
    "].drop_duplicates()\n",
    "\n",
    "compound_perturbation_names = {}\n",
    "for _, row in compound_meta.iterrows():\n",
    "    compound = wl.Compound.get(name=row[\"perturbation\"])\n",
    "    duration = row.time.split(', ')[-1]  # take the longest duration\n",
    "    compound_perturbation_name = \"control\" if row[\"perturbation\"] == \"control\" else f\"{row['perturbation']}_{row.dose_value}{row.dose_unit}_{duration}h\"\n",
    "    compound_perturbation = wl.CompoundPerturbation(\n",
    "            name=compound_perturbation_name,\n",
    "            concentration=row.dose_value,\n",
    "            concentration_unit=row.dose_unit,\n",
    "            duration=timedelta(hours=int(duration)),\n",
    "            compound=compound,\n",
    "            ).save()\n",
    "    compound_perturbation_names[row[\"perturbation\"]] = compound_perturbation_name\n",
    "\n",
    "adata.obs[\"compound_perturbation\"] = adata.obs[\"compound\"].map(compound_perturbation_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": [
     "hide-output"
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning existing PerturbationTarget record with same name: 'MAPK/ERK pathway'\n",
      "\u001b[92m→\u001b[0m returning existing PerturbationTarget record with same name: 'Transcription regulation (BET proteins)'\n",
      "\u001b[92m→\u001b[0m returning existing PerturbationTarget record with same name: 'Apoptosis regulation'\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MTzGXCz6</td>\n",
       "      <td>PI3K/AKT/mTOR pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-12 17:27:25.458552+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                   name description  run_id  \\\n",
       "id                                                        \n",
       "10  MTzGXCz6  PI3K/AKT/mTOR pathway        None       1   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "10 2024-12-12 17:27:25.458552+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# optional: add protein targets to compounds\n",
    "compounds_to_targets = {\n",
    "    \"trametinib\": (\"MAPK/ERK pathway\", [\"P36507\"]),\n",
    "    \"afatinib\": (\"EGFR, HER2, HER4 signaling\", [\"P00533\", \"Q9UK79\", \"Q15303\"]),\n",
    "    \"dabrafenib\": (\"MAPK/ERK pathway\", [\"P15056\"]),\n",
    "    \"gemcitabine\": (\"DNA synthesis inhibition\", [\"P23921\"]),  # No single protein target\n",
    "    \"navitoclax\": (\"Apoptosis regulation\", [\"P10415\", \"Q07812\"]),\n",
    "    \"bortezomib\": (\"Proteasome pathway\", [\"P49721\"]),\n",
    "    \"brd3379\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"JQ1\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"azd5591\": (\"Apoptosis regulation\", [\"Q07820\"]),\n",
    "    \"prexasertib\": (\"DNA damage response\", [\"O14757\"]),\n",
    "    \"taselisib\": (\"PI3K/AKT/mTOR pathway\", [\"P42336\", \"O00329\", \"P48736\"]),\n",
    "    \"idasanutlin\": (\"p53 regulation\", [\"Q00987\"]),\n",
    "    \"everolimus\": (\"mTOR pathway\", [\"P42345\"])\n",
    "}\n",
    "\n",
    "for compound_name, compound_perturbation_name in compound_perturbation_names.items():\n",
    "    if compound_perturbation_name == \"control\":  # No target for control\n",
    "        continue\n",
    "    compound_perturbation = wl.CompoundPerturbation.get(name=compound_perturbation_name)\n",
    "    target = wl.PerturbationTarget(name=compounds_to_targets[compound_name][0]).save()\n",
    "    proteins = bt.Protein.from_values(compounds_to_targets[compound_name][1], field=bt.Protein.uniprotkb_id).save()\n",
    "    target.proteins.set(proteins)\n",
    "    compound_perturbation.targets.add(target)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m added default value 'unknown' to the adata.obs['assay']\n",
      "\u001b[92m→\u001b[0m added default value 'unknown' to the adata.obs['cell_type']\n",
      "\u001b[92m→\u001b[0m added default value 'unknown' to the adata.obs['development_stage']\n",
      "\u001b[92m→\u001b[0m added default value 'unknown' to the adata.obs['donor_id']\n",
      "\u001b[92m→\u001b[0m added default value 'unknown' to the adata.obs['self_reported_ethnicity']\n",
      "\u001b[92m→\u001b[0m added default value 'cell' to the adata.obs['suspension_type']\n",
      "\u001b[92m→\u001b[0m added default value '' to the adata.obs['environmental_perturbation']\n",
      "\u001b[92m→\u001b[0m added default value '' to the adata.obs['combination_perturbation']\n",
      "\u001b[92m✓\u001b[0m added 16 records with \u001b[3mFeature.name\u001b[0m for \"columns\": 'assay', 'cell_type', 'development_stage', 'disease', 'donor_id', 'self_reported_ethnicity', 'sex', 'suspension_type', 'tissue_type', 'organism', 'cell_line', 'genetic_perturbation', 'compound', 'compound_perturbation', 'environmental_perturbation', 'combination_perturbation'\n",
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n",
      "\u001b[92m✓\u001b[0m added 1869 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mGene.ensembl_gene_id\u001b[0m for \"var_index\": 'ENSG00000102316', 'ENSG00000109472', 'ENSG00000080007', 'ENSG00000203926', 'ENSG00000232301', 'ENSG00000127419', 'ENSG00000266869', 'ENSG00000108960', 'ENSG00000261316', 'ENSG00000126870', 'ENSG00000121797', 'ENSG00000243927', 'ENSG00000143473', 'ENSG00000261488', 'ENSG00000115665', 'ENSG00000180613', 'ENSG00000167283', 'ENSG00000160472', 'ENSG00000273387', 'ENSG00000110768', ...\n",
      "\u001b[94m•\u001b[0m saving validated records of 'disease'\n",
      "\u001b[92m✓\u001b[0m added 21 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mDisease.name\u001b[0m for \"disease\": 'prostate cancer', 'brain cancer', 'gastric cancer', 'esophageal cancer', 'skin cancer', 'thyroid cancer', 'uterine corpus cancer', 'colorectal cancer', 'malignant pancreatic neoplasm', 'bone cancer', 'sarcoma', 'breast cancer', 'urinary bladder carcinoma', 'bile duct cancer', 'kidney cancer', 'ovarian cancer', 'head and neck cancer', 'rhabdoid tumor', 'liver cancer', 'neuroblastoma', ...\n",
      "\u001b[94m•\u001b[0m saving validated records of 'sex'\n",
      "\u001b[92m✓\u001b[0m added 2 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mPhenotype.name\u001b[0m for \"sex\": 'female', 'male'\n",
      "\u001b[94m•\u001b[0m saving validated records of 'cell_line'\n",
      "\u001b[92m✓\u001b[0m added 183 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mCellLine.name\u001b[0m for \"cell_line\": 'NCI-H522', 'SW579', 'VM-CUB1', 'SW 480', 'RERF-LC-Ad2', 'SW 900', 'LMSU', 'NCI-H1581', 'KYSE-510', 'GAMG', 'OVK18', 'G-401', 'SK-MEL-24', 'SK-N-FI', 'KYSE-270', 'SBC-5', 'GCIY', 'TE-11', 'HEC-1-A', '42-MG-BA', ...\n",
      "\u001b[94m•\u001b[0m mapping \"var_index\" on \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m2 terms\u001b[0m are not validated: \u001b[1;91m'ENSG00000255823', 'ENSG00000272370'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from_var_index()\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"assay\" on \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"assay\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"cell_type\" on \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"cell_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"development_stage\" on \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"development_stage\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"disease\" on \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'pancreatic cancer'\u001b[0m\n",
      "    \u001b[1;93m1 synonym\u001b[0m found: \u001b[1;93m\"pancreatic cancer\" → \"malignant pancreatic neoplasm\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"disease\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"donor_id\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"donor_id\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"self_reported_ethnicity\" on \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"self_reported_ethnicity\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"sex\" on \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m3 terms\u001b[0m are not validated: \u001b[1;91m'Male', 'Female', 'Unknown'\u001b[0m\n",
      "    \u001b[1;93m2 synonyms\u001b[0m found: \u001b[1;93m\"Male\" → \"male\", \"Female\" → \"female\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"sex\")\u001b[0m    for remaining terms:\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"sex\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"suspension_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"suspension_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"tissue_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell_line'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"tissue_type\")\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"organism\" is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_line\" is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"genetic_perturbation\" is validated against \u001b[3mGeneticPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound\" is validated against \u001b[3mCompound.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound_perturbation\" is validated against \u001b[3mCompoundPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"environmental_perturbation\" is validated against \u001b[3mEnvironmentalPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"combination_perturbation\" is validated against \u001b[3mCombinationPerturbation.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator = pts.PerturbationCurator(adata, using_key=\"test-perturbation\")\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually fix sex and set assay\n",
    "adata.obs[\"sex\"] = adata.obs[\"sex\"].cat.rename_categories({\"Unknown\": \"unknown\"})\n",
    "adata.obs[\"assay\"] = \"10x 3' v3\"\n",
    "\n",
    "# subset the adata to only include the validated genes\n",
    "adata = adata[:, ~adata.var_names.isin(curator.non_validated[\"var_index\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'assay'\n",
      "\u001b[92m✓\u001b[0m added 1 record \u001b[1;92mfrom public\u001b[0m with \u001b[3mExperimentalFactor.name\u001b[0m for \"assay\": '10x 3' v3'\n",
      "\u001b[92m✓\u001b[0m \"var_index\" is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"assay\" is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"cell_type\" on \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"cell_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"development_stage\" on \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"development_stage\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"disease\" on \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'pancreatic cancer'\u001b[0m\n",
      "    \u001b[1;93m1 synonym\u001b[0m found: \u001b[1;93m\"pancreatic cancer\" → \"malignant pancreatic neoplasm\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"disease\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"donor_id\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"donor_id\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"self_reported_ethnicity\" on \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"self_reported_ethnicity\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"sex\" on \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m3 terms\u001b[0m are not validated: \u001b[1;91m'Male', 'Female', 'unknown'\u001b[0m\n",
      "    \u001b[1;93m2 synonyms\u001b[0m found: \u001b[1;93m\"Male\" → \"male\", \"Female\" → \"female\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"sex\")\u001b[0m    for remaining terms:\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"sex\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"suspension_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"suspension_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"tissue_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell_line'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"tissue_type\")\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"organism\" is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_line\" is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"genetic_perturbation\" is validated against \u001b[3mGeneticPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound\" is validated against \u001b[3mCompound.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound_perturbation\" is validated against \u001b[3mCompoundPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"environmental_perturbation\" is validated against \u001b[3mEnvironmentalPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"combination_perturbation\" is validated against \u001b[3mCombinationPerturbation.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Recreate Curator object because we are using a new adata\n",
    "curator = pts.PerturbationCurator(adata, using_key=\"test-perturbation\")\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m standardized 1 synonym in \"disease\": \u001b[1;92m\"pancreatic cancer\" → \"malignant pancreatic neoplasm\"\u001b[0m\n",
      "\u001b[92m✓\u001b[0m standardized 2 synonyms in \"sex\": \u001b[1;92m\"Male\" → \"male\", \"Female\" → \"female\"\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "curator.standardize(\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mCellType.name\u001b[0m for \"cell_type\": 'unknown'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mDevelopmentalStage.name\u001b[0m for \"development_stage\": 'unknown'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mULabel.name\u001b[0m for \"donor_id\": 'unknown'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mEthnicity.name\u001b[0m for \"self_reported_ethnicity\": 'unknown'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mPhenotype.name\u001b[0m for \"sex\": 'unknown'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mULabel.name\u001b[0m for \"suspension_type\": 'cell'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mULabel.name\u001b[0m for \"tissue_type\": 'cell_line'\n"
     ]
    }
   ],
   "source": [
    "curator.add_new_from(\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"var_index\" is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"assay\" is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_type\" is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"development_stage\" is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"disease\" is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"donor_id\" is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"self_reported_ethnicity\" is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"sex\" is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"suspension_type\" is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"tissue_type\" is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"organism\" is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_line\" is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"genetic_perturbation\" is validated against \u001b[3mGeneticPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound\" is validated against \u001b[3mCompound.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound_perturbation\" is validated against \u001b[3mCompoundPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"environmental_perturbation\" is validated against \u001b[3mEnvironmentalPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"combination_perturbation\" is validated against \u001b[3mCombinationPerturbation.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator.validate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = ops.Reference(\n",
    "    name=\"Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action\",\n",
    "    abbr=\"McFarland 2020\",\n",
    "    url=\"https://www.nature.com/articles/s41467-020-17440-w\",\n",
    "    doi=\"10.1038/s41467-020-17440-w\",\n",
    "    text=(\n",
    "        \"Assays to study cancer cell responses to pharmacologic or genetic perturbations are typically \"\n",
    "        \"restricted to using simple phenotypic readouts such as proliferation rate. Information-rich assays, \"\n",
    "        \"such as gene-expression profiling, have generally not permitted efficient profiling of a given \"\n",
    "        \"perturbation across multiple cellular contexts. Here, we develop MIX-Seq, a method for multiplexed \"\n",
    "        \"transcriptional profiling of post-perturbation responses across a mixture of samples with single-cell \"\n",
    "        \"resolution, using SNP-based computational demultiplexing of single-cell RNA-sequencing data. We show \"\n",
    "        \"that MIX-Seq can be used to profile responses to chemical or genetic perturbations across pools of 100 \"\n",
    "        \"or more cancer cell lines. We combine it with Cell Hashing to further multiplex additional experimental \"\n",
    "        \"conditions, such as post-treatment time points or drug doses. Analyzing the high-content readout of \"\n",
    "        \"scRNA-seq reveals both shared and context-specific transcriptional response components that can identify \"\n",
    "        \"drug mechanism of action and enable prediction of long-term cell viability from short-term transcriptional \"\n",
    "        \"responses to treatment.\"\n",
    "    ),\n",
    ").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Register curated artifact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m    \u001b[1;93m30 unique terms\u001b[0m (65.20%) are not validated for \u001b[3mname\u001b[0m: \u001b[1;93mdepmap_id, cancer, cell_det_rate, cell_quality, channel, dose_unit, dose_value, doublet_CL1, doublet_CL2, doublet_GMM_prob, doublet_dev_imp, doublet_z_margin, hash_assignment, hash_tag, num_SNPs, perturbation, perturbation_type, singlet_ID, singlet_dev, singlet_dev_z, ...\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91mdid not create\u001b[0m Feature records for \u001b[1;93m30 non-validated\u001b[0m \u001b[3mnames\u001b[0m: \u001b[1;93m'cancer', 'cell_det_rate', 'cell_quality', 'channel', 'chembl-ID', 'depmap_id', 'dose_unit', 'dose_value', 'doublet_CL1', 'doublet_CL2', 'doublet_GMM_prob', 'doublet_dev_imp', 'doublet_z_margin', 'hash_assignment', 'hash_tag', 'ncounts', 'ngenes', 'nperts', 'num_SNPs', 'percent_mito', ...\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "artifact = curator.save_artifact(description=\"McFarland AnnData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# link the reference to the artifact\n",
    "artifact.references.add(reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">Artifact</span><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f; font-weight: bold\"> .h5ad/AnnData</span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">├── </span><span style=\"color: #00ffff; text-decoration-color: #00ffff; font-weight: bold\">General</span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span>.uid = 'jcBPXRit3gRZ1xvb0000'\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span>.size = 3500720\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span>.hash = '_5KilLlHszvZF60lInz57w'\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span>.n_observations = 1000\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span>.path = \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">/Users/sunnysun/Documents/repos.nosync/scverse/pertpy-datasets/scripts/lamindb_datasets/test-perturbation</span>/.\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>lamindb/jcBPXRit3gRZ1xvb0000.h5ad\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span>.created_by = sunnyosun (Sunny Sun)\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span>.created_at = <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">2024-12-12 18:30:18</span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   └── </span><span style=\"color: #00d7af; text-decoration-color: #00d7af\">.transform = 'Curate perturbation dataset with `PerturbationCurator`'</span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">├── </span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff; font-weight: bold\">Dataset featuers</span><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">/</span><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f; font-weight: bold\">.feature_sets</span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   ├── </span><span style=\"color: #d787ff; text-decoration-color: #d787ff; font-weight: bold\">var</span><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f; font-weight: bold\"> • </span><span style=\"color: #ffafd7; text-decoration-color: #ffafd7; font-weight: bold\">1869</span><span style=\"font-weight: bold\">                  </span><span style=\"color: #ffafd7; text-decoration-color: #ffafd7; font-weight: bold\">[bionty.Gene]</span><span style=\"font-weight: bold\">                                                       </span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>MAGED2                     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>CPE                        <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>DDX43                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>SPANXA2                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>LNCPRESS1                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>TMEM175                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>MMD                        <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>LINC01834                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>DYNC2I1                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>CCRL2                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>MRPS6                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>KCNH1                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>TBILA                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>SLC5A7                     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>GSX2                       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>ATP5MG                     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>TMEM190                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   │   </span>GTF2H1                     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> float                     </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│   └── </span><span style=\"color: #d787ff; text-decoration-color: #d787ff; font-weight: bold\">obs</span><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f; font-weight: bold\"> • </span><span style=\"color: #ffafd7; text-decoration-color: #ffafd7; font-weight: bold\">16</span><span style=\"font-weight: bold\">                    </span><span style=\"color: #ffafd7; text-decoration-color: #ffafd7; font-weight: bold\">[Feature]</span><span style=\"font-weight: bold\">                                                           </span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>assay                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.ExperimentalF… </span> 10x 3' v3                                \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>cell_line                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.CellLine]      </span> 22Rv1, 253J-BV, 42-MG-BA, 639-V, 647-V, …\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>cell_type                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.CellType]      </span> unknown                                  \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>compound                   <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[wetlab.Compound]      </span> JQ1, afatinib, azd5591, bortezomib, brd3…\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>compound_perturbation      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[wetlab.CompoundPertu… </span> JQ1_2.5µM_24h, afatinib_0.5µM_24h, azd55…\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>development_stage          <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.Developmental… </span> unknown                                  \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>disease                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.Disease]       </span> bile duct cancer, bone cancer, brain can…\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>donor_id                   <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[ULabel]               </span> unknown                                  \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>genetic_perturbation       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[wetlab.GeneticPertur… </span> sggpx4-1, sggpx4-2, sglacz, sgor2j2      \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>organism                   <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.Organism]      </span> human                                    \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>self_reported_ethnicity    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.Ethnicity]     </span> unknown                                  \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>sex                        <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[bionty.Phenotype]     </span> female, male, unknown                    \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>suspension_type            <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[ULabel]               </span> cell                                     \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>tissue_type                <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat[ULabel]               </span> cell_line                                \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>environmental_perturbation <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat                       </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">│       </span>combination_perturbation   <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> cat                       </span>                                          \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">└── </span><span style=\"color: #afff00; text-decoration-color: #afff00; font-weight: bold\">Labels</span>\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">    └── </span>.references                <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> ourprojects.Reference     </span> Multiplexed single-cell transcriptional …\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.organisms                 <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.Organism           </span> human                                    \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.cell_types                <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.CellType           </span> unknown                                  \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.diseases                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.Disease            </span> prostate cancer, brain cancer, gastric c…\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.cell_lines                <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.CellLine           </span> NCI-H522, SW579, VM-CUB1, SW 480, RERF-L…\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.phenotypes                <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.Phenotype          </span> female, male, unknown                    \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.experimental_factors      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.ExperimentalFactor </span> 10x 3' v3                                \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.developmental_stages      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.DevelopmentalStage </span> unknown                                  \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.ethnicities               <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> bionty.Ethnicity          </span> unknown                                  \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.compounds                 <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> wetlab.Compound           </span> JQ1, afatinib, azd5591, bortezomib, brd3…\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.genetic_perturbations     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> wetlab.GeneticPerturbati… </span> sggpx4-1, sggpx4-2, sglacz, sgor2j2      \n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.compound_perturbations    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> wetlab.CompoundPerturbat… </span> JQ1_2.5µM_24h, afatinib_0.5µM_24h, azd55…\n",
       "<span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">        </span>.ulabels                   <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\"> ULabel                    </span> unknown, cell, cell_line                 \n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1mArtifact\u001b[0m\u001b[1;2m .h5ad/AnnData\u001b[0m\n",
       "\u001b[2m├── \u001b[0m\u001b[1mGeneral\u001b[0m\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m.uid = 'jcBPXRit3gRZ1xvb0000'\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m.size = 3500720\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m.hash = '_5KilLlHszvZF60lInz57w'\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m.n_observations = 1000\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m.path = \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0m\u001b[2m/Users/sunnysun/Documents/repos.nosync/scverse/pertpy-datasets/scripts/lamindb_datasets/test-perturbation\u001b[0m/.\n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mlamindb/jcBPXRit3gRZ1xvb0000.h5ad\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m.created_by = sunnyosun (Sunny Sun)\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m.created_at = \u001b[2m2024-12-12 18:30:18\u001b[0m\n",
       "\u001b[2m│   \u001b[0m\u001b[2m└── \u001b[0m.transform = 'Curate perturbation dataset with `PerturbationCurator`'\n",
       "\u001b[2m├── \u001b[0m\u001b[1mDataset featuers\u001b[0m\u001b[2m/\u001b[0m\u001b[1;2m.feature_sets\u001b[0m\n",
       "\u001b[2m│   \u001b[0m\u001b[2m├── \u001b[0m\u001b[1mvar\u001b[0m\u001b[1;2m • \u001b[0m\u001b[1m1869\u001b[0m\u001b[1m                \u001b[0m\u001b[1m \u001b[0m\u001b[1m \u001b[0m\u001b[1m[bionty.Gene]\u001b[0m\u001b[1m            \u001b[0m\u001b[1m \u001b[0m\u001b[1m \u001b[0m\u001b[1m                                         \u001b[0m\n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mMAGED2                     \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mCPE                        \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mDDX43                      \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mSPANXA2                    \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mLNCPRESS1                  \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mTMEM175                    \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mMMD                        \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mLINC01834                  \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mDYNC2I1                    \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mCCRL2                      \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mMRPS6                      \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mKCNH1                      \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mTBILA                      \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mSLC5A7                     \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mGSX2                       \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mATP5MG                     \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mTMEM190                    \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m│   \u001b[0mGTF2H1                     \u001b[2m \u001b[0m\u001b[2mfloat                    \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m└── \u001b[0m\u001b[1mobs\u001b[0m\u001b[1;2m • \u001b[0m\u001b[1m16\u001b[0m\u001b[1m                  \u001b[0m\u001b[1m \u001b[0m\u001b[1m \u001b[0m\u001b[1m[Feature]\u001b[0m\u001b[1m                \u001b[0m\u001b[1m \u001b[0m\u001b[1m \u001b[0m\u001b[1m                                         \u001b[0m\n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0massay                      \u001b[2m \u001b[0m\u001b[2mcat[bionty.ExperimentalF…\u001b[0m\u001b[2m \u001b[0m 10x 3' v3                                \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mcell_line                  \u001b[2m \u001b[0m\u001b[2mcat[bionty.CellLine]     \u001b[0m\u001b[2m \u001b[0m 22Rv1, 253J-BV, 42-MG-BA, 639-V, 647-V, …\n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mcell_type                  \u001b[2m \u001b[0m\u001b[2mcat[bionty.CellType]     \u001b[0m\u001b[2m \u001b[0m unknown                                  \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mcompound                   \u001b[2m \u001b[0m\u001b[2mcat[wetlab.Compound]     \u001b[0m\u001b[2m \u001b[0m JQ1, afatinib, azd5591, bortezomib, brd3…\n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mcompound_perturbation      \u001b[2m \u001b[0m\u001b[2mcat[wetlab.CompoundPertu…\u001b[0m\u001b[2m \u001b[0m JQ1_2.5µM_24h, afatinib_0.5µM_24h, azd55…\n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mdevelopment_stage          \u001b[2m \u001b[0m\u001b[2mcat[bionty.Developmental…\u001b[0m\u001b[2m \u001b[0m unknown                                  \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mdisease                    \u001b[2m \u001b[0m\u001b[2mcat[bionty.Disease]      \u001b[0m\u001b[2m \u001b[0m bile duct cancer, bone cancer, brain can…\n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mdonor_id                   \u001b[2m \u001b[0m\u001b[2mcat[ULabel]              \u001b[0m\u001b[2m \u001b[0m unknown                                  \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mgenetic_perturbation       \u001b[2m \u001b[0m\u001b[2mcat[wetlab.GeneticPertur…\u001b[0m\u001b[2m \u001b[0m sggpx4-1, sggpx4-2, sglacz, sgor2j2      \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0morganism                   \u001b[2m \u001b[0m\u001b[2mcat[bionty.Organism]     \u001b[0m\u001b[2m \u001b[0m human                                    \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mself_reported_ethnicity    \u001b[2m \u001b[0m\u001b[2mcat[bionty.Ethnicity]    \u001b[0m\u001b[2m \u001b[0m unknown                                  \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0msex                        \u001b[2m \u001b[0m\u001b[2mcat[bionty.Phenotype]    \u001b[0m\u001b[2m \u001b[0m female, male, unknown                    \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0msuspension_type            \u001b[2m \u001b[0m\u001b[2mcat[ULabel]              \u001b[0m\u001b[2m \u001b[0m cell                                     \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mtissue_type                \u001b[2m \u001b[0m\u001b[2mcat[ULabel]              \u001b[0m\u001b[2m \u001b[0m cell_line                                \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0menvironmental_perturbation \u001b[2m \u001b[0m\u001b[2mcat                      \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m│   \u001b[0m\u001b[2m    \u001b[0mcombination_perturbation   \u001b[2m \u001b[0m\u001b[2mcat                      \u001b[0m\u001b[2m \u001b[0m                                          \n",
       "\u001b[2m└── \u001b[0m\u001b[1mLabels\u001b[0m\n",
       "\u001b[2m    \u001b[0m\u001b[2m└── \u001b[0m.references                \u001b[2m \u001b[0m\u001b[2mourprojects.Reference    \u001b[0m\u001b[2m \u001b[0m Multiplexed single-cell transcriptional …\n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.organisms                 \u001b[2m \u001b[0m\u001b[2mbionty.Organism          \u001b[0m\u001b[2m \u001b[0m human                                    \n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.cell_types                \u001b[2m \u001b[0m\u001b[2mbionty.CellType          \u001b[0m\u001b[2m \u001b[0m unknown                                  \n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.diseases                  \u001b[2m \u001b[0m\u001b[2mbionty.Disease           \u001b[0m\u001b[2m \u001b[0m prostate cancer, brain cancer, gastric c…\n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.cell_lines                \u001b[2m \u001b[0m\u001b[2mbionty.CellLine          \u001b[0m\u001b[2m \u001b[0m NCI-H522, SW579, VM-CUB1, SW 480, RERF-L…\n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.phenotypes                \u001b[2m \u001b[0m\u001b[2mbionty.Phenotype         \u001b[0m\u001b[2m \u001b[0m female, male, unknown                    \n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.experimental_factors      \u001b[2m \u001b[0m\u001b[2mbionty.ExperimentalFactor\u001b[0m\u001b[2m \u001b[0m 10x 3' v3                                \n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.developmental_stages      \u001b[2m \u001b[0m\u001b[2mbionty.DevelopmentalStage\u001b[0m\u001b[2m \u001b[0m unknown                                  \n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.ethnicities               \u001b[2m \u001b[0m\u001b[2mbionty.Ethnicity         \u001b[0m\u001b[2m \u001b[0m unknown                                  \n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.compounds                 \u001b[2m \u001b[0m\u001b[2mwetlab.Compound          \u001b[0m\u001b[2m \u001b[0m JQ1, afatinib, azd5591, bortezomib, brd3…\n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.genetic_perturbations     \u001b[2m \u001b[0m\u001b[2mwetlab.GeneticPerturbati…\u001b[0m\u001b[2m \u001b[0m sggpx4-1, sggpx4-2, sglacz, sgor2j2      \n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.compound_perturbations    \u001b[2m \u001b[0m\u001b[2mwetlab.CompoundPerturbat…\u001b[0m\u001b[2m \u001b[0m JQ1_2.5µM_24h, afatinib_0.5µM_24h, azd55…\n",
       "\u001b[2m    \u001b[0m\u001b[2m    \u001b[0m.ulabels                   \u001b[2m \u001b[0m\u001b[2mULabel                   \u001b[0m\u001b[2m \u001b[0m unknown, cell, cell_line                 \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "artifact.describe()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py310",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
