{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curate perturbation dataset with `PerturbationCurator`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we use pertpy's `PerturbationCurator` to ensure that a perturbatin dataset conforms to both, `CELLxGENE` (schema 5.1.0) and pertpy's defined criteria.\n",
    "More specifically, the `PerturbationCurator` builds upon [cellxgene-lamin](https://github.com/laminlabs/cellxgene-lamin) and extends it by further requiring `cell_line` and `X_treatments` columns for the perturbations.\n",
    "\n",
    "This guide demonstrates how to curate a complex, real world perturbation dataset [McFarland et al. 2020](https://www.nature.com/articles/s41467-020-17440-w) using `PerturbationCurator`. Please have a look at [lamindb's perturbation guide](https://docs.lamin.ai/perturbation) for more details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install pertpy-datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: zethson/test-perturbation\n"
     ]
    }
   ],
   "source": [
    "# Using a local instance here but in practice, we use `laminlabs/pertpy-datasets`\n",
    "!lamin init --storage ./test-perturbation --schema bionty,wetlab,findrefs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: zethson/test-perturbation\n"
     ]
    }
   ],
   "source": [
    "import lamindb as ln\n",
    "import bionty as bt\n",
    "import wetlab as wl\n",
    "import findrefs as fr\n",
    "import pertpy_datasets as pts\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m notebook imports: bionty==0.52.0 findrefs==0.1.0 lamindb==0.76.15 pertpy_datasets==0.1.0 scanpy==1.10.3 wetlab==0.34.0\n",
      "\u001b[92m→\u001b[0m loaded Transform('HIRTYxL3'), started Run('RGyUX0tz') at 2024-11-11 15:10:01 UTC\n"
     ]
    }
   ],
   "source": [
    "ln.track(\"HIRTYxL3aZc70000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m completing transfer to track Artifact('Xk7Qaik9') as input\n",
      "\u001b[92m→\u001b[0m mapped records: Artifact(uid='Xk7Qaik9vBLV4PKf0001')\n",
      "\u001b[92m→\u001b[0m transferred records: \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cell_det_rate</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_quality</th>\n",
       "      <th>channel</th>\n",
       "      <th>disease</th>\n",
       "      <th>dose_unit</th>\n",
       "      <th>dose_value</th>\n",
       "      <th>doublet_CL1</th>\n",
       "      <th>...</th>\n",
       "      <th>singlet_z_margin</th>\n",
       "      <th>time</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>tot_reads</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>chembl-ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACTGGTGTCTCTCTG</th>\n",
       "      <td>ACH-000390</td>\n",
       "      <td>True</td>\n",
       "      <td>0.093159</td>\n",
       "      <td>LUDLU-1</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>12.351139</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>787</td>\n",
       "      <td>1</td>\n",
       "      <td>3045</td>\n",
       "      <td>12895.0</td>\n",
       "      <td>3.202792</td>\n",
       "      <td>24.955409</td>\n",
       "      <td>CHEMBL2103875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAGGCTCAGATTTCG</th>\n",
       "      <td>ACH-000444</td>\n",
       "      <td>True</td>\n",
       "      <td>0.145728</td>\n",
       "      <td>LU99</td>\n",
       "      <td>normal</td>\n",
       "      <td>2</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.5</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>8.164565</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1597</td>\n",
       "      <td>1</td>\n",
       "      <td>4763</td>\n",
       "      <td>23161.0</td>\n",
       "      <td>7.473771</td>\n",
       "      <td>18.051898</td>\n",
       "      <td>CHEMBL1173655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCCAAATCAAGCCGTC</th>\n",
       "      <td>ACH-000396</td>\n",
       "      <td>True</td>\n",
       "      <td>0.117330</td>\n",
       "      <td>J82</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>urinary bladder carcinoma</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>...</td>\n",
       "      <td>11.188513</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1159</td>\n",
       "      <td>1</td>\n",
       "      <td>3834</td>\n",
       "      <td>18062.0</td>\n",
       "      <td>2.762706</td>\n",
       "      <td>22.085040</td>\n",
       "      <td>CHEMBL2028663</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   depmap_id  cancer  cell_det_rate cell_line cell_quality  \\\n",
       "AACTGGTGTCTCTCTG  ACH-000390    True       0.093159   LUDLU-1       normal   \n",
       "ATAGGCTCAGATTTCG  ACH-000444    True       0.145728      LU99       normal   \n",
       "GCCAAATCAAGCCGTC  ACH-000396    True       0.117330       J82       normal   \n",
       "\n",
       "                 channel                    disease dose_unit  dose_value  \\\n",
       "AACTGGTGTCTCTCTG     nan                lung cancer        µM         0.1   \n",
       "ATAGGCTCAGATTTCG       2                lung cancer        µM         0.5   \n",
       "GCCAAATCAAGCCGTC     nan  urinary bladder carcinoma        µM         0.1   \n",
       "\n",
       "                        doublet_CL1  ... singlet_z_margin  time  tissue_type  \\\n",
       "AACTGGTGTCTCTCTG        LUDLU1_LUNG  ...        12.351139    24    cell_line   \n",
       "ATAGGCTCAGATTTCG          LU99_LUNG  ...         8.164565    24    cell_line   \n",
       "GCCAAATCAAGCCGTC  J82_URINARY_TRACT  ...        11.188513    24    cell_line   \n",
       "\n",
       "                  tot_reads nperts ngenes  ncounts percent_mito percent_ribo  \\\n",
       "AACTGGTGTCTCTCTG        787      1   3045  12895.0     3.202792    24.955409   \n",
       "ATAGGCTCAGATTTCG       1597      1   4763  23161.0     7.473771    18.051898   \n",
       "GCCAAATCAAGCCGTC       1159      1   3834  18062.0     2.762706    22.085040   \n",
       "\n",
       "                      chembl-ID  \n",
       "AACTGGTGTCTCTCTG  CHEMBL2103875  \n",
       "ATAGGCTCAGATTTCG  CHEMBL1173655  \n",
       "GCCAAATCAAGCCGTC  CHEMBL2028663  \n",
       "\n",
       "[3 rows x 35 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = ln.Artifact.using(\"laminlabs/lamindata\").get(uid=\"Xk7Qaik9vBLV4PKf0001\").load()\n",
    "adata.obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate an embedding because CELLxGENE requires one\n",
    "sc.tl.pca(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curator non-perturbation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m added defaults to the AnnData object: {'assay': 'unknown', 'cell_type': 'unknown', 'development_stage': 'unknown', 'donor_id': 'unknown', 'self_reported_ethnicity': 'unknown', 'suspension_type': 'cell', 'genetic_treatments': '', 'compound_treatments': '', 'environmental_treatments': '', 'combination_treatments': ''}\n",
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mvar_index\u001b[0m on \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m2 terms\u001b[0m are not validated: \u001b[1;91m'ENSG00000255823', 'ENSG00000272370'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from_var_index()\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3massay\u001b[0m on \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "→ fix typo, remove non-existent value, or save term via \u001b[1;91m.add_new_from('assay')\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_type' is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'development_stage' is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'disease' is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'donor_id' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'self_reported_ethnicity' is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3msex\u001b[0m on \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    found \u001b[1;93m1\u001b[0m validated term: \u001b[1;93m'unknown'\u001b[0m\n",
      "→ save term via \u001b[1;93m.add_validated_from('sex')\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'Unknown'\u001b[0m\n",
      "→ fix typo, remove non-existent value, or save term via \u001b[1;91m.add_new_from('sex')\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'suspension_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'tissue_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'organism' is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_line' is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'genetic_treatments' is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'compound_treatments' is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'environmental_treatments' is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'combination_treatments' is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator = pts.PerturbationCurator(\n",
    "    adata\n",
    ")  # Fetch all ontologies from this instance\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cell_det_rate</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_quality</th>\n",
       "      <th>channel</th>\n",
       "      <th>disease</th>\n",
       "      <th>dose_unit</th>\n",
       "      <th>dose_value</th>\n",
       "      <th>doublet_CL1</th>\n",
       "      <th>...</th>\n",
       "      <th>assay</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>genetic_treatments</th>\n",
       "      <th>compound_treatments</th>\n",
       "      <th>environmental_treatments</th>\n",
       "      <th>combination_treatments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACTGGTGTCTCTCTG</th>\n",
       "      <td>ACH-000390</td>\n",
       "      <td>True</td>\n",
       "      <td>0.093159</td>\n",
       "      <td>LUDLU-1</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAGGCTCAGATTTCG</th>\n",
       "      <td>ACH-000444</td>\n",
       "      <td>True</td>\n",
       "      <td>0.145728</td>\n",
       "      <td>LU99</td>\n",
       "      <td>normal</td>\n",
       "      <td>2</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.5</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCCAAATCAAGCCGTC</th>\n",
       "      <td>ACH-000396</td>\n",
       "      <td>True</td>\n",
       "      <td>0.117330</td>\n",
       "      <td>J82</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>urinary bladder carcinoma</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGGAGAAGTCGCGTCA</th>\n",
       "      <td>ACH-000997</td>\n",
       "      <td>True</td>\n",
       "      <td>0.005422</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>low_quality</td>\n",
       "      <td>7</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>HCT15_LARGE_INTESTINE</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTTGGAGATCGATA</th>\n",
       "      <td>ACH-000723</td>\n",
       "      <td>True</td>\n",
       "      <td>0.132708</td>\n",
       "      <td>YD-10B</td>\n",
       "      <td>low_quality</td>\n",
       "      <td>nan</td>\n",
       "      <td>head and neck cancer</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YD10B_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATCTAGTCATCGGAT</th>\n",
       "      <td>ACH-000015</td>\n",
       "      <td>True</td>\n",
       "      <td>0.094752</td>\n",
       "      <td>NCI-H1581</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>NCIH1581_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACTCTTAGTCCCACG</th>\n",
       "      <td>ACH-000252</td>\n",
       "      <td>True</td>\n",
       "      <td>0.069540</td>\n",
       "      <td>LS1034</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LS1034_LARGE_INTESTINE</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACTGTCCAGTCACGC</th>\n",
       "      <td>ACH-000681</td>\n",
       "      <td>True</td>\n",
       "      <td>0.094262</td>\n",
       "      <td>A549</td>\n",
       "      <td>normal</td>\n",
       "      <td>5</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>2.5</td>\n",
       "      <td>A549_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACCTTGGTCGACTAT</th>\n",
       "      <td>ACH-000875</td>\n",
       "      <td>True</td>\n",
       "      <td>0.163557</td>\n",
       "      <td>NCI-H2347</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NCIH2347_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGTAAGG</th>\n",
       "      <td>ACH-000605</td>\n",
       "      <td>True</td>\n",
       "      <td>0.127501</td>\n",
       "      <td>TE-6</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>esophageal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TE6_OESOPHAGUS</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   depmap_id  cancer  cell_det_rate  cell_line cell_quality  \\\n",
       "AACTGGTGTCTCTCTG  ACH-000390    True       0.093159    LUDLU-1       normal   \n",
       "ATAGGCTCAGATTTCG  ACH-000444    True       0.145728       LU99       normal   \n",
       "GCCAAATCAAGCCGTC  ACH-000396    True       0.117330        J82       normal   \n",
       "CGGAGAAGTCGCGTCA  ACH-000997    True       0.005422     HCT-15  low_quality   \n",
       "TAGTTGGAGATCGATA  ACH-000723    True       0.132708     YD-10B  low_quality   \n",
       "...                      ...     ...            ...        ...          ...   \n",
       "GATCTAGTCATCGGAT  ACH-000015    True       0.094752  NCI-H1581       normal   \n",
       "AACTCTTAGTCCCACG  ACH-000252    True       0.069540     LS1034       normal   \n",
       "CACTGTCCAGTCACGC  ACH-000681    True       0.094262       A549       normal   \n",
       "CACCTTGGTCGACTAT  ACH-000875    True       0.163557  NCI-H2347       normal   \n",
       "TTTGTCACACGTAAGG  ACH-000605    True       0.127501       TE-6       normal   \n",
       "\n",
       "                 channel                    disease dose_unit  dose_value  \\\n",
       "AACTGGTGTCTCTCTG     nan                lung cancer        µM         0.1   \n",
       "ATAGGCTCAGATTTCG       2                lung cancer        µM         0.5   \n",
       "GCCAAATCAAGCCGTC     nan  urinary bladder carcinoma        µM         0.1   \n",
       "CGGAGAAGTCGCGTCA       7          colorectal cancer        µM         0.1   \n",
       "TAGTTGGAGATCGATA     nan       head and neck cancer       nan         NaN   \n",
       "...                  ...                        ...       ...         ...   \n",
       "GATCTAGTCATCGGAT     nan                lung cancer        µM         0.1   \n",
       "AACTCTTAGTCCCACG     nan          colorectal cancer        µM         0.0   \n",
       "CACTGTCCAGTCACGC       5                lung cancer        µM         2.5   \n",
       "CACCTTGGTCGACTAT     nan                lung cancer        µM         0.0   \n",
       "TTTGTCACACGTAAGG     nan          esophageal cancer        µM         0.0   \n",
       "\n",
       "                                      doublet_CL1  ...    assay  cell_type  \\\n",
       "AACTGGTGTCTCTCTG                      LUDLU1_LUNG  ...  unknown    unknown   \n",
       "ATAGGCTCAGATTTCG                        LU99_LUNG  ...  unknown    unknown   \n",
       "GCCAAATCAAGCCGTC                J82_URINARY_TRACT  ...  unknown    unknown   \n",
       "CGGAGAAGTCGCGTCA            HCT15_LARGE_INTESTINE  ...  unknown    unknown   \n",
       "TAGTTGGAGATCGATA  YD10B_UPPER_AERODIGESTIVE_TRACT  ...  unknown    unknown   \n",
       "...                                           ...  ...      ...        ...   \n",
       "GATCTAGTCATCGGAT                    NCIH1581_LUNG  ...  unknown    unknown   \n",
       "AACTCTTAGTCCCACG           LS1034_LARGE_INTESTINE  ...  unknown    unknown   \n",
       "CACTGTCCAGTCACGC                        A549_LUNG  ...  unknown    unknown   \n",
       "CACCTTGGTCGACTAT                    NCIH2347_LUNG  ...  unknown    unknown   \n",
       "TTTGTCACACGTAAGG                   TE6_OESOPHAGUS  ...  unknown    unknown   \n",
       "\n",
       "                  development_stage  donor_id self_reported_ethnicity  \\\n",
       "AACTGGTGTCTCTCTG            unknown   unknown                 unknown   \n",
       "ATAGGCTCAGATTTCG            unknown   unknown                 unknown   \n",
       "GCCAAATCAAGCCGTC            unknown   unknown                 unknown   \n",
       "CGGAGAAGTCGCGTCA            unknown   unknown                 unknown   \n",
       "TAGTTGGAGATCGATA            unknown   unknown                 unknown   \n",
       "...                             ...       ...                     ...   \n",
       "GATCTAGTCATCGGAT            unknown   unknown                 unknown   \n",
       "AACTCTTAGTCCCACG            unknown   unknown                 unknown   \n",
       "CACTGTCCAGTCACGC            unknown   unknown                 unknown   \n",
       "CACCTTGGTCGACTAT            unknown   unknown                 unknown   \n",
       "TTTGTCACACGTAAGG            unknown   unknown                 unknown   \n",
       "\n",
       "                 suspension_type  genetic_treatments compound_treatments  \\\n",
       "AACTGGTGTCTCTCTG            cell                                           \n",
       "ATAGGCTCAGATTTCG            cell                                           \n",
       "GCCAAATCAAGCCGTC            cell                                           \n",
       "CGGAGAAGTCGCGTCA            cell                                           \n",
       "TAGTTGGAGATCGATA            cell                                           \n",
       "...                          ...                 ...                 ...   \n",
       "GATCTAGTCATCGGAT            cell                                           \n",
       "AACTCTTAGTCCCACG            cell                                           \n",
       "CACTGTCCAGTCACGC            cell                                           \n",
       "CACCTTGGTCGACTAT            cell                                           \n",
       "TTTGTCACACGTAAGG            cell                                           \n",
       "\n",
       "                 environmental_treatments combination_treatments  \n",
       "AACTGGTGTCTCTCTG                                                  \n",
       "ATAGGCTCAGATTTCG                                                  \n",
       "GCCAAATCAAGCCGTC                                                  \n",
       "CGGAGAAGTCGCGTCA                                                  \n",
       "TAGTTGGAGATCGATA                                                  \n",
       "...                                   ...                    ...  \n",
       "GATCTAGTCATCGGAT                                                  \n",
       "AACTCTTAGTCCCACG                                                  \n",
       "CACTGTCCAGTCACGC                                                  \n",
       "CACCTTGGTCGACTAT                                                  \n",
       "TTTGTCACACGTAAGG                                                  \n",
       "\n",
       "[1000 rows x 45 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_48515/2705999714.py:1: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.\n",
      "  adata.obs[\"sex\"] = adata.obs[\"sex\"].replace({\"Unknown\": \"unknown\"})\n"
     ]
    }
   ],
   "source": [
    "adata.obs[\"sex\"] = adata.obs[\"sex\"].replace({\"Unknown\": \"unknown\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "efo_lo = bt.ExperimentalFactor.public().lookup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"assay\"] = efo_lo.single_cell_rna_sequencing.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, ~adata.var_names.isin(curator.non_validated[\"var_index\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'var_index' is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'assay' is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_type' is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'development_stage' is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'disease' is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'donor_id' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'self_reported_ethnicity' is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'sex' is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'suspension_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'tissue_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'organism' is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_line' is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'genetic_treatments' is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'compound_treatments' is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'environmental_treatments' is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'combination_treatments' is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Need to recreate Curator object because we are using a new object\n",
    "curator = pts.PerturbationCurator(adata)\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "curator.add_new_from(\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'var_index' is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'assay' is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_type' is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'development_stage' is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'disease' is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'donor_id' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'self_reported_ethnicity' is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'sex' is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'suspension_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'tissue_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'organism' is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_line' is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'genetic_treatments' is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'compound_treatments' is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'environmental_treatments' is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'combination_treatments' is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator.validate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All treatment columns validate but that's only because they're all empty."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curate perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Move\n",
    "adata.obs[\"genetic_treatments\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"CRISPR\", None\n",
    ")\n",
    "adata.obs[\"compound_treatments\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"drug\", None\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genetic treatments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, 'sggpx4-2', 'sglacz', 'sggpx4-1', 'sgor2j2']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(adata.obs[\"genetic_treatments\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning existing GeneticTreatment record with same name: 'sggpx4-1'\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Glutathione Peroxidase 4'\n",
      "\u001b[92m→\u001b[0m returning existing GeneticTreatment record with same name: 'sggpx4-2'\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Glutathione Peroxidase 4'\n",
      "\u001b[92m→\u001b[0m returning existing GeneticTreatment record with same name: 'sgor2j2'\n",
      "\u001b[92m✓\u001b[0m loaded \u001b[1;92m1 Gene record\u001b[0m matching \u001b[3msymbol\u001b[0m: \u001b[1;92m'OR2J2'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m loaded \u001b[1;92m1 Gene record\u001b[0m matching \u001b[3msynonyms\u001b[0m: \u001b[1;92m'or2j2'\u001b[0m\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Olfactory receptor family 2 subfamily J member 2'\n",
      "\u001b[92m→\u001b[0m returning existing GeneticTreatment record with same name: 'sglacz'\n",
      "\u001b[92m→\u001b[0m returning existing Gene record with same symbol: 'lacz'\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'beta-galactosidase control'\n"
     ]
    }
   ],
   "source": [
    "treatments = [\n",
    "    (\"sggpx4-1\", \"GPX4\", \"Glutathione Peroxidase 4\"),\n",
    "    (\"sggpx4-2\", \"GPX4\", \"Glutathione Peroxidase 4\"),\n",
    "    (\"sgor2j2\", \"or2j2\", \"Olfactory receptor family 2 subfamily J member 2\"),\n",
    "    (\"sglacz\", \"lacz\", \"beta-galactosidase control\"),  # Control from E. coli\n",
    "]\n",
    "organism = bt.Organism.lookup().human\n",
    "\n",
    "genetic_treatments = []\n",
    "for name, symbol, target_name in treatments:\n",
    "    treatment = wl.GeneticTreatment(system=\"CRISPR KO\", name=name).save()\n",
    "    if symbol != \"lacz\":\n",
    "        gene_result = bt.Gene.from_source(symbol=symbol, organism=organism)\n",
    "        gene = gene_result[0] if isinstance(gene_result, list) else gene_result\n",
    "        gene = gene.save()\n",
    "    else:\n",
    "        gene = bt.Gene(symbol=symbol, organism=organism).save()\n",
    "    target = wl.TreatmentTarget(name=target_name).save()\n",
    "    target.genes.add(gene)\n",
    "    treatment.targets.add(target)\n",
    "    genetic_treatments.append(treatment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'var_index' is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'assay' is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_type' is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'development_stage' is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'disease' is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'donor_id' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'self_reported_ethnicity' is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'sex' is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'suspension_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'tissue_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'organism' is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_line' is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'genetic_treatments' is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mcompound_treatments\u001b[0m on \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m14 terms\u001b[0m are not validated: \u001b[1;91m'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'brd3379', 'JQ1', 'azd5591', 'control', ...\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('compound_treatments')\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'environmental_treatments' is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'combination_treatments' is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator.validate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m8 Compound records from Bionty\u001b[0m matching \u001b[3mname\u001b[0m: \u001b[1;95m'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'JQ1', 'everolimus'\u001b[0m\n",
      "\u001b[93m!\u001b[0m \u001b[1;91mdid not create\u001b[0m Compound records for \u001b[1;93m6 non-validated\u001b[0m \u001b[3mnames\u001b[0m: \u001b[1;93m'azd5591', 'brd3379', 'control', 'idasanutlin', 'prexasertib', 'taselisib'\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "compounds = wl.Compound.from_values(adata.obs[\"compound_treatments\"], field=\"name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The remaining compounds are not in chebi and we create records for them\n",
    "for missing in [\n",
    "    \"azd5591\",\n",
    "    \"brd3379\",\n",
    "    \"control\",\n",
    "    \"idasanutlin\",\n",
    "    \"prexasertib\",\n",
    "    \"taselisib\",\n",
    "]:\n",
    "    compounds.append(wl.Compound(name=missing))\n",
    "ln.save(compounds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_metadata = adata.obs[adata.obs[\"compound_treatments\"].notna()]\n",
    "\n",
    "unique_treatments = drug_metadata[\n",
    "    [\"perturbation\", \"dose_unit\", \"dose_value\"]\n",
    "].drop_duplicates()\n",
    "\n",
    "compound_treatments = []\n",
    "for _, row in unique_treatments.iterrows():\n",
    "    compound = wl.Compound.get(name=row[\"perturbation\"])\n",
    "    treatment = wl.CompoundTreatment(\n",
    "        name=compound.name,\n",
    "        concentration=row[\"dose_value\"],\n",
    "        concentration_unit=row[\"dose_unit\"],\n",
    "    )\n",
    "    compound_treatments.append(treatment)\n",
    "\n",
    "ln.save(compound_treatments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P36507'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P00533'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q9UK79'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q15303'\u001b[0m\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'MAPK/ERK pathway'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P15056'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P23921'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P10415'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q07812'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i6XpYvMC</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:10:57.608736+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid              name description  run_id  \\\n",
       "id                                                   \n",
       "4   i6XpYvMC  MAPK/ERK pathway        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "4  2024-11-11 15:10:57.608736+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P49721'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SnJAC9fR</td>\n",
       "      <td>Apoptosis regulation</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:04.651378+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                  name description  run_id  \\\n",
       "id                                                       \n",
       "7   SnJAC9fR  Apoptosis regulation        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "7  2024-11-11 15:11:04.651378+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O60885'\u001b[0m\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Transcription regulation (BET proteins)'\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Apoptosis regulation'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q07820'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O14757'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i6XpYvMC</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:10:57.608736+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fuAhPhXj</td>\n",
       "      <td>Proteasome pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:07.553672+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                name description  run_id  \\\n",
       "id                                                     \n",
       "4   i6XpYvMC    MAPK/ERK pathway        None       2   \n",
       "8   fuAhPhXj  Proteasome pathway        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "4  2024-11-11 15:10:57.608736+00:00              1  \n",
       "8  2024-11-11 15:11:07.553672+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P42336'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O00329'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P48736'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SnJAC9fR</td>\n",
       "      <td>Apoptosis regulation</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:04.651378+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5KWr4lMW</td>\n",
       "      <td>Transcription regulation (BET proteins)</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:08.744610+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                                     name description  run_id  \\\n",
       "id                                                                          \n",
       "7   SnJAC9fR                     Apoptosis regulation        None       2   \n",
       "9   5KWr4lMW  Transcription regulation (BET proteins)        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "7  2024-11-11 15:11:04.651378+00:00              1  \n",
       "9  2024-11-11 15:11:08.744610+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q00987'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>irV5vyHx</td>\n",
       "      <td>PI3K/AKT/mTOR pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:12.919341+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i6XpYvMC</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:10:57.608736+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fuAhPhXj</td>\n",
       "      <td>Proteasome pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:07.553672+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                   name description  run_id  \\\n",
       "id                                                        \n",
       "12  irV5vyHx  PI3K/AKT/mTOR pathway        None       2   \n",
       "4   i6XpYvMC       MAPK/ERK pathway        None       2   \n",
       "8   fuAhPhXj     Proteasome pathway        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "12 2024-11-11 15:11:12.919341+00:00              1  \n",
       "4  2024-11-11 15:10:57.608736+00:00              1  \n",
       "8  2024-11-11 15:11:07.553672+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P42345'\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "compounds_to_targets = {\n",
    "    \"trametinib\": (\"MAPK/ERK pathway\", [\"P36507\"]),\n",
    "    \"afatinib\": (\"EGFR, HER2, HER4 signaling\", [\"P00533\", \"Q9UK79\", \"Q15303\"]),\n",
    "    \"dabrafenib\": (\"MAPK/ERK pathway\", [\"P15056\"]),\n",
    "    \"gemcitabine\": (\"DNA synthesis inhibition\", [\"P23921\"]),  # No single protein target\n",
    "    \"navitoclax\": (\"Apoptosis regulation\", [\"P10415\", \"Q07812\"]),\n",
    "    \"bortezomib\": (\"Proteasome pathway\", [\"P49721\"]),\n",
    "    \"brd3379\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"JQ1\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"azd5591\": (\"Apoptosis regulation\", [\"Q07820\"]),\n",
    "    \"control\": (\"Baseline\", [None]),  # No target for control\n",
    "    \"prexasertib\": (\"DNA damage response\", [\"O14757\"]),\n",
    "    \"taselisib\": (\"PI3K/AKT/mTOR pathway\", [\"P42336\", \"O00329\", \"P48736\"]),\n",
    "    \"idasanutlin\": (\"p53 regulation\", [\"Q00987\"]),\n",
    "    \"everolimus\": (\"mTOR pathway\", [\"P42345\"])\n",
    "}\n",
    "\n",
    "\n",
    "for compound_treatment_name, targets_tuple in compounds_to_targets.items():\n",
    "    compound_treatment = wl.CompoundTreatment.get(name=compound_treatment_name)\n",
    "    target = wl.TreatmentTarget(name=targets_tuple[0]).save()\n",
    "    proteins = []\n",
    "    for id in targets_tuple[1]:\n",
    "        if id is not None:\n",
    "            proteins.append(bt.Protein.from_source(uniprotkb_id=id).save())\n",
    "    target.proteins.set(proteins)\n",
    "    compound_treatment.targets.add(target)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = fr.Reference(\n",
    "    name=\"Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action\",\n",
    "    abbr=\"McFarland 2020\",\n",
    "    url=\"https://www.nature.com/articles/s41467-020-17440-w\",\n",
    "    doi=\"10.1038/s41467-020-17440-w\",\n",
    "    text=(\n",
    "        \"Assays to study cancer cell responses to pharmacologic or genetic perturbations are typically \"\n",
    "        \"restricted to using simple phenotypic readouts such as proliferation rate. Information-rich assays, \"\n",
    "        \"such as gene-expression profiling, have generally not permitted efficient profiling of a given \"\n",
    "        \"perturbation across multiple cellular contexts. Here, we develop MIX-Seq, a method for multiplexed \"\n",
    "        \"transcriptional profiling of post-perturbation responses across a mixture of samples with single-cell \"\n",
    "        \"resolution, using SNP-based computational demultiplexing of single-cell RNA-sequencing data. We show \"\n",
    "        \"that MIX-Seq can be used to profile responses to chemical or genetic perturbations across pools of 100 \"\n",
    "        \"or more cancer cell lines. We combine it with Cell Hashing to further multiplex additional experimental \"\n",
    "        \"conditions, such as post-treatment time points or drug doses. Analyzing the high-content readout of \"\n",
    "        \"scRNA-seq reveals both shared and context-specific transcriptional response components that can identify \"\n",
    "        \"drug mechanism of action and enable prediction of long-term cell viability from short-term transcriptional \"\n",
    "        \"responses to treatment.\"\n",
    "    ),\n",
    ").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove unused columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.drop(\n",
    "    [\n",
    "        \"depmap_id\",\n",
    "        \"cancer\",\n",
    "        \"cell_quality\",\n",
    "        \"channel\",\n",
    "        \"perturbation\",\n",
    "        \"perturbation_type\",\n",
    "        \"singlet_dev\",\n",
    "        \"singlet_dev_z\",\n",
    "        \"singlet_margin\",\n",
    "        \"singlet_z_margin\",\n",
    "        \"nperts\",\n",
    "        \"ngenes\",\n",
    "        \"ncounts\",\n",
    "        \"cell_det_rate\",\n",
    "        \"doublet_GMM_prob\",\n",
    "        \"doublet_dev_imp\",\n",
    "        \"doublet_z_margin\",\n",
    "        'doublet_CL1',\n",
    "        'doublet_CL2'\n",
    "    ],\n",
    "    axis=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Register curated artifact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'var_index' is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'assay' is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_type' is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'development_stage' is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'disease' is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'donor_id' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'self_reported_ethnicity' is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'sex' is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'suspension_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'tissue_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'organism' is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_line' is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'genetic_treatments' is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'compound_treatments' is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'environmental_treatments' is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'combination_treatments' is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar description exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>version</th>\n",
       "      <th>is_latest</th>\n",
       "      <th>description</th>\n",
       "      <th>key</th>\n",
       "      <th>suffix</th>\n",
       "      <th>type</th>\n",
       "      <th>size</th>\n",
       "      <th>hash</th>\n",
       "      <th>n_objects</th>\n",
       "      <th>n_observations</th>\n",
       "      <th>_hash_type</th>\n",
       "      <th>_accessor</th>\n",
       "      <th>visibility</th>\n",
       "      <th>_key_is_virtual</th>\n",
       "      <th>storage_id</th>\n",
       "      <th>transform_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Xk7Qaik9vBLV4PKf0001</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>McFarland 2020 preprocessed</td>\n",
       "      <td>None</td>\n",
       "      <td>.h5ad</td>\n",
       "      <td>dataset</td>\n",
       "      <td>2511528</td>\n",
       "      <td>Iz4mVUpIruvtABfA6D3vQA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>md5</td>\n",
       "      <td>AnnData</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2024-11-11 14:44:54.959908+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     uid version  is_latest                  description  \\\n",
       "id                                                                         \n",
       "2   Xk7Qaik9vBLV4PKf0001    None       True  McFarland 2020 preprocessed   \n",
       "\n",
       "     key suffix     type     size                    hash n_objects  \\\n",
       "id                                                                    \n",
       "2   None  .h5ad  dataset  2511528  Iz4mVUpIruvtABfA6D3vQA      None   \n",
       "\n",
       "   n_observations _hash_type _accessor  visibility  _key_is_virtual  \\\n",
       "id                                                                    \n",
       "2            None        md5   AnnData           1             True   \n",
       "\n",
       "    storage_id  transform_id  run_id                       created_at  \\\n",
       "id                                                                      \n",
       "2            3             3       3 2024-11-11 14:44:54.959908+00:00   \n",
       "\n",
       "    created_by_id  \n",
       "id                 \n",
       "2               1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m    \u001b[1;93m11 unique terms\u001b[0m (42.30%) are not validated for \u001b[3mname\u001b[0m: \u001b[1;93m'dose_unit', 'dose_value', 'hash_assignment', 'hash_tag', 'num_SNPs', 'singlet_ID', 'time', 'tot_reads', 'percent_mito', 'percent_ribo', ...\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "artifact = curator.save_artifact(description=\"McFarland AnnData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the perturbations and references\n",
    "artifact.genetic_treatments.set(genetic_treatments)\n",
    "artifact.compound_treatments.set(compound_treatments)\n",
    "artifact.references.add(reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;92mArtifact\u001b[0m(uid='A2xWHSPBuPgBhcHi0000', is_latest=True, description='McFarland AnnData', suffix='.h5ad', type='dataset', size=3345456, hash='jzxUs9DOPJewAKOb6ZMaGg', n_observations=1000, _hash_type='md5', _accessor='AnnData', visibility=1, _key_is_virtual=True, created_at=2024-11-11 15:11:20 UTC)\n",
      "  \u001b[3mProvenance\u001b[0m\n",
      "    .storage = '/home/zeth/PycharmProjects/pertpy-datasets/scripts/lamindb_datasets/test-perturbation'\n",
      "    .transform = 'Curate perturbation dataset with `PerturbationCurator`'\n",
      "    .run = 2024-11-11 15:10:01 UTC\n",
      "    .created_by = 'zethson'\n",
      "  \u001b[3mLabels\u001b[0m\n",
      "    .references = 'Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action'\n",
      "    .genetic_treatments = 'sggpx4-1', 'sggpx4-2', 'sgor2j2', 'sglacz'\n",
      "    .compound_treatments = 'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'brd3379', 'JQ1', 'azd5591', 'control', ...\n",
      "    .organisms = 'human'\n",
      "    .cell_types = 'unknown'\n",
      "    .diseases = 'colorectal cancer', 'ovarian cancer', 'sarcoma', 'bile duct cancer', 'bone cancer', 'liver cancer', 'urinary bladder carcinoma', 'prostate cancer', 'gastric cancer', 'uterine corpus cancer', ...\n",
      "    .cell_lines = 'NUGC-3', 'HuCCT1', 'WM-266-4', 'NCI-H838', 'Ishikawa (Heraklio) 02 ER-', 'SNU-761', 'NMC-G1', 'COLO-680N', 'OVK18', 'NCI-H1048', ...\n",
      "    .phenotypes = 'male', 'female', 'unknown'\n",
      "    .experimental_factors = 'single-cell RNA sequencing'\n",
      "    .developmental_stages = 'unknown'\n",
      "    .ethnicities = 'unknown'\n",
      "    .ulabels = 'unknown', 'cell', 'cell_line'\n",
      "  \u001b[3mFeatures\u001b[0m\n",
      "    'assay' = 'single-cell RNA sequencing'\n",
      "    'cell_line' = '22Rv1', '253J-BV', '42-MG-BA', '639-V', '647-V', '769-P', '786-O', '8505C', 'A-375', 'A2780', ...\n",
      "    'cell_type' = 'unknown'\n",
      "    'development_stage' = 'unknown'\n",
      "    'disease' = 'bile duct cancer', 'bone cancer', 'brain cancer', 'breast cancer', 'colorectal cancer', 'esophageal cancer', 'gastric cancer', 'head and neck cancer', 'kidney cancer', 'liver cancer', ...\n",
      "    'donor_id' = 'unknown'\n",
      "    'organism' = 'human'\n",
      "    'self_reported_ethnicity' = 'unknown'\n",
      "    'sex' = 'female', 'male', 'unknown'\n",
      "    'suspension_type' = 'cell'\n",
      "    'tissue_type' = 'cell_line'\n",
      "  \u001b[3mFeature sets\u001b[0m\n",
      "    'var' = 'MAGED2', 'CPE', 'DDX43', 'SPANXA2', 'LNCPRESS1', 'TMEM175', 'None', 'MMD', 'LINC01834', 'DYNC2I1', 'CCRL2', 'MRPS6', 'KCNH1', 'TBILA', 'SLC5A7', 'GSX2', 'ATP5MG', 'TMEM190', 'GTF2H1'\n",
      "    'obs' = 'assay', 'cell_type', 'development_stage', 'disease', 'donor_id', 'self_reported_ethnicity', 'sex', 'suspension_type', 'tissue_type', 'organism', 'cell_line', 'genetic_treatments', 'compound_treatments', 'environmental_treatments', 'combination_treatments'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "artifact.describe()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lamindb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
