{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Curate perturbation dataset with `PerturbationCurator`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we use pertpy's `PerturbationCurator` to ensure that a perturbatin dataset conforms to both, `CELLxGENE` (schema 5.1.0) and pertpy's defined criteria.\n",
    "More specifically, the `PerturbationCurator` builds upon [cellxgene-lamin](https://github.com/laminlabs/cellxgene-lamin) and extends it by further requiring `cell_line` and `X_treatments` columns for the perturbations.\n",
    "\n",
    "This guide demonstrates how to curate a complex, real world perturbation dataset [McFarland et al. 2020](https://www.nature.com/articles/s41467-020-17440-w) using `PerturbationCurator`. Please have a look at [lamindb's perturbation guide](https://docs.lamin.ai/perturbation) for more details."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install pertpy-datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: zethson/test-perturbation\n"
     ]
    }
   ],
   "source": [
    "# Using a local instance here but in practice, we use `laminlabs/pertpy-datasets`\n",
    "!lamin init --storage ./test-perturbation --schema bionty,wetlab,ourprojects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: zethson/test-perturbation\n"
     ]
    }
   ],
   "source": [
    "import lamindb as ln\n",
    "import bionty as bt\n",
    "import wetlab as wl\n",
    "import ourprojects as opr\n",
    "import pertpy_datasets as pts\n",
    "import scanpy as sc\n",
    "\n",
    "from lamindb.core import RecordList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m created Transform('HIRTYxL3'), started new Run('ifbU3n87') at 2024-12-10 12:49:06 UTC\n",
      "\u001b[92m→\u001b[0m notebook imports: bionty==0.53.2 lamindb==0.77.2 ourprojects==0.1.0 pertpy_datasets==0.1.0 scanpy==1.10.4 wetlab==0.37.0\n"
     ]
    }
   ],
   "source": [
    "ln.track(\"HIRTYxL3aZc70000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m completing transfer to track Artifact('Xk7Qaik9') as input\n",
      "\u001b[92m→\u001b[0m mapped records: \n",
      "\u001b[92m→\u001b[0m transferred records: Artifact(uid='Xk7Qaik9vBLV4PKf0001'), Storage(uid='D9BilDV2')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cell_det_rate</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_quality</th>\n",
       "      <th>channel</th>\n",
       "      <th>disease</th>\n",
       "      <th>dose_unit</th>\n",
       "      <th>dose_value</th>\n",
       "      <th>doublet_CL1</th>\n",
       "      <th>doublet_CL2</th>\n",
       "      <th>doublet_GMM_prob</th>\n",
       "      <th>doublet_dev_imp</th>\n",
       "      <th>doublet_z_margin</th>\n",
       "      <th>hash_assignment</th>\n",
       "      <th>hash_tag</th>\n",
       "      <th>num_SNPs</th>\n",
       "      <th>organism</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>singlet_ID</th>\n",
       "      <th>singlet_dev</th>\n",
       "      <th>singlet_dev_z</th>\n",
       "      <th>singlet_margin</th>\n",
       "      <th>singlet_z_margin</th>\n",
       "      <th>time</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>tot_reads</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>chembl-ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACTGGTGTCTCTCTG</th>\n",
       "      <td>ACH-000390</td>\n",
       "      <td>True</td>\n",
       "      <td>0.093159</td>\n",
       "      <td>LUDLU-1</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>TE14_OESOPHAGUS</td>\n",
       "      <td>2.269468e-10</td>\n",
       "      <td>0.009426</td>\n",
       "      <td>0.403316</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>481</td>\n",
       "      <td>human</td>\n",
       "      <td>trametinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>0.655877</td>\n",
       "      <td>14.860933</td>\n",
       "      <td>0.462273</td>\n",
       "      <td>12.351139</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>787</td>\n",
       "      <td>1</td>\n",
       "      <td>3045</td>\n",
       "      <td>12895.0</td>\n",
       "      <td>3.202792</td>\n",
       "      <td>24.955409</td>\n",
       "      <td>CHEMBL2103875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAGGCTCAGATTTCG</th>\n",
       "      <td>ACH-000444</td>\n",
       "      <td>True</td>\n",
       "      <td>0.145728</td>\n",
       "      <td>LU99</td>\n",
       "      <td>normal</td>\n",
       "      <td>2</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.5</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>MCAS_OVARY</td>\n",
       "      <td>8.562908e-04</td>\n",
       "      <td>0.010173</td>\n",
       "      <td>0.188284</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>1003</td>\n",
       "      <td>human</td>\n",
       "      <td>afatinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>0.762847</td>\n",
       "      <td>10.648094</td>\n",
       "      <td>0.474590</td>\n",
       "      <td>8.164565</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1597</td>\n",
       "      <td>1</td>\n",
       "      <td>4763</td>\n",
       "      <td>23161.0</td>\n",
       "      <td>7.473771</td>\n",
       "      <td>18.051898</td>\n",
       "      <td>CHEMBL1173655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCCAAATCAAGCCGTC</th>\n",
       "      <td>ACH-000396</td>\n",
       "      <td>True</td>\n",
       "      <td>0.117330</td>\n",
       "      <td>J82</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>urinary bladder carcinoma</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>IGR1_SKIN</td>\n",
       "      <td>6.490367e-08</td>\n",
       "      <td>0.009686</td>\n",
       "      <td>1.185862</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>647</td>\n",
       "      <td>human</td>\n",
       "      <td>dabrafenib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>0.651059</td>\n",
       "      <td>14.740111</td>\n",
       "      <td>0.404508</td>\n",
       "      <td>11.188513</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1159</td>\n",
       "      <td>1</td>\n",
       "      <td>3834</td>\n",
       "      <td>18062.0</td>\n",
       "      <td>2.762706</td>\n",
       "      <td>22.085040</td>\n",
       "      <td>CHEMBL2028663</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   depmap_id  cancer  cell_det_rate cell_line cell_quality  \\\n",
       "AACTGGTGTCTCTCTG  ACH-000390    True       0.093159   LUDLU-1       normal   \n",
       "ATAGGCTCAGATTTCG  ACH-000444    True       0.145728      LU99       normal   \n",
       "GCCAAATCAAGCCGTC  ACH-000396    True       0.117330       J82       normal   \n",
       "\n",
       "                 channel                    disease dose_unit  dose_value  \\\n",
       "AACTGGTGTCTCTCTG     nan                lung cancer        µM         0.1   \n",
       "ATAGGCTCAGATTTCG       2                lung cancer        µM         0.5   \n",
       "GCCAAATCAAGCCGTC     nan  urinary bladder carcinoma        µM         0.1   \n",
       "\n",
       "                        doublet_CL1      doublet_CL2  doublet_GMM_prob  \\\n",
       "AACTGGTGTCTCTCTG        LUDLU1_LUNG  TE14_OESOPHAGUS      2.269468e-10   \n",
       "ATAGGCTCAGATTTCG          LU99_LUNG       MCAS_OVARY      8.562908e-04   \n",
       "GCCAAATCAAGCCGTC  J82_URINARY_TRACT        IGR1_SKIN      6.490367e-08   \n",
       "\n",
       "                  doublet_dev_imp  doublet_z_margin hash_assignment hash_tag  \\\n",
       "AACTGGTGTCTCTCTG         0.009426          0.403316             nan      nan   \n",
       "ATAGGCTCAGATTTCG         0.010173          0.188284             nan      nan   \n",
       "GCCAAATCAAGCCGTC         0.009686          1.185862             nan      nan   \n",
       "\n",
       "                  num_SNPs organism perturbation perturbation_type   sex  \\\n",
       "AACTGGTGTCTCTCTG       481    human   trametinib              drug  Male   \n",
       "ATAGGCTCAGATTTCG      1003    human     afatinib              drug  Male   \n",
       "GCCAAATCAAGCCGTC       647    human   dabrafenib              drug  Male   \n",
       "\n",
       "                         singlet_ID  singlet_dev  singlet_dev_z  \\\n",
       "AACTGGTGTCTCTCTG        LUDLU1_LUNG     0.655877      14.860933   \n",
       "ATAGGCTCAGATTTCG          LU99_LUNG     0.762847      10.648094   \n",
       "GCCAAATCAAGCCGTC  J82_URINARY_TRACT     0.651059      14.740111   \n",
       "\n",
       "                  singlet_margin  singlet_z_margin time tissue_type  \\\n",
       "AACTGGTGTCTCTCTG        0.462273         12.351139   24   cell_line   \n",
       "ATAGGCTCAGATTTCG        0.474590          8.164565   24   cell_line   \n",
       "GCCAAATCAAGCCGTC        0.404508         11.188513   24   cell_line   \n",
       "\n",
       "                  tot_reads  nperts  ngenes  ncounts  percent_mito  \\\n",
       "AACTGGTGTCTCTCTG        787       1    3045  12895.0      3.202792   \n",
       "ATAGGCTCAGATTTCG       1597       1    4763  23161.0      7.473771   \n",
       "GCCAAATCAAGCCGTC       1159       1    3834  18062.0      2.762706   \n",
       "\n",
       "                  percent_ribo      chembl-ID  \n",
       "AACTGGTGTCTCTCTG     24.955409  CHEMBL2103875  \n",
       "ATAGGCTCAGATTTCG     18.051898  CHEMBL1173655  \n",
       "GCCAAATCAAGCCGTC     22.085040  CHEMBL2028663  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = ln.Artifact.using(\"laminlabs/lamindata\").get(uid=\"Xk7Qaik9vBLV4PKf0001\").load()\n",
    "adata.obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate an embedding because CELLxGENE requires one\n",
    "sc.tl.pca(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curator non-perturbation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m added defaults to the AnnData object: {'assay': 'unknown', 'cell_type': 'unknown', 'development_stage': 'unknown', 'donor_id': 'unknown', 'self_reported_ethnicity': 'unknown', 'suspension_type': 'cell', 'genetic_treatments': '', 'compound_treatments': '', 'environmental_treatments': '', 'combination_treatments': ''}\n",
      "\u001b[92m✓\u001b[0m added 15 records with \u001b[3mFeature.name\u001b[0m for \"columns\": 'assay', 'cell_type', 'development_stage', 'disease', 'donor_id', 'self_reported_ethnicity', 'sex', 'suspension_type', 'tissue_type', 'organism', 'cell_line', 'genetic_treatments', 'compound_treatments', 'environmental_treatments', 'combination_treatments'\n",
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n",
      "\u001b[92m✓\u001b[0m added 1869 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mGene.ensembl_gene_id\u001b[0m for \"var_index\": 'ENSG00000102316', 'ENSG00000109472', 'ENSG00000080007', 'ENSG00000203926', 'ENSG00000232301', 'ENSG00000127419', 'ENSG00000266869', 'ENSG00000108960', 'ENSG00000261316', 'ENSG00000126870', 'ENSG00000121797', 'ENSG00000243927', 'ENSG00000143473', 'ENSG00000261488', 'ENSG00000115665', 'ENSG00000180613', 'ENSG00000167283', 'ENSG00000160472', 'ENSG00000273387', 'ENSG00000110768', ...\n",
      "\u001b[94m•\u001b[0m saving validated records of 'disease'\n",
      "\u001b[92m✓\u001b[0m added 21 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mDisease.name\u001b[0m for \"disease\": 'sarcoma', 'malignant pancreatic neoplasm', 'prostate cancer', 'uterine corpus cancer', 'bone cancer', 'lung cancer', 'brain cancer', 'esophageal cancer', 'skin cancer', 'neuroblastoma', 'head and neck cancer', 'urinary bladder carcinoma', 'gastric cancer', 'colorectal cancer', 'thyroid cancer', 'bile duct cancer', 'rhabdoid tumor', 'liver cancer', 'ovarian cancer', 'kidney cancer', ...\n",
      "\u001b[94m•\u001b[0m saving validated records of 'sex'\n",
      "\u001b[92m✓\u001b[0m added 2 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mPhenotype.name\u001b[0m for \"sex\": 'female', 'male'\n",
      "\u001b[94m•\u001b[0m saving validated records of 'cell_line'\n",
      "\u001b[92m✓\u001b[0m added 183 records \u001b[1;92mfrom public\u001b[0m with \u001b[3mCellLine.name\u001b[0m for \"cell_line\": '42-MG-BA', 'NCI-H841', 'CCF-STTG1', 'KNS-42', 'Hs 852.T', 'G-401', 'LK-2', 'KNS-60', 'A2780', 'HLF', 'OVK18', 'LU99', 'NCI-H2009', 'Caov-3', 'RMUG-S', 'BT-549', 'J82', 'NCI-H1650', 'WM-266-4', 'GCIY', ...\n",
      "\u001b[94m•\u001b[0m mapping \"var_index\" on \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m2 terms\u001b[0m are not validated: \u001b[1;91m'ENSG00000255823', 'ENSG00000272370'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from_var_index()\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"assay\" on \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"assay\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"cell_type\" on \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"cell_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"development_stage\" on \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"development_stage\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"disease\" on \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'pancreatic cancer'\u001b[0m\n",
      "    \u001b[1;93m1 synonym\u001b[0m found: \u001b[1;93m\"pancreatic cancer\" → \"malignant pancreatic neoplasm\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"disease\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"donor_id\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"donor_id\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"self_reported_ethnicity\" on \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"self_reported_ethnicity\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"sex\" on \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m3 terms\u001b[0m are not validated: \u001b[1;91m'Male', 'Female', 'Unknown'\u001b[0m\n",
      "    \u001b[1;93m2 synonyms\u001b[0m found: \u001b[1;93m\"Male\" → \"male\", \"Female\" → \"female\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"sex\")\u001b[0m    for remaining terms:\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"sex\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"suspension_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"suspension_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"tissue_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell_line'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"tissue_type\")\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"organism\" is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_line\" is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"genetic_treatments\" is validated against \u001b[3mGeneticPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound_treatments\" is validated against \u001b[3mCompoundPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"environmental_treatments\" is validated against \u001b[3mEnvironmentalPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"combination_treatments\" is validated against \u001b[3mCombinationPerturbation.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator = pts.PerturbationCurator(\n",
    "    adata, using_key=\"test-perturbation\"\n",
    ")\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cell_det_rate</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_quality</th>\n",
       "      <th>channel</th>\n",
       "      <th>disease</th>\n",
       "      <th>dose_unit</th>\n",
       "      <th>dose_value</th>\n",
       "      <th>doublet_CL1</th>\n",
       "      <th>doublet_CL2</th>\n",
       "      <th>doublet_GMM_prob</th>\n",
       "      <th>doublet_dev_imp</th>\n",
       "      <th>doublet_z_margin</th>\n",
       "      <th>hash_assignment</th>\n",
       "      <th>hash_tag</th>\n",
       "      <th>num_SNPs</th>\n",
       "      <th>organism</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>perturbation_type</th>\n",
       "      <th>sex</th>\n",
       "      <th>singlet_ID</th>\n",
       "      <th>singlet_dev</th>\n",
       "      <th>singlet_dev_z</th>\n",
       "      <th>singlet_margin</th>\n",
       "      <th>singlet_z_margin</th>\n",
       "      <th>time</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>tot_reads</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>chembl-ID</th>\n",
       "      <th>assay</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>genetic_treatments</th>\n",
       "      <th>compound_treatments</th>\n",
       "      <th>environmental_treatments</th>\n",
       "      <th>combination_treatments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACTGGTGTCTCTCTG</th>\n",
       "      <td>ACH-000390</td>\n",
       "      <td>True</td>\n",
       "      <td>0.093159</td>\n",
       "      <td>LUDLU-1</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>TE14_OESOPHAGUS</td>\n",
       "      <td>2.269468e-10</td>\n",
       "      <td>0.009426</td>\n",
       "      <td>0.403316</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>481</td>\n",
       "      <td>human</td>\n",
       "      <td>trametinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>0.655877</td>\n",
       "      <td>14.860933</td>\n",
       "      <td>0.462273</td>\n",
       "      <td>12.351139</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>787</td>\n",
       "      <td>1</td>\n",
       "      <td>3045</td>\n",
       "      <td>12895.0</td>\n",
       "      <td>3.202792</td>\n",
       "      <td>24.955409</td>\n",
       "      <td>CHEMBL2103875</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAGGCTCAGATTTCG</th>\n",
       "      <td>ACH-000444</td>\n",
       "      <td>True</td>\n",
       "      <td>0.145728</td>\n",
       "      <td>LU99</td>\n",
       "      <td>normal</td>\n",
       "      <td>2</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.5</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>MCAS_OVARY</td>\n",
       "      <td>8.562908e-04</td>\n",
       "      <td>0.010173</td>\n",
       "      <td>0.188284</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>1003</td>\n",
       "      <td>human</td>\n",
       "      <td>afatinib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>0.762847</td>\n",
       "      <td>10.648094</td>\n",
       "      <td>0.474590</td>\n",
       "      <td>8.164565</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1597</td>\n",
       "      <td>1</td>\n",
       "      <td>4763</td>\n",
       "      <td>23161.0</td>\n",
       "      <td>7.473771</td>\n",
       "      <td>18.051898</td>\n",
       "      <td>CHEMBL1173655</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCCAAATCAAGCCGTC</th>\n",
       "      <td>ACH-000396</td>\n",
       "      <td>True</td>\n",
       "      <td>0.117330</td>\n",
       "      <td>J82</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>urinary bladder carcinoma</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>IGR1_SKIN</td>\n",
       "      <td>6.490367e-08</td>\n",
       "      <td>0.009686</td>\n",
       "      <td>1.185862</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>647</td>\n",
       "      <td>human</td>\n",
       "      <td>dabrafenib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>0.651059</td>\n",
       "      <td>14.740111</td>\n",
       "      <td>0.404508</td>\n",
       "      <td>11.188513</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1159</td>\n",
       "      <td>1</td>\n",
       "      <td>3834</td>\n",
       "      <td>18062.0</td>\n",
       "      <td>2.762706</td>\n",
       "      <td>22.085040</td>\n",
       "      <td>CHEMBL2028663</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGGAGAAGTCGCGTCA</th>\n",
       "      <td>ACH-000997</td>\n",
       "      <td>True</td>\n",
       "      <td>0.005422</td>\n",
       "      <td>HCT-15</td>\n",
       "      <td>low_quality</td>\n",
       "      <td>7</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>HCT15_LARGE_INTESTINE</td>\n",
       "      <td>NCIH322_LUNG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029753</td>\n",
       "      <td>0.000794</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>30</td>\n",
       "      <td>human</td>\n",
       "      <td>gemcitabine</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>HCT15_LARGE_INTESTINE</td>\n",
       "      <td>0.970247</td>\n",
       "      <td>2.852338</td>\n",
       "      <td>0.168971</td>\n",
       "      <td>0.833455</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>76</td>\n",
       "      <td>1</td>\n",
       "      <td>178</td>\n",
       "      <td>726.0</td>\n",
       "      <td>70.247934</td>\n",
       "      <td>5.785124</td>\n",
       "      <td>CHEMBL888</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTTGGAGATCGATA</th>\n",
       "      <td>ACH-000723</td>\n",
       "      <td>True</td>\n",
       "      <td>0.132708</td>\n",
       "      <td>YD-10B</td>\n",
       "      <td>low_quality</td>\n",
       "      <td>nan</td>\n",
       "      <td>head and neck cancer</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YD10B_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>647V_URINARY_TRACT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.156492</td>\n",
       "      <td>1.556214</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>874</td>\n",
       "      <td>human</td>\n",
       "      <td>sggpx4-2</td>\n",
       "      <td>CRISPR</td>\n",
       "      <td>Male</td>\n",
       "      <td>YD10B_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>0.292802</td>\n",
       "      <td>3.272682</td>\n",
       "      <td>0.016459</td>\n",
       "      <td>0.330120</td>\n",
       "      <td>72, 96</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>2105</td>\n",
       "      <td>1</td>\n",
       "      <td>4341</td>\n",
       "      <td>20693.0</td>\n",
       "      <td>0.695887</td>\n",
       "      <td>16.242208</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATCTAGTCATCGGAT</th>\n",
       "      <td>ACH-000015</td>\n",
       "      <td>True</td>\n",
       "      <td>0.094752</td>\n",
       "      <td>NCI-H1581</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>NCIH1581_LUNG</td>\n",
       "      <td>HS766T_PANCREAS</td>\n",
       "      <td>2.865590e-08</td>\n",
       "      <td>0.007438</td>\n",
       "      <td>0.078045</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>480</td>\n",
       "      <td>human</td>\n",
       "      <td>dabrafenib</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>NCIH1581_LUNG</td>\n",
       "      <td>0.573389</td>\n",
       "      <td>19.212898</td>\n",
       "      <td>0.449900</td>\n",
       "      <td>17.285484</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>697</td>\n",
       "      <td>1</td>\n",
       "      <td>3098</td>\n",
       "      <td>10575.0</td>\n",
       "      <td>6.780142</td>\n",
       "      <td>27.158392</td>\n",
       "      <td>CHEMBL2028663</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACTCTTAGTCCCACG</th>\n",
       "      <td>ACH-000252</td>\n",
       "      <td>True</td>\n",
       "      <td>0.069540</td>\n",
       "      <td>LS1034</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LS1034_LARGE_INTESTINE</td>\n",
       "      <td>SNU1079_BILIARY_TRACT</td>\n",
       "      <td>1.436511e-02</td>\n",
       "      <td>0.036492</td>\n",
       "      <td>0.727242</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>366</td>\n",
       "      <td>human</td>\n",
       "      <td>control</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>LS1034_LARGE_INTESTINE</td>\n",
       "      <td>0.640246</td>\n",
       "      <td>10.526643</td>\n",
       "      <td>0.413657</td>\n",
       "      <td>8.703848</td>\n",
       "      <td>6</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>616</td>\n",
       "      <td>0</td>\n",
       "      <td>2274</td>\n",
       "      <td>8281.0</td>\n",
       "      <td>7.837218</td>\n",
       "      <td>21.676126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACTGTCCAGTCACGC</th>\n",
       "      <td>ACH-000681</td>\n",
       "      <td>True</td>\n",
       "      <td>0.094262</td>\n",
       "      <td>A549</td>\n",
       "      <td>normal</td>\n",
       "      <td>5</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>2.5</td>\n",
       "      <td>A549_LUNG</td>\n",
       "      <td>MDAMB435S_SKIN</td>\n",
       "      <td>2.303331e-05</td>\n",
       "      <td>0.018360</td>\n",
       "      <td>0.089038</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>528</td>\n",
       "      <td>human</td>\n",
       "      <td>JQ1</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>A549_LUNG</td>\n",
       "      <td>0.714991</td>\n",
       "      <td>14.528289</td>\n",
       "      <td>0.487355</td>\n",
       "      <td>11.729031</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>786</td>\n",
       "      <td>1</td>\n",
       "      <td>3081</td>\n",
       "      <td>10204.0</td>\n",
       "      <td>8.584869</td>\n",
       "      <td>17.679341</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACCTTGGTCGACTAT</th>\n",
       "      <td>ACH-000875</td>\n",
       "      <td>True</td>\n",
       "      <td>0.163557</td>\n",
       "      <td>NCI-H2347</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NCIH2347_LUNG</td>\n",
       "      <td>BICR6_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>3.441387e-08</td>\n",
       "      <td>0.001044</td>\n",
       "      <td>0.314087</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>1230</td>\n",
       "      <td>human</td>\n",
       "      <td>control</td>\n",
       "      <td>drug</td>\n",
       "      <td>Female</td>\n",
       "      <td>NCIH2347_LUNG</td>\n",
       "      <td>0.721699</td>\n",
       "      <td>14.720316</td>\n",
       "      <td>0.525416</td>\n",
       "      <td>13.127789</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>2591</td>\n",
       "      <td>0</td>\n",
       "      <td>5347</td>\n",
       "      <td>37053.0</td>\n",
       "      <td>3.902518</td>\n",
       "      <td>29.606240</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGTAAGG</th>\n",
       "      <td>ACH-000605</td>\n",
       "      <td>True</td>\n",
       "      <td>0.127501</td>\n",
       "      <td>TE-6</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>esophageal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TE6_OESOPHAGUS</td>\n",
       "      <td>SNU685_ENDOMETRIUM</td>\n",
       "      <td>1.904494e-07</td>\n",
       "      <td>0.010530</td>\n",
       "      <td>0.072845</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>782</td>\n",
       "      <td>human</td>\n",
       "      <td>brd3379</td>\n",
       "      <td>drug</td>\n",
       "      <td>Male</td>\n",
       "      <td>TE6_OESOPHAGUS</td>\n",
       "      <td>0.760473</td>\n",
       "      <td>9.688672</td>\n",
       "      <td>0.461341</td>\n",
       "      <td>7.369671</td>\n",
       "      <td>6</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1422</td>\n",
       "      <td>1</td>\n",
       "      <td>4167</td>\n",
       "      <td>23002.0</td>\n",
       "      <td>6.995044</td>\n",
       "      <td>28.027998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   depmap_id  cancer  cell_det_rate  cell_line cell_quality  \\\n",
       "AACTGGTGTCTCTCTG  ACH-000390    True       0.093159    LUDLU-1       normal   \n",
       "ATAGGCTCAGATTTCG  ACH-000444    True       0.145728       LU99       normal   \n",
       "GCCAAATCAAGCCGTC  ACH-000396    True       0.117330        J82       normal   \n",
       "CGGAGAAGTCGCGTCA  ACH-000997    True       0.005422     HCT-15  low_quality   \n",
       "TAGTTGGAGATCGATA  ACH-000723    True       0.132708     YD-10B  low_quality   \n",
       "...                      ...     ...            ...        ...          ...   \n",
       "GATCTAGTCATCGGAT  ACH-000015    True       0.094752  NCI-H1581       normal   \n",
       "AACTCTTAGTCCCACG  ACH-000252    True       0.069540     LS1034       normal   \n",
       "CACTGTCCAGTCACGC  ACH-000681    True       0.094262       A549       normal   \n",
       "CACCTTGGTCGACTAT  ACH-000875    True       0.163557  NCI-H2347       normal   \n",
       "TTTGTCACACGTAAGG  ACH-000605    True       0.127501       TE-6       normal   \n",
       "\n",
       "                 channel                    disease dose_unit  dose_value  \\\n",
       "AACTGGTGTCTCTCTG     nan                lung cancer        µM         0.1   \n",
       "ATAGGCTCAGATTTCG       2                lung cancer        µM         0.5   \n",
       "GCCAAATCAAGCCGTC     nan  urinary bladder carcinoma        µM         0.1   \n",
       "CGGAGAAGTCGCGTCA       7          colorectal cancer        µM         0.1   \n",
       "TAGTTGGAGATCGATA     nan       head and neck cancer       nan         NaN   \n",
       "...                  ...                        ...       ...         ...   \n",
       "GATCTAGTCATCGGAT     nan                lung cancer        µM         0.1   \n",
       "AACTCTTAGTCCCACG     nan          colorectal cancer        µM         0.0   \n",
       "CACTGTCCAGTCACGC       5                lung cancer        µM         2.5   \n",
       "CACCTTGGTCGACTAT     nan                lung cancer        µM         0.0   \n",
       "TTTGTCACACGTAAGG     nan          esophageal cancer        µM         0.0   \n",
       "\n",
       "                                      doublet_CL1  \\\n",
       "AACTGGTGTCTCTCTG                      LUDLU1_LUNG   \n",
       "ATAGGCTCAGATTTCG                        LU99_LUNG   \n",
       "GCCAAATCAAGCCGTC                J82_URINARY_TRACT   \n",
       "CGGAGAAGTCGCGTCA            HCT15_LARGE_INTESTINE   \n",
       "TAGTTGGAGATCGATA  YD10B_UPPER_AERODIGESTIVE_TRACT   \n",
       "...                                           ...   \n",
       "GATCTAGTCATCGGAT                    NCIH1581_LUNG   \n",
       "AACTCTTAGTCCCACG           LS1034_LARGE_INTESTINE   \n",
       "CACTGTCCAGTCACGC                        A549_LUNG   \n",
       "CACCTTGGTCGACTAT                    NCIH2347_LUNG   \n",
       "TTTGTCACACGTAAGG                   TE6_OESOPHAGUS   \n",
       "\n",
       "                                      doublet_CL2  doublet_GMM_prob  \\\n",
       "AACTGGTGTCTCTCTG                  TE14_OESOPHAGUS      2.269468e-10   \n",
       "ATAGGCTCAGATTTCG                       MCAS_OVARY      8.562908e-04   \n",
       "GCCAAATCAAGCCGTC                        IGR1_SKIN      6.490367e-08   \n",
       "CGGAGAAGTCGCGTCA                     NCIH322_LUNG               NaN   \n",
       "TAGTTGGAGATCGATA               647V_URINARY_TRACT               NaN   \n",
       "...                                           ...               ...   \n",
       "GATCTAGTCATCGGAT                  HS766T_PANCREAS      2.865590e-08   \n",
       "AACTCTTAGTCCCACG            SNU1079_BILIARY_TRACT      1.436511e-02   \n",
       "CACTGTCCAGTCACGC                   MDAMB435S_SKIN      2.303331e-05   \n",
       "CACCTTGGTCGACTAT  BICR6_UPPER_AERODIGESTIVE_TRACT      3.441387e-08   \n",
       "TTTGTCACACGTAAGG               SNU685_ENDOMETRIUM      1.904494e-07   \n",
       "\n",
       "                  doublet_dev_imp  doublet_z_margin hash_assignment hash_tag  \\\n",
       "AACTGGTGTCTCTCTG         0.009426          0.403316             nan      nan   \n",
       "ATAGGCTCAGATTTCG         0.010173          0.188284             nan      nan   \n",
       "GCCAAATCAAGCCGTC         0.009686          1.185862             nan      nan   \n",
       "CGGAGAAGTCGCGTCA         0.029753          0.000794             nan      nan   \n",
       "TAGTTGGAGATCGATA         0.156492          1.556214             nan      nan   \n",
       "...                           ...               ...             ...      ...   \n",
       "GATCTAGTCATCGGAT         0.007438          0.078045             nan      nan   \n",
       "AACTCTTAGTCCCACG         0.036492          0.727242             nan      nan   \n",
       "CACTGTCCAGTCACGC         0.018360          0.089038             nan      nan   \n",
       "CACCTTGGTCGACTAT         0.001044          0.314087             nan      nan   \n",
       "TTTGTCACACGTAAGG         0.010530          0.072845             nan      nan   \n",
       "\n",
       "                  num_SNPs organism perturbation perturbation_type     sex  \\\n",
       "AACTGGTGTCTCTCTG       481    human   trametinib              drug    Male   \n",
       "ATAGGCTCAGATTTCG      1003    human     afatinib              drug    Male   \n",
       "GCCAAATCAAGCCGTC       647    human   dabrafenib              drug    Male   \n",
       "CGGAGAAGTCGCGTCA        30    human  gemcitabine              drug    Male   \n",
       "TAGTTGGAGATCGATA       874    human     sggpx4-2            CRISPR    Male   \n",
       "...                    ...      ...          ...               ...     ...   \n",
       "GATCTAGTCATCGGAT       480    human   dabrafenib              drug    Male   \n",
       "AACTCTTAGTCCCACG       366    human      control              drug    Male   \n",
       "CACTGTCCAGTCACGC       528    human          JQ1              drug    Male   \n",
       "CACCTTGGTCGACTAT      1230    human      control              drug  Female   \n",
       "TTTGTCACACGTAAGG       782    human      brd3379              drug    Male   \n",
       "\n",
       "                                       singlet_ID  singlet_dev  singlet_dev_z  \\\n",
       "AACTGGTGTCTCTCTG                      LUDLU1_LUNG     0.655877      14.860933   \n",
       "ATAGGCTCAGATTTCG                        LU99_LUNG     0.762847      10.648094   \n",
       "GCCAAATCAAGCCGTC                J82_URINARY_TRACT     0.651059      14.740111   \n",
       "CGGAGAAGTCGCGTCA            HCT15_LARGE_INTESTINE     0.970247       2.852338   \n",
       "TAGTTGGAGATCGATA  YD10B_UPPER_AERODIGESTIVE_TRACT     0.292802       3.272682   \n",
       "...                                           ...          ...            ...   \n",
       "GATCTAGTCATCGGAT                    NCIH1581_LUNG     0.573389      19.212898   \n",
       "AACTCTTAGTCCCACG           LS1034_LARGE_INTESTINE     0.640246      10.526643   \n",
       "CACTGTCCAGTCACGC                        A549_LUNG     0.714991      14.528289   \n",
       "CACCTTGGTCGACTAT                    NCIH2347_LUNG     0.721699      14.720316   \n",
       "TTTGTCACACGTAAGG                   TE6_OESOPHAGUS     0.760473       9.688672   \n",
       "\n",
       "                  singlet_margin  singlet_z_margin    time tissue_type  \\\n",
       "AACTGGTGTCTCTCTG        0.462273         12.351139      24   cell_line   \n",
       "ATAGGCTCAGATTTCG        0.474590          8.164565      24   cell_line   \n",
       "GCCAAATCAAGCCGTC        0.404508         11.188513      24   cell_line   \n",
       "CGGAGAAGTCGCGTCA        0.168971          0.833455      24   cell_line   \n",
       "TAGTTGGAGATCGATA        0.016459          0.330120  72, 96   cell_line   \n",
       "...                          ...               ...     ...         ...   \n",
       "GATCTAGTCATCGGAT        0.449900         17.285484      24   cell_line   \n",
       "AACTCTTAGTCCCACG        0.413657          8.703848       6   cell_line   \n",
       "CACTGTCCAGTCACGC        0.487355         11.729031      24   cell_line   \n",
       "CACCTTGGTCGACTAT        0.525416         13.127789      24   cell_line   \n",
       "TTTGTCACACGTAAGG        0.461341          7.369671       6   cell_line   \n",
       "\n",
       "                  tot_reads  nperts  ngenes  ncounts  percent_mito  \\\n",
       "AACTGGTGTCTCTCTG        787       1    3045  12895.0      3.202792   \n",
       "ATAGGCTCAGATTTCG       1597       1    4763  23161.0      7.473771   \n",
       "GCCAAATCAAGCCGTC       1159       1    3834  18062.0      2.762706   \n",
       "CGGAGAAGTCGCGTCA         76       1     178    726.0     70.247934   \n",
       "TAGTTGGAGATCGATA       2105       1    4341  20693.0      0.695887   \n",
       "...                     ...     ...     ...      ...           ...   \n",
       "GATCTAGTCATCGGAT        697       1    3098  10575.0      6.780142   \n",
       "AACTCTTAGTCCCACG        616       0    2274   8281.0      7.837218   \n",
       "CACTGTCCAGTCACGC        786       1    3081  10204.0      8.584869   \n",
       "CACCTTGGTCGACTAT       2591       0    5347  37053.0      3.902518   \n",
       "TTTGTCACACGTAAGG       1422       1    4167  23002.0      6.995044   \n",
       "\n",
       "                  percent_ribo      chembl-ID    assay cell_type  \\\n",
       "AACTGGTGTCTCTCTG     24.955409  CHEMBL2103875  unknown   unknown   \n",
       "ATAGGCTCAGATTTCG     18.051898  CHEMBL1173655  unknown   unknown   \n",
       "GCCAAATCAAGCCGTC     22.085040  CHEMBL2028663  unknown   unknown   \n",
       "CGGAGAAGTCGCGTCA      5.785124      CHEMBL888  unknown   unknown   \n",
       "TAGTTGGAGATCGATA     16.242208            NaN  unknown   unknown   \n",
       "...                        ...            ...      ...       ...   \n",
       "GATCTAGTCATCGGAT     27.158392  CHEMBL2028663  unknown   unknown   \n",
       "AACTCTTAGTCCCACG     21.676126            NaN  unknown   unknown   \n",
       "CACTGTCCAGTCACGC     17.679341            NaN  unknown   unknown   \n",
       "CACCTTGGTCGACTAT     29.606240            NaN  unknown   unknown   \n",
       "TTTGTCACACGTAAGG     28.027998            NaN  unknown   unknown   \n",
       "\n",
       "                 development_stage donor_id self_reported_ethnicity  \\\n",
       "AACTGGTGTCTCTCTG           unknown  unknown                 unknown   \n",
       "ATAGGCTCAGATTTCG           unknown  unknown                 unknown   \n",
       "GCCAAATCAAGCCGTC           unknown  unknown                 unknown   \n",
       "CGGAGAAGTCGCGTCA           unknown  unknown                 unknown   \n",
       "TAGTTGGAGATCGATA           unknown  unknown                 unknown   \n",
       "...                            ...      ...                     ...   \n",
       "GATCTAGTCATCGGAT           unknown  unknown                 unknown   \n",
       "AACTCTTAGTCCCACG           unknown  unknown                 unknown   \n",
       "CACTGTCCAGTCACGC           unknown  unknown                 unknown   \n",
       "CACCTTGGTCGACTAT           unknown  unknown                 unknown   \n",
       "TTTGTCACACGTAAGG           unknown  unknown                 unknown   \n",
       "\n",
       "                 suspension_type genetic_treatments compound_treatments  \\\n",
       "AACTGGTGTCTCTCTG            cell                                          \n",
       "ATAGGCTCAGATTTCG            cell                                          \n",
       "GCCAAATCAAGCCGTC            cell                                          \n",
       "CGGAGAAGTCGCGTCA            cell                                          \n",
       "TAGTTGGAGATCGATA            cell                                          \n",
       "...                          ...                ...                 ...   \n",
       "GATCTAGTCATCGGAT            cell                                          \n",
       "AACTCTTAGTCCCACG            cell                                          \n",
       "CACTGTCCAGTCACGC            cell                                          \n",
       "CACCTTGGTCGACTAT            cell                                          \n",
       "TTTGTCACACGTAAGG            cell                                          \n",
       "\n",
       "                 environmental_treatments combination_treatments  \n",
       "AACTGGTGTCTCTCTG                                                  \n",
       "ATAGGCTCAGATTTCG                                                  \n",
       "GCCAAATCAAGCCGTC                                                  \n",
       "CGGAGAAGTCGCGTCA                                                  \n",
       "TAGTTGGAGATCGATA                                                  \n",
       "...                                   ...                    ...  \n",
       "GATCTAGTCATCGGAT                                                  \n",
       "AACTCTTAGTCCCACG                                                  \n",
       "CACTGTCCAGTCACGC                                                  \n",
       "CACCTTGGTCGACTAT                                                  \n",
       "TTTGTCACACGTAAGG                                                  \n",
       "\n",
       "[1000 rows x 45 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"sex\"] = adata.obs[\"sex\"].cat.rename_categories({\"Unknown\": \"unknown\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "efo_lo = bt.ExperimentalFactor.public().lookup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"assay\"] = efo_lo.single_cell_rna_sequencing.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, ~adata.var_names.isin(curator.non_validated[\"var_index\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'assay'\n",
      "\u001b[92m✓\u001b[0m added 1 record \u001b[1;92mfrom public\u001b[0m with \u001b[3mExperimentalFactor.name\u001b[0m for \"assay\": 'single-cell RNA sequencing'\n",
      "\u001b[92m✓\u001b[0m \"var_index\" is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"assay\" is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"cell_type\" on \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"cell_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"development_stage\" on \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"development_stage\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"disease\" on \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'pancreatic cancer'\u001b[0m\n",
      "    \u001b[1;93m1 synonym\u001b[0m found: \u001b[1;93m\"pancreatic cancer\" → \"malignant pancreatic neoplasm\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"disease\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"donor_id\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"donor_id\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"self_reported_ethnicity\" on \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"self_reported_ethnicity\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"sex\" on \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m3 terms\u001b[0m are not validated: \u001b[1;91m'Male', 'Female', 'unknown'\u001b[0m\n",
      "    \u001b[1;93m2 synonyms\u001b[0m found: \u001b[1;93m\"Male\" → \"male\", \"Female\" → \"female\"\u001b[0m\n",
      "    → curate synonyms via \u001b[1;96m.standardize(\"sex\")\u001b[0m    for remaining terms:\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"sex\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"suspension_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"suspension_type\")\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \"tissue_type\" on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m   \u001b[1;91m1 term\u001b[0m is not validated: \u001b[1;91m'cell_line'\u001b[0m\n",
      "    → fix typos, remove non-existent values, or save terms via \u001b[1;96m.add_new_from(\"tissue_type\")\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"organism\" is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_line\" is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"genetic_treatments\" is validated against \u001b[3mGeneticPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound_treatments\" is validated against \u001b[3mCompoundPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"environmental_treatments\" is validated against \u001b[3mEnvironmentalPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"combination_treatments\" is validated against \u001b[3mCombinationPerturbation.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Recreate Curator object because we are using a new object\n",
    "curator = pts.PerturbationCurator(adata, using_key=\"test-perturbation\")\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'name'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/bionty/models.py:386\u001b[0m, in \u001b[0;36mBioRecord.public\u001b[0;34m(cls, organism, source)\u001b[0m\n\u001b[1;32m    385\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 386\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mbt_base\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;18;43m__name__\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    387\u001b[0m \u001b[43m        \u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msource\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msource_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mversion\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mversion\u001b[49m\n\u001b[1;32m    388\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    389\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m InvalidParamError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/bionty/base/entities/_celltype.py:38\u001b[0m, in \u001b[0;36mCellType.__init__\u001b[0;34m(self, organism, source, version, **kwargs)\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__init__\u001b[39m(\n\u001b[1;32m     22\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m     23\u001b[0m     organism: Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m     37\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m---> 38\u001b[0m     \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m     39\u001b[0m \u001b[43m        \u001b[49m\u001b[43msource\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     40\u001b[0m \u001b[43m        \u001b[49m\u001b[43mversion\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mversion\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     41\u001b[0m \u001b[43m        \u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     42\u001b[0m \u001b[43m        \u001b[49m\u001b[43minclude_id_prefixes\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcl\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mCL\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     43\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     44\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/bionty/base/_public_ontology.py:74\u001b[0m, in \u001b[0;36mPublicOntology.__init__\u001b[0;34m(self, source, version, organism, include_id_prefixes, include_rel)\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[38;5;66;03m# search in all available sources to get url and md5\u001b[39;00m\n\u001b[0;32m---> 74\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_source_record \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_match_sources\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     75\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_all_sources\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     76\u001b[0m \u001b[43m    \u001b[49m\u001b[43msource\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     77\u001b[0m \u001b[43m    \u001b[49m\u001b[43mversion\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mversion\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     78\u001b[0m \u001b[43m    \u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     79\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     81\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_organism \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_source_record[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124morganism\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/bionty/base/_public_ontology.py:251\u001b[0m, in \u001b[0;36mPublicOntology._match_sources\u001b[0;34m(self, ref_sources, source, version, organism)\u001b[0m\n\u001b[1;32m    250\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m row\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 251\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    252\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNo source is available with \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkwargs\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mCheck\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    253\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m `.display_available_sources()`\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    254\u001b[0m     )\n\u001b[1;32m    255\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m row\u001b[38;5;241m.\u001b[39mto_dict(orient\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrecords\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m0\u001b[39m]\n",
      "\u001b[0;31mValueError\u001b[0m: No source is available with {'organism': 'human'}\nCheck `.display_available_sources()`",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mcurator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstandardize\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mall\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m curator\u001b[38;5;241m.\u001b[39madd_new_from(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/PycharmProjects/lamindb/lamindb/_curate.py:710\u001b[0m, in \u001b[0;36mAnnDataCurator.standardize\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    698\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Replace synonyms with standardized values.\u001b[39;00m\n\u001b[1;32m    699\u001b[0m \n\u001b[1;32m    700\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    706\u001b[0m \u001b[38;5;124;03mInplace modification of the dataset.\u001b[39;00m\n\u001b[1;32m    707\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    708\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_adata\u001b[38;5;241m.\u001b[39mobs\u001b[38;5;241m.\u001b[39mcolumns \u001b[38;5;129;01mor\u001b[39;00m key \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    709\u001b[0m     \u001b[38;5;66;03m# standardize obs columns\u001b[39;00m\n\u001b[0;32m--> 710\u001b[0m     \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstandardize\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    711\u001b[0m \u001b[38;5;66;03m# in addition to the obs columns, standardize the var.index\u001b[39;00m\n\u001b[1;32m    712\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar_index\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m key \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mall\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/PycharmProjects/lamindb/lamindb/_curate.py:367\u001b[0m, in \u001b[0;36mDataFrameCurator.standardize\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    365\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m avail_keys:\n\u001b[1;32m    366\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_fields:  \u001b[38;5;66;03m# needed to exclude var_index\u001b[39;00m\n\u001b[0;32m--> 367\u001b[0m             syn_mapper \u001b[38;5;241m=\u001b[39m \u001b[43mstandardize_categories\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    368\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnon_validated\u001b[49m\u001b[43m[\u001b[49m\u001b[43mk\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    369\u001b[0m \u001b[43m                \u001b[49m\u001b[43mfield\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_fields\u001b[49m\u001b[43m[\u001b[49m\u001b[43mk\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    370\u001b[0m \u001b[43m                \u001b[49m\u001b[43musing_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_using_key\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    371\u001b[0m \u001b[43m                \u001b[49m\u001b[43msource\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sources\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mk\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    372\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    373\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    374\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_df[k] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_replace_synonyms(k, syn_mapper, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_df[k])\n\u001b[1;32m    375\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/PycharmProjects/lamindb/lamindb/_curate.py:1902\u001b[0m, in \u001b[0;36mstandardize_categories\u001b[0;34m(values, field, using_key, organism, source)\u001b[0m\n\u001b[1;32m   1900\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m {}\n\u001b[1;32m   1901\u001b[0m \u001b[38;5;66;03m# standardize values using the default instance\u001b[39;00m\n\u001b[0;32m-> 1902\u001b[0m syn_mapper \u001b[38;5;241m=\u001b[39m \u001b[43mregistry\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstandardize\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1903\u001b[0m \u001b[43m    \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1904\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfield\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfield\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfield\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1905\u001b[0m \u001b[43m    \u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1906\u001b[0m \u001b[43m    \u001b[49m\u001b[43msource\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1907\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmute\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   1908\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_mapper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m   1909\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1911\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(values) \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mlen\u001b[39m(syn_mapper):  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m   1912\u001b[0m     \u001b[38;5;66;03m# standardize values using the using_key instance\u001b[39;00m\n\u001b[1;32m   1913\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m using_key \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m using_key \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdefault\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m~/PycharmProjects/lamindb/lamindb/_can_curate.py:290\u001b[0m, in \u001b[0;36mstandardize\u001b[0;34m(cls, values, field, return_field, return_mapper, case_sensitive, mute, public_aware, keep, synonyms_field, organism, source)\u001b[0m\n\u001b[1;32m    272\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m    273\u001b[0m \u001b[38;5;129m@doc_args\u001b[39m(CanCurate\u001b[38;5;241m.\u001b[39mstandardize\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__doc__\u001b[39m)\n\u001b[1;32m    274\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mstandardize\u001b[39m(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    287\u001b[0m     source: Record \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    288\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mlist\u001b[39m[\u001b[38;5;28mstr\u001b[39m] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28mdict\u001b[39m[\u001b[38;5;28mstr\u001b[39m, \u001b[38;5;28mstr\u001b[39m]:\n\u001b[1;32m    289\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"{}\"\"\"\u001b[39;00m  \u001b[38;5;66;03m# noqa: D415\u001b[39;00m\n\u001b[0;32m--> 290\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_standardize\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    291\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mcls\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    292\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    293\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfield\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfield\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    294\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_field\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_field\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    295\u001b[0m \u001b[43m        \u001b[49m\u001b[43mreturn_mapper\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mreturn_mapper\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    296\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcase_sensitive\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcase_sensitive\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    297\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmute\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmute\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    298\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpublic_aware\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpublic_aware\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    299\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkeep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    300\u001b[0m \u001b[43m        \u001b[49m\u001b[43msynonyms_field\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msynonyms_field\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    301\u001b[0m \u001b[43m        \u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    302\u001b[0m \u001b[43m        \u001b[49m\u001b[43msource\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    303\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/PycharmProjects/lamindb/lamindb/_can_curate.py:447\u001b[0m, in \u001b[0;36m_standardize\u001b[0;34m(cls, values, field, return_field, return_mapper, case_sensitive, mute, public_aware, keep, synonyms_field, using_key, organism, source)\u001b[0m\n\u001b[1;32m    444\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m _return(result\u001b[38;5;241m=\u001b[39mstd_names_db, mapper\u001b[38;5;241m=\u001b[39mmapper)\n\u001b[1;32m    446\u001b[0m nonval \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(std_names_db)[\u001b[38;5;241m~\u001b[39mval_res]\n\u001b[0;32m--> 447\u001b[0m std_names_bt_mapper \u001b[38;5;241m=\u001b[39m \u001b[43mregistry\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpublic\u001b[49m\u001b[43m(\u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mstandardize(\n\u001b[1;32m    448\u001b[0m     nonval, return_mapper\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, mute\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m_kwargs\n\u001b[1;32m    449\u001b[0m )\n\u001b[1;32m    451\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(std_names_bt_mapper) \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m mute:\n\u001b[1;32m    452\u001b[0m     s \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(std_names_bt_mapper) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124ms\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/bionty/models.py:403\u001b[0m, in \u001b[0;36mBioRecord.public\u001b[0;34m(cls, organism, source)\u001b[0m\n\u001b[1;32m    401\u001b[0m             kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124morganism\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m organism\n\u001b[1;32m    402\u001b[0m     source \u001b[38;5;241m=\u001b[39m Source\u001b[38;5;241m.\u001b[39mfilter(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\u001b[38;5;241m.\u001b[39mfirst()\n\u001b[0;32m--> 403\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mStaticReference\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/bionty/models.py:44\u001b[0m, in \u001b[0;36mStaticReference.__init__\u001b[0;34m(self, source_record)\u001b[0m\n\u001b[1;32m     41\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m, source_record: Source) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m     42\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_source_record \u001b[38;5;241m=\u001b[39m source_record\n\u001b[1;32m     43\u001b[0m     \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__init__\u001b[39m(\n\u001b[0;32m---> 44\u001b[0m         source\u001b[38;5;241m=\u001b[39m\u001b[43msource_record\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m,\n\u001b[1;32m     45\u001b[0m         version\u001b[38;5;241m=\u001b[39msource_record\u001b[38;5;241m.\u001b[39mversion,\n\u001b[1;32m     46\u001b[0m         organism\u001b[38;5;241m=\u001b[39msource_record\u001b[38;5;241m.\u001b[39morganism,\n\u001b[1;32m     47\u001b[0m     )\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'name'"
     ]
    }
   ],
   "source": [
    "curator.standardize(\"all\")\n",
    "curator.add_new_from(\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"var_index\" is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"assay\" is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_type\" is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"development_stage\" is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"disease\" is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"donor_id\" is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"self_reported_ethnicity\" is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"sex\" is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"suspension_type\" is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"tissue_type\" is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"organism\" is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"cell_line\" is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"genetic_treatments\" is validated against \u001b[3mGeneticPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"compound_treatments\" is validated against \u001b[3mCompoundPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"environmental_treatments\" is validated against \u001b[3mEnvironmentalPerturbation.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m \"combination_treatments\" is validated against \u001b[3mCombinationPerturbation.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator.validate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All treatment columns validate but that's only because they're all empty."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curate perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Move\n",
    "adata.obs[\"genetic_treatments\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"CRISPR\", None\n",
    ")\n",
    "adata.obs[\"compound_treatments\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"drug\", None\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genetic treatments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, 'sggpx4-2', 'sglacz', 'sggpx4-1', 'sgor2j2']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(adata.obs[\"genetic_treatments\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Gene record from Bionty\u001b[0m matching \u001b[3msymbol\u001b[0m: \u001b[1;95m'GPX4'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>system</th>\n",
       "      <th>sequence</th>\n",
       "      <th>on_target_score</th>\n",
       "      <th>off_target_score</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WemtRNZv2ziV</td>\n",
       "      <td>sggpx4-1</td>\n",
       "      <td>CRISPR-Cas9</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>2024-12-10 12:44:57.321264+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid      name       system sequence on_target_score  \\\n",
       "id                                                                 \n",
       "1   WemtRNZv2ziV  sggpx4-1  CRISPR-Cas9     None            None   \n",
       "\n",
       "   off_target_score  run_id                       created_at  created_by_id  \n",
       "id                                                                           \n",
       "1              None       1 2024-12-10 12:44:57.321264+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning existing PerturbationTarget record with same name: 'Glutathione Peroxidase 4'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Gene record from Bionty\u001b[0m matching \u001b[3msynonyms\u001b[0m: \u001b[1;95m'or2j2'\u001b[0m\n",
      "\u001b[93m!\u001b[0m ambiguous validation in Bionty for 1 record: 'OR2J2'\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "Field 'id' expected a number but got [Gene(uid='4U5zlMFedTR1', symbol='OR2J2', ensembl_gene_id='ENSG00000196231', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='2cT5bXP0JAXr', symbol='OR2J2', ensembl_gene_id='ENSG00000204700', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='1NVMcJPLrAlr', symbol='OR2J2', ensembl_gene_id='ENSG00000225550', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='Nt62Ro62g1VZ', symbol='OR2J2', ensembl_gene_id='ENSG00000226000', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='3wXVmsWB2nY2', symbol='OR2J2', ensembl_gene_id='ENSG00000226347', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='4CkZekReK7p6', symbol='OR2J2', ensembl_gene_id='ENSG00000231676', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='2ttz4jYNb9Oo', symbol='OR2J2', ensembl_gene_id='ENSG00000232945', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='5iYHfK0bEf2e', symbol='OR2J2', ensembl_gene_id='ENSG00000234746', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC)].",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/django/db/models/fields/__init__.py:2123\u001b[0m, in \u001b[0;36mIntegerField.get_prep_value\u001b[0;34m(self, value)\u001b[0m\n\u001b[1;32m   2122\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 2123\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mint\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2124\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mTypeError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mTypeError\u001b[0m: int() argument must be a string, a bytes-like object or a real number, not 'RecordList'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 19\u001b[0m\n\u001b[1;32m     17\u001b[0m     gene \u001b[38;5;241m=\u001b[39m bt\u001b[38;5;241m.\u001b[39mGene(symbol\u001b[38;5;241m=\u001b[39msymbol, organism\u001b[38;5;241m=\u001b[39morganism)\u001b[38;5;241m.\u001b[39msave()\n\u001b[1;32m     18\u001b[0m target \u001b[38;5;241m=\u001b[39m wl\u001b[38;5;241m.\u001b[39mTreatmentTarget(name\u001b[38;5;241m=\u001b[39mtarget_name)\u001b[38;5;241m.\u001b[39msave()\n\u001b[0;32m---> 19\u001b[0m \u001b[43mtarget\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenes\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43madd\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     20\u001b[0m treatment\u001b[38;5;241m.\u001b[39mtargets\u001b[38;5;241m.\u001b[39madd(target)\n\u001b[1;32m     21\u001b[0m genetic_treatments\u001b[38;5;241m.\u001b[39mappend(treatment)\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/django/db/models/fields/related_descriptors.py:1253\u001b[0m, in \u001b[0;36mcreate_forward_many_to_many_manager.<locals>.ManyRelatedManager.add\u001b[0;34m(self, through_defaults, *objs)\u001b[0m\n\u001b[1;32m   1251\u001b[0m db \u001b[38;5;241m=\u001b[39m router\u001b[38;5;241m.\u001b[39mdb_for_write(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mthrough, instance\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minstance)\n\u001b[1;32m   1252\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m transaction\u001b[38;5;241m.\u001b[39matomic(using\u001b[38;5;241m=\u001b[39mdb, savepoint\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[0;32m-> 1253\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_add_items\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1254\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msource_field_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1255\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtarget_field_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1256\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mobjs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1257\u001b[0m \u001b[43m        \u001b[49m\u001b[43mthrough_defaults\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mthrough_defaults\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1258\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1259\u001b[0m     \u001b[38;5;66;03m# If this is a symmetrical m2m relation to self, add the mirror\u001b[39;00m\n\u001b[1;32m   1260\u001b[0m     \u001b[38;5;66;03m# entry in the m2m table.\u001b[39;00m\n\u001b[1;32m   1261\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msymmetrical:\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/django/db/models/fields/related_descriptors.py:1513\u001b[0m, in \u001b[0;36mcreate_forward_many_to_many_manager.<locals>.ManyRelatedManager._add_items\u001b[0;34m(self, source_field_name, target_field_name, through_defaults, *objs)\u001b[0m\n\u001b[1;32m   1510\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m   1512\u001b[0m through_defaults \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(resolve_callables(through_defaults \u001b[38;5;129;01mor\u001b[39;00m {}))\n\u001b[0;32m-> 1513\u001b[0m target_ids \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_target_ids\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtarget_field_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mobjs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1514\u001b[0m db \u001b[38;5;241m=\u001b[39m router\u001b[38;5;241m.\u001b[39mdb_for_write(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mthrough, instance\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39minstance)\n\u001b[1;32m   1515\u001b[0m can_ignore_conflicts, must_send_signals, can_fast_add \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_add_plan(\n\u001b[1;32m   1516\u001b[0m     db, source_field_name\n\u001b[1;32m   1517\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/django/db/models/fields/related_descriptors.py:1447\u001b[0m, in \u001b[0;36mcreate_forward_many_to_many_manager.<locals>.ManyRelatedManager._get_target_ids\u001b[0;34m(self, target_field_name, objs)\u001b[0m\n\u001b[1;32m   1442\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[1;32m   1443\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m instance expected, got \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1444\u001b[0m             \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel\u001b[38;5;241m.\u001b[39m_meta\u001b[38;5;241m.\u001b[39mobject_name, obj)\n\u001b[1;32m   1445\u001b[0m         )\n\u001b[1;32m   1446\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 1447\u001b[0m         target_ids\u001b[38;5;241m.\u001b[39madd(\u001b[43mtarget_field\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_prep_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m   1448\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m target_ids\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/django/db/models/fields/related.py:1148\u001b[0m, in \u001b[0;36mForeignKey.get_prep_value\u001b[0;34m(self, value)\u001b[0m\n\u001b[1;32m   1147\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_prep_value\u001b[39m(\u001b[38;5;28mself\u001b[39m, value):\n\u001b[0;32m-> 1148\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtarget_field\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_prep_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/lamindb/lib/python3.11/site-packages/django/db/models/fields/__init__.py:2125\u001b[0m, in \u001b[0;36mIntegerField.get_prep_value\u001b[0;34m(self, value)\u001b[0m\n\u001b[1;32m   2123\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mint\u001b[39m(value)\n\u001b[1;32m   2124\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mTypeError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m-> 2125\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m(\n\u001b[1;32m   2126\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mField \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m expected a number but got \u001b[39m\u001b[38;5;132;01m%r\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mname, value),\n\u001b[1;32m   2127\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "\u001b[0;31mTypeError\u001b[0m: Field 'id' expected a number but got [Gene(uid='4U5zlMFedTR1', symbol='OR2J2', ensembl_gene_id='ENSG00000196231', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='2cT5bXP0JAXr', symbol='OR2J2', ensembl_gene_id='ENSG00000204700', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='1NVMcJPLrAlr', symbol='OR2J2', ensembl_gene_id='ENSG00000225550', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='Nt62Ro62g1VZ', symbol='OR2J2', ensembl_gene_id='ENSG00000226000', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='3wXVmsWB2nY2', symbol='OR2J2', ensembl_gene_id='ENSG00000226347', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='4CkZekReK7p6', symbol='OR2J2', ensembl_gene_id='ENSG00000231676', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='2ttz4jYNb9Oo', symbol='OR2J2', ensembl_gene_id='ENSG00000232945', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC), Gene(uid='5iYHfK0bEf2e', symbol='OR2J2', ensembl_gene_id='ENSG00000234746', ncbi_gene_ids='26707', biotype='protein_coding', synonyms='DJ80I19.4|HS6M1-6|OR6-8', description='olfactory receptor family 2 subfamily J member 2 ', created_by_id=1, run_id=1, source_id=11, organism_id=1, created_at=2024-12-10 12:45:01 UTC)]."
     ]
    }
   ],
   "source": [
    "treatments = [\n",
    "    (\"sggpx4-1\", \"GPX4\", \"Glutathione Peroxidase 4\"),\n",
    "    (\"sggpx4-2\", \"GPX4\", \"Glutathione Peroxidase 4\"),\n",
    "    (\"sgor2j2\", \"or2j2\", \"Olfactory receptor family 2 subfamily J member 2\"),\n",
    "    (\"sglacz\", \"lacz\", \"beta-galactosidase control\"),  # Control from E. coli\n",
    "]\n",
    "organism = bt.Organism.lookup().human\n",
    "\n",
    "genetic_treatments = []\n",
    "for name, symbol, target_name in treatments:\n",
    "    treatment = wl.GeneticTreatment(system=\"CRISPR-Cas9\", name=name).save()\n",
    "    if symbol != \"lacz\":\n",
    "        gene_result = bt.Gene.from_source(symbol=symbol, organism=organism)\n",
    "        gene = gene_result[0] if isinstance(gene_result, RecordList) else gene_result\n",
    "        gene = gene.save()\n",
    "    else:\n",
    "        gene = bt.Gene(symbol=symbol, organism=organism).save()\n",
    "    target = wl.TreatmentTarget(name=target_name).save()\n",
    "    target.genes.add(gene)\n",
    "    treatment.targets.add(target)\n",
    "    genetic_treatments.append(treatment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'var_index' is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'assay' is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_type' is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'development_stage' is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'disease' is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'donor_id' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'self_reported_ethnicity' is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'sex' is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'suspension_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'tissue_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'organism' is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_line' is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'genetic_treatments' is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mcompound_treatments\u001b[0m on \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m14 terms\u001b[0m are not validated: \u001b[1;91m'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'brd3379', 'JQ1', 'azd5591', 'control', ...\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('compound_treatments')\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'environmental_treatments' is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'combination_treatments' is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator.validate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m8 Compound records from Bionty\u001b[0m matching \u001b[3mname\u001b[0m: \u001b[1;95m'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'JQ1', 'everolimus'\u001b[0m\n",
      "\u001b[93m!\u001b[0m \u001b[1;91mdid not create\u001b[0m Compound records for \u001b[1;93m6 non-validated\u001b[0m \u001b[3mnames\u001b[0m: \u001b[1;93m'azd5591', 'brd3379', 'control', 'idasanutlin', 'prexasertib', 'taselisib'\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "compounds = wl.Compound.from_values(adata.obs[\"compound_treatments\"], field=\"name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The remaining compounds are not in chebi and we create records for them\n",
    "for missing in [\n",
    "    \"azd5591\",\n",
    "    \"brd3379\",\n",
    "    \"control\",\n",
    "    \"idasanutlin\",\n",
    "    \"prexasertib\",\n",
    "    \"taselisib\",\n",
    "]:\n",
    "    compounds.append(wl.Compound(name=missing))\n",
    "ln.save(compounds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_metadata = adata.obs[adata.obs[\"compound_treatments\"].notna()]\n",
    "\n",
    "unique_treatments = drug_metadata[\n",
    "    [\"perturbation\", \"dose_unit\", \"dose_value\"]\n",
    "].drop_duplicates()\n",
    "\n",
    "compound_treatments = []\n",
    "for _, row in unique_treatments.iterrows():\n",
    "    compound = wl.Compound.get(name=row[\"perturbation\"])\n",
    "    treatment = wl.CompoundTreatment(\n",
    "        name=compound.name,\n",
    "        concentration=row[\"dose_value\"],\n",
    "        concentration_unit=row[\"dose_unit\"],\n",
    "    )\n",
    "    compound_treatments.append(treatment)\n",
    "\n",
    "ln.save(compound_treatments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P36507'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P00533'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q9UK79'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q15303'\u001b[0m\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'MAPK/ERK pathway'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P15056'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P23921'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P10415'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q07812'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i6XpYvMC</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:10:57.608736+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid              name description  run_id  \\\n",
       "id                                                   \n",
       "4   i6XpYvMC  MAPK/ERK pathway        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "4  2024-11-11 15:10:57.608736+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P49721'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SnJAC9fR</td>\n",
       "      <td>Apoptosis regulation</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:04.651378+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                  name description  run_id  \\\n",
       "id                                                       \n",
       "7   SnJAC9fR  Apoptosis regulation        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "7  2024-11-11 15:11:04.651378+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O60885'\u001b[0m\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Transcription regulation (BET proteins)'\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Apoptosis regulation'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q07820'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O14757'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i6XpYvMC</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:10:57.608736+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fuAhPhXj</td>\n",
       "      <td>Proteasome pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:07.553672+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                name description  run_id  \\\n",
       "id                                                     \n",
       "4   i6XpYvMC    MAPK/ERK pathway        None       2   \n",
       "8   fuAhPhXj  Proteasome pathway        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "4  2024-11-11 15:10:57.608736+00:00              1  \n",
       "8  2024-11-11 15:11:07.553672+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P42336'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O00329'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P48736'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SnJAC9fR</td>\n",
       "      <td>Apoptosis regulation</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:04.651378+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5KWr4lMW</td>\n",
       "      <td>Transcription regulation (BET proteins)</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:08.744610+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                                     name description  run_id  \\\n",
       "id                                                                          \n",
       "7   SnJAC9fR                     Apoptosis regulation        None       2   \n",
       "9   5KWr4lMW  Transcription regulation (BET proteins)        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "7  2024-11-11 15:11:04.651378+00:00              1  \n",
       "9  2024-11-11 15:11:08.744610+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q00987'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>irV5vyHx</td>\n",
       "      <td>PI3K/AKT/mTOR pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:12.919341+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>i6XpYvMC</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:10:57.608736+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fuAhPhXj</td>\n",
       "      <td>Proteasome pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>2024-11-11 15:11:07.553672+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                   name description  run_id  \\\n",
       "id                                                        \n",
       "12  irV5vyHx  PI3K/AKT/mTOR pathway        None       2   \n",
       "4   i6XpYvMC       MAPK/ERK pathway        None       2   \n",
       "8   fuAhPhXj     Proteasome pathway        None       2   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "12 2024-11-11 15:11:12.919341+00:00              1  \n",
       "4  2024-11-11 15:10:57.608736+00:00              1  \n",
       "8  2024-11-11 15:11:07.553672+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P42345'\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "compounds_to_targets = {\n",
    "    \"trametinib\": (\"MAPK/ERK pathway\", [\"P36507\"]),\n",
    "    \"afatinib\": (\"EGFR, HER2, HER4 signaling\", [\"P00533\", \"Q9UK79\", \"Q15303\"]),\n",
    "    \"dabrafenib\": (\"MAPK/ERK pathway\", [\"P15056\"]),\n",
    "    \"gemcitabine\": (\"DNA synthesis inhibition\", [\"P23921\"]),  # No single protein target\n",
    "    \"navitoclax\": (\"Apoptosis regulation\", [\"P10415\", \"Q07812\"]),\n",
    "    \"bortezomib\": (\"Proteasome pathway\", [\"P49721\"]),\n",
    "    \"brd3379\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"JQ1\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"azd5591\": (\"Apoptosis regulation\", [\"Q07820\"]),\n",
    "    \"control\": (\"Baseline\", [None]),  # No target for control\n",
    "    \"prexasertib\": (\"DNA damage response\", [\"O14757\"]),\n",
    "    \"taselisib\": (\"PI3K/AKT/mTOR pathway\", [\"P42336\", \"O00329\", \"P48736\"]),\n",
    "    \"idasanutlin\": (\"p53 regulation\", [\"Q00987\"]),\n",
    "    \"everolimus\": (\"mTOR pathway\", [\"P42345\"])\n",
    "}\n",
    "\n",
    "for compound_treatment_name, targets_tuple in compounds_to_targets.items():\n",
    "    compound_treatment = wl.CompoundTreatment.get(name=compound_treatment_name)\n",
    "    target = wl.TreatmentTarget(name=targets_tuple[0]).save()\n",
    "    proteins = []\n",
    "    for id in targets_tuple[1]:\n",
    "        if id is not None:\n",
    "            proteins.append(bt.Protein.from_source(uniprotkb_id=id).save())\n",
    "    target.proteins.set(proteins)\n",
    "    compound_treatment.targets.add(target)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = opr.Reference(\n",
    "    name=\"Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action\",\n",
    "    abbr=\"McFarland 2020\",\n",
    "    url=\"https://www.nature.com/articles/s41467-020-17440-w\",\n",
    "    doi=\"10.1038/s41467-020-17440-w\",\n",
    "    text=(\n",
    "        \"Assays to study cancer cell responses to pharmacologic or genetic perturbations are typically \"\n",
    "        \"restricted to using simple phenotypic readouts such as proliferation rate. Information-rich assays, \"\n",
    "        \"such as gene-expression profiling, have generally not permitted efficient profiling of a given \"\n",
    "        \"perturbation across multiple cellular contexts. Here, we develop MIX-Seq, a method for multiplexed \"\n",
    "        \"transcriptional profiling of post-perturbation responses across a mixture of samples with single-cell \"\n",
    "        \"resolution, using SNP-based computational demultiplexing of single-cell RNA-sequencing data. We show \"\n",
    "        \"that MIX-Seq can be used to profile responses to chemical or genetic perturbations across pools of 100 \"\n",
    "        \"or more cancer cell lines. We combine it with Cell Hashing to further multiplex additional experimental \"\n",
    "        \"conditions, such as post-treatment time points or drug doses. Analyzing the high-content readout of \"\n",
    "        \"scRNA-seq reveals both shared and context-specific transcriptional response components that can identify \"\n",
    "        \"drug mechanism of action and enable prediction of long-term cell viability from short-term transcriptional \"\n",
    "        \"responses to treatment.\"\n",
    "    ),\n",
    ").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove unused columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.drop(\n",
    "    [\n",
    "        \"depmap_id\",\n",
    "        \"cancer\",\n",
    "        \"cell_quality\",\n",
    "        \"channel\",\n",
    "        \"perturbation\",\n",
    "        \"perturbation_type\",\n",
    "        \"singlet_dev\",\n",
    "        \"singlet_dev_z\",\n",
    "        \"singlet_margin\",\n",
    "        \"singlet_z_margin\",\n",
    "        \"nperts\",\n",
    "        \"ngenes\",\n",
    "        \"ncounts\",\n",
    "        \"cell_det_rate\",\n",
    "        \"doublet_GMM_prob\",\n",
    "        \"doublet_dev_imp\",\n",
    "        \"doublet_z_margin\",\n",
    "        'doublet_CL1',\n",
    "        'doublet_CL2'\n",
    "    ],\n",
    "    axis=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Register curated artifact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'var_index' is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'assay' is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_type' is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'development_stage' is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'disease' is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'donor_id' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'self_reported_ethnicity' is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'sex' is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'suspension_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'tissue_type' is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'organism' is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'cell_line' is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'genetic_treatments' is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'compound_treatments' is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'environmental_treatments' is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m 'combination_treatments' is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar description exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>version</th>\n",
       "      <th>is_latest</th>\n",
       "      <th>description</th>\n",
       "      <th>key</th>\n",
       "      <th>suffix</th>\n",
       "      <th>type</th>\n",
       "      <th>size</th>\n",
       "      <th>hash</th>\n",
       "      <th>n_objects</th>\n",
       "      <th>n_observations</th>\n",
       "      <th>_hash_type</th>\n",
       "      <th>_accessor</th>\n",
       "      <th>visibility</th>\n",
       "      <th>_key_is_virtual</th>\n",
       "      <th>storage_id</th>\n",
       "      <th>transform_id</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Xk7Qaik9vBLV4PKf0001</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>McFarland 2020 preprocessed</td>\n",
       "      <td>None</td>\n",
       "      <td>.h5ad</td>\n",
       "      <td>dataset</td>\n",
       "      <td>2511528</td>\n",
       "      <td>Iz4mVUpIruvtABfA6D3vQA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>md5</td>\n",
       "      <td>AnnData</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2024-11-11 14:44:54.959908+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     uid version  is_latest                  description  \\\n",
       "id                                                                         \n",
       "2   Xk7Qaik9vBLV4PKf0001    None       True  McFarland 2020 preprocessed   \n",
       "\n",
       "     key suffix     type     size                    hash n_objects  \\\n",
       "id                                                                    \n",
       "2   None  .h5ad  dataset  2511528  Iz4mVUpIruvtABfA6D3vQA      None   \n",
       "\n",
       "   n_observations _hash_type _accessor  visibility  _key_is_virtual  \\\n",
       "id                                                                    \n",
       "2            None        md5   AnnData           1             True   \n",
       "\n",
       "    storage_id  transform_id  run_id                       created_at  \\\n",
       "id                                                                      \n",
       "2            3             3       3 2024-11-11 14:44:54.959908+00:00   \n",
       "\n",
       "    created_by_id  \n",
       "id                 \n",
       "2               1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[93m!\u001b[0m    \u001b[1;93m11 unique terms\u001b[0m (42.30%) are not validated for \u001b[3mname\u001b[0m: \u001b[1;93m'dose_unit', 'dose_value', 'hash_assignment', 'hash_tag', 'num_SNPs', 'singlet_ID', 'time', 'tot_reads', 'percent_mito', 'percent_ribo', ...\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "artifact = curator.save_artifact(description=\"McFarland AnnData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the perturbations and references\n",
    "artifact.genetic_treatments.set(genetic_treatments)\n",
    "artifact.compound_treatments.set(compound_treatments)\n",
    "artifact.references.add(reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;92mArtifact\u001b[0m(uid='A2xWHSPBuPgBhcHi0000', is_latest=True, description='McFarland AnnData', suffix='.h5ad', type='dataset', size=3345456, hash='jzxUs9DOPJewAKOb6ZMaGg', n_observations=1000, _hash_type='md5', _accessor='AnnData', visibility=1, _key_is_virtual=True, created_at=2024-11-11 15:11:20 UTC)\n",
      "  \u001b[3mProvenance\u001b[0m\n",
      "    .storage = '/home/zeth/PycharmProjects/pertpy-datasets/scripts/lamindb_datasets/test-perturbation'\n",
      "    .transform = 'Curate perturbation dataset with `PerturbationCurator`'\n",
      "    .run = 2024-11-11 15:10:01 UTC\n",
      "    .created_by = 'zethson'\n",
      "  \u001b[3mLabels\u001b[0m\n",
      "    .references = 'Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action'\n",
      "    .genetic_treatments = 'sggpx4-1', 'sggpx4-2', 'sgor2j2', 'sglacz'\n",
      "    .compound_treatments = 'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'brd3379', 'JQ1', 'azd5591', 'control', ...\n",
      "    .organisms = 'human'\n",
      "    .cell_types = 'unknown'\n",
      "    .diseases = 'colorectal cancer', 'ovarian cancer', 'sarcoma', 'bile duct cancer', 'bone cancer', 'liver cancer', 'urinary bladder carcinoma', 'prostate cancer', 'gastric cancer', 'uterine corpus cancer', ...\n",
      "    .cell_lines = 'NUGC-3', 'HuCCT1', 'WM-266-4', 'NCI-H838', 'Ishikawa (Heraklio) 02 ER-', 'SNU-761', 'NMC-G1', 'COLO-680N', 'OVK18', 'NCI-H1048', ...\n",
      "    .phenotypes = 'male', 'female', 'unknown'\n",
      "    .experimental_factors = 'single-cell RNA sequencing'\n",
      "    .developmental_stages = 'unknown'\n",
      "    .ethnicities = 'unknown'\n",
      "    .ulabels = 'unknown', 'cell', 'cell_line'\n",
      "  \u001b[3mFeatures\u001b[0m\n",
      "    'assay' = 'single-cell RNA sequencing'\n",
      "    'cell_line' = '22Rv1', '253J-BV', '42-MG-BA', '639-V', '647-V', '769-P', '786-O', '8505C', 'A-375', 'A2780', ...\n",
      "    'cell_type' = 'unknown'\n",
      "    'development_stage' = 'unknown'\n",
      "    'disease' = 'bile duct cancer', 'bone cancer', 'brain cancer', 'breast cancer', 'colorectal cancer', 'esophageal cancer', 'gastric cancer', 'head and neck cancer', 'kidney cancer', 'liver cancer', ...\n",
      "    'donor_id' = 'unknown'\n",
      "    'organism' = 'human'\n",
      "    'self_reported_ethnicity' = 'unknown'\n",
      "    'sex' = 'female', 'male', 'unknown'\n",
      "    'suspension_type' = 'cell'\n",
      "    'tissue_type' = 'cell_line'\n",
      "  \u001b[3mFeature sets\u001b[0m\n",
      "    'var' = 'MAGED2', 'CPE', 'DDX43', 'SPANXA2', 'LNCPRESS1', 'TMEM175', 'None', 'MMD', 'LINC01834', 'DYNC2I1', 'CCRL2', 'MRPS6', 'KCNH1', 'TBILA', 'SLC5A7', 'GSX2', 'ATP5MG', 'TMEM190', 'GTF2H1'\n",
      "    'obs' = 'assay', 'cell_type', 'development_stage', 'disease', 'donor_id', 'self_reported_ethnicity', 'sex', 'suspension_type', 'tissue_type', 'organism', 'cell_line', 'genetic_treatments', 'compound_treatments', 'environmental_treatments', 'combination_treatments'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "artifact.describe()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lamindb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
