{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# McFarland 2020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install pertpy[validator]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: laminlabs/pertpy-datasets\n"
     ]
    }
   ],
   "source": [
    "# Using a local instance here but in practice, we use `laminlabs/pertpy-datasets`\n",
    "!lamin load laminlabs/pertpy-datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m connected lamindb: laminlabs/pertpy-datasets\n"
     ]
    }
   ],
   "source": [
    "import lamindb as ln\n",
    "import bionty as bt\n",
    "import wetlab as wl\n",
    "import findrefs as fr\n",
    "import pertpy as pt\n",
    "import scanpy as sc\n",
    "\n",
    "THIS_INSTANCE = \"laminlabs/pertpy-datasets\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m notebook imports: bionty==0.52.0 findrefs==0.1.0 lamindb==0.76.15 pertpy==0.9.4 scanpy==1.10.3 wetlab==0.34.0\n",
      "\u001b[92m→\u001b[0m loaded Transform('HIRTYxL3'), started Run('r4d3DOTu') at 2024-10-30 13:27:28 UTC\n"
     ]
    }
   ],
   "source": [
    "ln.track(\"HIRTYxL3aZc70000\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>cancer</th>\n",
       "      <th>cell_det_rate</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_quality</th>\n",
       "      <th>channel</th>\n",
       "      <th>disease</th>\n",
       "      <th>dose_unit</th>\n",
       "      <th>dose_value</th>\n",
       "      <th>doublet_CL1</th>\n",
       "      <th>...</th>\n",
       "      <th>singlet_z_margin</th>\n",
       "      <th>time</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>tot_reads</th>\n",
       "      <th>nperts</th>\n",
       "      <th>ngenes</th>\n",
       "      <th>ncounts</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>percent_ribo</th>\n",
       "      <th>chembl-ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGACAAGCC</th>\n",
       "      <td>ACH-000174</td>\n",
       "      <td>True</td>\n",
       "      <td>0.114940</td>\n",
       "      <td>CAL62</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>thyroid cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>5.0</td>\n",
       "      <td>CAL62_THYROID</td>\n",
       "      <td>...</td>\n",
       "      <td>14.820933</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1094</td>\n",
       "      <td>1</td>\n",
       "      <td>3758</td>\n",
       "      <td>17155.0</td>\n",
       "      <td>4.447683</td>\n",
       "      <td>28.079277</td>\n",
       "      <td>CHEMBL443684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAGATGAG</th>\n",
       "      <td>ACH-000601</td>\n",
       "      <td>True</td>\n",
       "      <td>0.122660</td>\n",
       "      <td>MIAPACA2</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>pancreatic cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>5.0</td>\n",
       "      <td>MIAPACA2_PANCREAS</td>\n",
       "      <td>...</td>\n",
       "      <td>14.746431</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>1009</td>\n",
       "      <td>1</td>\n",
       "      <td>4009</td>\n",
       "      <td>19764.0</td>\n",
       "      <td>7.468124</td>\n",
       "      <td>25.460433</td>\n",
       "      <td>CHEMBL443684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCGTTGCC</th>\n",
       "      <td>ACH-000601</td>\n",
       "      <td>True</td>\n",
       "      <td>0.176608</td>\n",
       "      <td>MIAPACA2</td>\n",
       "      <td>normal</td>\n",
       "      <td>nan</td>\n",
       "      <td>pancreatic cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>5.0</td>\n",
       "      <td>MIAPACA2_PANCREAS</td>\n",
       "      <td>...</td>\n",
       "      <td>16.011316</td>\n",
       "      <td>24</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>3022</td>\n",
       "      <td>1</td>\n",
       "      <td>5772</td>\n",
       "      <td>47352.0</td>\n",
       "      <td>2.755955</td>\n",
       "      <td>27.965028</td>\n",
       "      <td>CHEMBL443684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   DepMap_ID  cancer  cell_det_rate cell_line cell_quality  \\\n",
       "AAACCTGAGACAAGCC  ACH-000174    True       0.114940     CAL62       normal   \n",
       "AAACCTGAGAGATGAG  ACH-000601    True       0.122660  MIAPACA2       normal   \n",
       "AAACCTGAGCGTTGCC  ACH-000601    True       0.176608  MIAPACA2       normal   \n",
       "\n",
       "                 channel            disease dose_unit  dose_value  \\\n",
       "AAACCTGAGACAAGCC     nan     thyroid cancer        µM         5.0   \n",
       "AAACCTGAGAGATGAG     nan  pancreatic cancer        µM         5.0   \n",
       "AAACCTGAGCGTTGCC     nan  pancreatic cancer        µM         5.0   \n",
       "\n",
       "                        doublet_CL1  ... singlet_z_margin  time  tissue_type  \\\n",
       "AAACCTGAGACAAGCC      CAL62_THYROID  ...        14.820933    24    cell_line   \n",
       "AAACCTGAGAGATGAG  MIAPACA2_PANCREAS  ...        14.746431    24    cell_line   \n",
       "AAACCTGAGCGTTGCC  MIAPACA2_PANCREAS  ...        16.011316    24    cell_line   \n",
       "\n",
       "                  tot_reads nperts ngenes  ncounts percent_mito  percent_ribo  \\\n",
       "AAACCTGAGACAAGCC       1094      1   3758  17155.0     4.447683     28.079277   \n",
       "AAACCTGAGAGATGAG       1009      1   4009  19764.0     7.468124     25.460433   \n",
       "AAACCTGAGCGTTGCC       3022      1   5772  47352.0     2.755955     27.965028   \n",
       "\n",
       "                     chembl-ID  \n",
       "AAACCTGAGACAAGCC  CHEMBL443684  \n",
       "AAACCTGAGAGATGAG  CHEMBL443684  \n",
       "AAACCTGAGCGTTGCC  CHEMBL443684  \n",
       "\n",
       "[3 rows x 36 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = pt.dt.mcfarland_2020()\n",
    "adata.obs.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate an embedding because CELLxGENE requires one\n",
    "sc.tl.pca(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curator non-perturbation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"disease\"] = adata.obs[\"disease\"].cat.rename_categories(\n",
    "    {\n",
    "        \"colon/colorectal cancer\": \"colorectal cancer\",\n",
    "        \"rhabdoid\": \"rhabdoid tumor\",\n",
    "        \"bladder cancer\": \"urinary bladder carcinoma\",\n",
    "        \"endometrial/uterine cancer\": \"uterine corpus cancer\",\n",
    "    }\n",
    ")\n",
    "\n",
    "adata.obs[\"cell_line\"] = bt.CellLine.public(\n",
    "    source=bt.Source.filter(name=\"depmap\").one()\n",
    ").standardize(adata.obs[\"cell_line\"], field=\"name\")\n",
    "\n",
    "adata.obs[\"cell_line\"] = adata.obs[\"cell_line\"].astype('category')\n",
    "\n",
    "# JQ1 is always uppercased\n",
    "adata.obs[\"perturbation\"] = adata.obs[\"perturbation\"].cat.rename_categories(\n",
    "    lambda category: category if category == \"JQ1\" else category.lower()\n",
    ")\n",
    "\n",
    "adata.obs = adata.obs.rename(columns={\"DepMap_ID\": \"depmap_id\"})\n",
    "adata.obs = adata.obs.drop(columns=\"percent.mito\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m added defaults to the AnnData object: {'assay': 'unknown', 'cell_type': 'unknown', 'development_stage': 'unknown', 'donor_id': 'unknown', 'self_reported_ethnicity': 'unknown', 'suspension_type': 'cell', 'genetic_treatments': '', 'compound_treatments': '', 'environmental_treatments': '', 'combination_treatments': ''}\n",
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mlaminlabs/pertpy-datasets\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n"
     ]
    },
    {
     "ename": "IntegrityError",
     "evalue": "duplicate key value violates unique constraint \"lnschema_bionty_gene_uid_key\"\nDETAIL:  Key (uid)=(3BEZYDTfG7t4) already exists.\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUniqueViolation\u001b[0m                           Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/backends/utils.py:105\u001b[0m, in \u001b[0;36mCursorWrapper._execute\u001b[0;34m(self, sql, params, *ignored_wrapper_args)\u001b[0m\n\u001b[1;32m    104\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 105\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcursor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43msql\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mUniqueViolation\u001b[0m: duplicate key value violates unique constraint \"lnschema_bionty_gene_uid_key\"\nDETAIL:  Key (uid)=(3BEZYDTfG7t4) already exists.\n",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mIntegrityError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m curator \u001b[38;5;241m=\u001b[39m pt\u001b[38;5;241m.\u001b[39mdt\u001b[38;5;241m.\u001b[39mPerturbationCurator(\n\u001b[1;32m      2\u001b[0m     adata, using_key\u001b[38;5;241m=\u001b[39mTHIS_INSTANCE\n\u001b[1;32m      3\u001b[0m )\n\u001b[0;32m----> 4\u001b[0m \u001b[43mcurator\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/PycharmProjects/pertpy/pertpy/data/_perturbation_validator.py:99\u001b[0m, in \u001b[0;36mPerturbationCurator.validate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     97\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mvalidate\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n\u001b[1;32m     98\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Validates the AnnData object against cellxgene and pertpy's requirements.\"\"\"\u001b[39;00m\n\u001b[0;32m---> 99\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/PycharmProjects/cellxgene-lamin/cellxgene_lamin/curate.py:285\u001b[0m, in \u001b[0;36mCurator.validate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    280\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(matching_keys) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    281\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    282\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnable to find an embedding key. Please calculate an embedding.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    283\u001b[0m     )\n\u001b[0;32m--> 285\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalidate\u001b[49m\u001b[43m(\u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/lamindb/_curate.py:547\u001b[0m, in \u001b[0;36mAnnDataCurator.validate\u001b[0;34m(self, organism)\u001b[0m\n\u001b[1;32m    542\u001b[0m     logger\u001b[38;5;241m.\u001b[39mimportant(\n\u001b[1;32m    543\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalidating metadata using registries of instance \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcolors\u001b[38;5;241m.\u001b[39mitalic(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_using_key)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    544\u001b[0m     )\n\u001b[1;32m    546\u001b[0m \u001b[38;5;66;03m# add all validated records to the current instance\u001b[39;00m\n\u001b[0;32m--> 547\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_update_registry_all\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    549\u001b[0m validated_var, non_validated_var \u001b[38;5;241m=\u001b[39m validate_categories(\n\u001b[1;32m    550\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_adata\u001b[38;5;241m.\u001b[39mvar\u001b[38;5;241m.\u001b[39mindex,\n\u001b[1;32m    551\u001b[0m     field\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_var_field,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    557\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_kwargs,  \u001b[38;5;66;03m# type: ignore\u001b[39;00m\n\u001b[1;32m    558\u001b[0m )\n\u001b[1;32m    559\u001b[0m validated_obs, non_validated_obs \u001b[38;5;241m=\u001b[39m validate_categories_in_df(\n\u001b[1;32m    560\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_adata\u001b[38;5;241m.\u001b[39mobs,\n\u001b[1;32m    561\u001b[0m     fields\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcategoricals,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    565\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_kwargs,\n\u001b[1;32m    566\u001b[0m )\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/lamindb/_curate.py:516\u001b[0m, in \u001b[0;36mAnnDataCurator._update_registry_all\u001b[0;34m(self, validated_only, **kwargs)\u001b[0m\n\u001b[1;32m    514\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Save labels for all features.\"\"\"\u001b[39;00m\n\u001b[1;32m    515\u001b[0m logger\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msaving validated records of \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mvar_index\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 516\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_save_from_var_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvalidated_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidated_only\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    517\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_obs_fields\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m    518\u001b[0m     logger\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msaving validated terms of \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/lamindb/_curate.py:501\u001b[0m, in \u001b[0;36mAnnDataCurator._save_from_var_index\u001b[0;34m(self, validated_only, organism)\u001b[0m\n\u001b[1;32m    497\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_save_from_var_index\u001b[39m(\n\u001b[1;32m    498\u001b[0m     \u001b[38;5;28mself\u001b[39m, validated_only: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, organism: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    499\u001b[0m ):\n\u001b[1;32m    500\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Save variable records.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 501\u001b[0m     \u001b[43mupdate_registry\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    502\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_adata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    503\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfield\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    504\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkey\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mvar_index\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    505\u001b[0m \u001b[43m        \u001b[49m\u001b[43msave_function\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m.add_new_from_var_index()\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    506\u001b[0m \u001b[43m        \u001b[49m\u001b[43musing_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_using_key\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    507\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalidated_only\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mvalidated_only\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    508\u001b[0m \u001b[43m        \u001b[49m\u001b[43morganism\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43morganism\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    509\u001b[0m \u001b[43m        \u001b[49m\u001b[43msource\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sources\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mvar_index\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    510\u001b[0m \u001b[43m        \u001b[49m\u001b[43mexclude\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_exclude\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mvar_index\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    511\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/lamindb/_curate.py:1511\u001b[0m, in \u001b[0;36mupdate_registry\u001b[0;34m(values, field, key, save_function, using_key, validated_only, df, organism, dtype, source, standardize, warning, exclude, **kwargs)\u001b[0m\n\u001b[1;32m   1509\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m source:\n\u001b[1;32m   1510\u001b[0m     public_records \u001b[38;5;241m=\u001b[39m [r \u001b[38;5;28;01mfor\u001b[39;00m r \u001b[38;5;129;01min\u001b[39;00m public_records \u001b[38;5;28;01mif\u001b[39;00m r\u001b[38;5;241m.\u001b[39msource\u001b[38;5;241m.\u001b[39muid \u001b[38;5;241m==\u001b[39m source\u001b[38;5;241m.\u001b[39muid]\n\u001b[0;32m-> 1511\u001b[0m \u001b[43mln_save\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpublic_records\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1512\u001b[0m labels_saved[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfrom public\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m   1513\u001b[0m     \u001b[38;5;28mgetattr\u001b[39m(r, field\u001b[38;5;241m.\u001b[39mfield\u001b[38;5;241m.\u001b[39mname) \u001b[38;5;28;01mfor\u001b[39;00m r \u001b[38;5;129;01min\u001b[39;00m public_records\n\u001b[1;32m   1514\u001b[0m ]\n\u001b[1;32m   1515\u001b[0m non_public_labels \u001b[38;5;241m=\u001b[39m [i \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m values \u001b[38;5;28;01mif\u001b[39;00m i \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m labels_saved[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfrom public\u001b[39m\u001b[38;5;124m\"\u001b[39m]]\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/lamindb/_save.py:83\u001b[0m, in \u001b[0;36msave\u001b[0;34m(records, ignore_conflicts)\u001b[0m\n\u001b[1;32m     79\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m non_artifacts:\n\u001b[1;32m     80\u001b[0m     non_artifacts_old, non_artifacts_new \u001b[38;5;241m=\u001b[39m partition(\n\u001b[1;32m     81\u001b[0m         \u001b[38;5;28;01mlambda\u001b[39;00m r: r\u001b[38;5;241m.\u001b[39m_state\u001b[38;5;241m.\u001b[39madding \u001b[38;5;129;01mor\u001b[39;00m r\u001b[38;5;241m.\u001b[39mpk \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m, non_artifacts\n\u001b[1;32m     82\u001b[0m     )\n\u001b[0;32m---> 83\u001b[0m     \u001b[43mbulk_create\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnon_artifacts_new\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mignore_conflicts\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mignore_conflicts\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     84\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m non_artifacts_old:\n\u001b[1;32m     85\u001b[0m         bulk_update(non_artifacts_old)\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/lamindb/_save.py:114\u001b[0m, in \u001b[0;36mbulk_create\u001b[0;34m(records, ignore_conflicts)\u001b[0m\n\u001b[1;32m    112\u001b[0m     records_by_orm[record\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m]\u001b[38;5;241m.\u001b[39mappend(record)\n\u001b[1;32m    113\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m registry, records \u001b[38;5;129;01min\u001b[39;00m records_by_orm\u001b[38;5;241m.\u001b[39mitems():\n\u001b[0;32m--> 114\u001b[0m     \u001b[43mregistry\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobjects\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbulk_create\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrecords\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mignore_conflicts\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mignore_conflicts\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/models/manager.py:87\u001b[0m, in \u001b[0;36mBaseManager._get_queryset_methods.<locals>.create_method.<locals>.manager_method\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m     85\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(method)\n\u001b[1;32m     86\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmanager_method\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 87\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mgetattr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_queryset\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/models/query.py:835\u001b[0m, in \u001b[0;36mQuerySet.bulk_create\u001b[0;34m(self, objs, batch_size, ignore_conflicts, update_conflicts, update_fields, unique_fields)\u001b[0m\n\u001b[1;32m    833\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m objs_without_pk:\n\u001b[1;32m    834\u001b[0m     fields \u001b[38;5;241m=\u001b[39m [f \u001b[38;5;28;01mfor\u001b[39;00m f \u001b[38;5;129;01min\u001b[39;00m fields \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(f, AutoField)]\n\u001b[0;32m--> 835\u001b[0m     returned_columns \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_batched_insert\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    836\u001b[0m \u001b[43m        \u001b[49m\u001b[43mobjs_without_pk\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    837\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    838\u001b[0m \u001b[43m        \u001b[49m\u001b[43mbatch_size\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    839\u001b[0m \u001b[43m        \u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon_conflict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    840\u001b[0m \u001b[43m        \u001b[49m\u001b[43mupdate_fields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mupdate_fields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    841\u001b[0m \u001b[43m        \u001b[49m\u001b[43munique_fields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43munique_fields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    842\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    843\u001b[0m     connection \u001b[38;5;241m=\u001b[39m connections[\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdb]\n\u001b[1;32m    844\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    845\u001b[0m         connection\u001b[38;5;241m.\u001b[39mfeatures\u001b[38;5;241m.\u001b[39mcan_return_rows_from_bulk_insert\n\u001b[1;32m    846\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m on_conflict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    847\u001b[0m     ):\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/models/query.py:1875\u001b[0m, in \u001b[0;36mQuerySet._batched_insert\u001b[0;34m(self, objs, fields, batch_size, on_conflict, update_fields, unique_fields)\u001b[0m\n\u001b[1;32m   1870\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m item \u001b[38;5;129;01min\u001b[39;00m [objs[i : i \u001b[38;5;241m+\u001b[39m batch_size] \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;28mlen\u001b[39m(objs), batch_size)]:\n\u001b[1;32m   1871\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m bulk_return \u001b[38;5;129;01mand\u001b[39;00m (\n\u001b[1;32m   1872\u001b[0m         on_conflict \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mor\u001b[39;00m on_conflict \u001b[38;5;241m==\u001b[39m OnConflict\u001b[38;5;241m.\u001b[39mUPDATE\n\u001b[1;32m   1873\u001b[0m     ):\n\u001b[1;32m   1874\u001b[0m         inserted_rows\u001b[38;5;241m.\u001b[39mextend(\n\u001b[0;32m-> 1875\u001b[0m             \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_insert\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1876\u001b[0m \u001b[43m                \u001b[49m\u001b[43mitem\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1877\u001b[0m \u001b[43m                \u001b[49m\u001b[43mfields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1878\u001b[0m \u001b[43m                \u001b[49m\u001b[43musing\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdb\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1879\u001b[0m \u001b[43m                \u001b[49m\u001b[43mon_conflict\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mon_conflict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1880\u001b[0m \u001b[43m                \u001b[49m\u001b[43mupdate_fields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mupdate_fields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1881\u001b[0m \u001b[43m                \u001b[49m\u001b[43munique_fields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43munique_fields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1882\u001b[0m \u001b[43m                \u001b[49m\u001b[43mreturning_fields\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_meta\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdb_returning_fields\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1883\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1884\u001b[0m         )\n\u001b[1;32m   1885\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1886\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_insert(\n\u001b[1;32m   1887\u001b[0m             item,\n\u001b[1;32m   1888\u001b[0m             fields\u001b[38;5;241m=\u001b[39mfields,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1892\u001b[0m             unique_fields\u001b[38;5;241m=\u001b[39munique_fields,\n\u001b[1;32m   1893\u001b[0m         )\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/models/query.py:1847\u001b[0m, in \u001b[0;36mQuerySet._insert\u001b[0;34m(self, objs, fields, returning_fields, raw, using, on_conflict, update_fields, unique_fields)\u001b[0m\n\u001b[1;32m   1840\u001b[0m query \u001b[38;5;241m=\u001b[39m sql\u001b[38;5;241m.\u001b[39mInsertQuery(\n\u001b[1;32m   1841\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmodel,\n\u001b[1;32m   1842\u001b[0m     on_conflict\u001b[38;5;241m=\u001b[39mon_conflict,\n\u001b[1;32m   1843\u001b[0m     update_fields\u001b[38;5;241m=\u001b[39mupdate_fields,\n\u001b[1;32m   1844\u001b[0m     unique_fields\u001b[38;5;241m=\u001b[39munique_fields,\n\u001b[1;32m   1845\u001b[0m )\n\u001b[1;32m   1846\u001b[0m query\u001b[38;5;241m.\u001b[39minsert_values(fields, objs, raw\u001b[38;5;241m=\u001b[39mraw)\n\u001b[0;32m-> 1847\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mquery\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_compiler\u001b[49m\u001b[43m(\u001b[49m\u001b[43musing\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43musing\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute_sql\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreturning_fields\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/models/sql/compiler.py:1836\u001b[0m, in \u001b[0;36mSQLInsertCompiler.execute_sql\u001b[0;34m(self, returning_fields)\u001b[0m\n\u001b[1;32m   1834\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mconnection\u001b[38;5;241m.\u001b[39mcursor() \u001b[38;5;28;01mas\u001b[39;00m cursor:\n\u001b[1;32m   1835\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m sql, params \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mas_sql():\n\u001b[0;32m-> 1836\u001b[0m         \u001b[43mcursor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43msql\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1837\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturning_fields:\n\u001b[1;32m   1838\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m []\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/backends/utils.py:79\u001b[0m, in \u001b[0;36mCursorWrapper.execute\u001b[0;34m(self, sql, params)\u001b[0m\n\u001b[1;32m     78\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mexecute\u001b[39m(\u001b[38;5;28mself\u001b[39m, sql, params\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[0;32m---> 79\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_execute_with_wrappers\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     80\u001b[0m \u001b[43m        \u001b[49m\u001b[43msql\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmany\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexecutor\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_execute\u001b[49m\n\u001b[1;32m     81\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/backends/utils.py:92\u001b[0m, in \u001b[0;36mCursorWrapper._execute_with_wrappers\u001b[0;34m(self, sql, params, many, executor)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m wrapper \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mreversed\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdb\u001b[38;5;241m.\u001b[39mexecute_wrappers):\n\u001b[1;32m     91\u001b[0m     executor \u001b[38;5;241m=\u001b[39m functools\u001b[38;5;241m.\u001b[39mpartial(wrapper, executor)\n\u001b[0;32m---> 92\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mexecutor\u001b[49m\u001b[43m(\u001b[49m\u001b[43msql\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmany\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontext\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/backends/utils.py:100\u001b[0m, in \u001b[0;36mCursorWrapper._execute\u001b[0;34m(self, sql, params, *ignored_wrapper_args)\u001b[0m\n\u001b[1;32m     98\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mAPPS_NOT_READY_WARNING_MSG, category\u001b[38;5;241m=\u001b[39m\u001b[38;5;167;01mRuntimeWarning\u001b[39;00m)\n\u001b[1;32m     99\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdb\u001b[38;5;241m.\u001b[39mvalidate_no_broken_transaction()\n\u001b[0;32m--> 100\u001b[0m \u001b[43m\u001b[49m\u001b[38;5;28;43;01mwith\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdb\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrap_database_errors\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m    101\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m:\u001b[49m\n\u001b[1;32m    102\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;66;43;03m# params default might be backend specific.\u001b[39;49;00m\n\u001b[1;32m    103\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43;01mreturn\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcursor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43msql\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/utils.py:91\u001b[0m, in \u001b[0;36mDatabaseErrorWrapper.__exit__\u001b[0;34m(self, exc_type, exc_value, traceback)\u001b[0m\n\u001b[1;32m     89\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m dj_exc_type \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m (DataError, IntegrityError):\n\u001b[1;32m     90\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwrapper\u001b[38;5;241m.\u001b[39merrors_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[0;32m---> 91\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m dj_exc_value\u001b[38;5;241m.\u001b[39mwith_traceback(traceback) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mexc_value\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/pertpy/lib/python3.12/site-packages/django/db/backends/utils.py:105\u001b[0m, in \u001b[0;36mCursorWrapper._execute\u001b[0;34m(self, sql, params, *ignored_wrapper_args)\u001b[0m\n\u001b[1;32m    103\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcursor\u001b[38;5;241m.\u001b[39mexecute(sql)\n\u001b[1;32m    104\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 105\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcursor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43msql\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mIntegrityError\u001b[0m: duplicate key value violates unique constraint \"lnschema_bionty_gene_uid_key\"\nDETAIL:  Key (uid)=(3BEZYDTfG7t4) already exists.\n"
     ]
    }
   ],
   "source": [
    "curator = pt.dt.PerturbationCurator(\n",
    "    adata, using_key=THIS_INSTANCE\n",
    ")\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "bt.Tissue.filter().all().delete()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"sex\"] = adata.obs[\"sex\"].replace({\"Unknown\": \"unknown\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "efo_lo = bt.ExperimentalFactor.public().lookup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"assay\"] = efo_lo.single_cell_rna_sequencing.name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, ~adata.var_names.isin(curator.non_validated[\"var_index\"])].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'assay'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'development_stage'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'disease'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'donor_id'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'self_reported_ethnicity'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'sex'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'suspension_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'tissue_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'organism'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_line'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'genetic_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'compound_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'environmental_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'combination_treatments'\n",
      "\u001b[92m✓\u001b[0m var_index is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m assay is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mcell_type\u001b[0m on \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 terms\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('cell_type')\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mdevelopment_stage\u001b[0m on \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 terms\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('development_stage')\u001b[0m\n",
      "\u001b[92m✓\u001b[0m disease is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mdonor_id\u001b[0m on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 terms\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('donor_id')\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mself_reported_ethnicity\u001b[0m on \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 terms\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('self_reported_ethnicity')\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3msex\u001b[0m on \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 terms\u001b[0m is not validated: \u001b[1;91m'unknown'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('sex')\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3msuspension_type\u001b[0m on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 terms\u001b[0m is not validated: \u001b[1;91m'cell'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('suspension_type')\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mtissue_type\u001b[0m on \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m1 terms\u001b[0m is not validated: \u001b[1;91m'cell_line'\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('tissue_type')\u001b[0m\n",
      "\u001b[92m✓\u001b[0m organism is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m cell_line is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m genetic_treatments is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m compound_treatments is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m environmental_treatments is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m combination_treatments is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Need to recreate Curator object because we are using a new object\n",
    "curator = pt.dt.PerturbationCurator(adata, using_key=THIS_INSTANCE)\n",
    "curator.validate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'assay'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_type'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mCellType.name\u001b[0m for \u001b[3mcell_type\u001b[0m: 'unknown'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'development_stage'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mDevelopmentalStage.name\u001b[0m for \u001b[3mdevelopment_stage\u001b[0m: 'unknown'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'disease'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'donor_id'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mULabel.name\u001b[0m for \u001b[3mdonor_id\u001b[0m: 'unknown'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'self_reported_ethnicity'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mEthnicity.name\u001b[0m for \u001b[3mself_reported_ethnicity\u001b[0m: 'unknown'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'sex'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mPhenotype.name\u001b[0m for \u001b[3msex\u001b[0m: 'unknown'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'suspension_type'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mULabel.name\u001b[0m for \u001b[3msuspension_type\u001b[0m: 'cell'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'tissue_type'\n",
      "\u001b[92m✓\u001b[0m added 1 record with \u001b[3mULabel.name\u001b[0m for \u001b[3mtissue_type\u001b[0m: 'cell_line'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'organism'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_line'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'genetic_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'compound_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'environmental_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'combination_treatments'\n"
     ]
    }
   ],
   "source": [
    "curator.add_new_from(\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'assay'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'development_stage'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'disease'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'donor_id'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'self_reported_ethnicity'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'sex'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'suspension_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'tissue_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'organism'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_line'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'genetic_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'compound_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'environmental_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'combination_treatments'\n",
      "\u001b[92m✓\u001b[0m var_index is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m assay is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m cell_type is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m development_stage is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m disease is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m donor_id is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m self_reported_ethnicity is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m sex is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m suspension_type is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m tissue_type is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m organism is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m cell_line is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m genetic_treatments is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m compound_treatments is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m environmental_treatments is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m combination_treatments is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator.validate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All treatment columns validate but that's only because they're all empty."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Curate perturbations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Move\n",
    "adata.obs[\"genetic_treatments\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"CRISPR\", None\n",
    ")\n",
    "adata.obs[\"compound_treatments\"] = adata.obs[\"perturbation\"].where(\n",
    "    adata.obs[\"perturbation_type\"] == \"drug\", None\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genetic treatments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, 'sggpx4-2', 'sglacz', 'sggpx4-1', 'sgor2j2']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(adata.obs[\"genetic_treatments\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Gene record from Bionty\u001b[0m matching \u001b[3msymbol\u001b[0m: \u001b[1;95m'GPX4'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>system</th>\n",
       "      <th>sequence</th>\n",
       "      <th>on_target_score</th>\n",
       "      <th>off_target_score</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PuVY1Nkbqahe</td>\n",
       "      <td>sggpx4-1</td>\n",
       "      <td>CRISPR KO</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:23.517967+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             uid      name     system sequence on_target_score  \\\n",
       "id                                                               \n",
       "1   PuVY1Nkbqahe  sggpx4-1  CRISPR KO     None            None   \n",
       "\n",
       "   off_target_score run_id                       created_at  created_by_id  \n",
       "id                                                                          \n",
       "1              None   None 2024-10-29 14:07:23.517967+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Glutathione Peroxidase 4'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Gene record from Bionty\u001b[0m matching \u001b[3msynonyms\u001b[0m: \u001b[1;95m'or2j2'\u001b[0m\n",
      "\u001b[93m!\u001b[0m ambiguous validation in Bionty for 1 record: 'OR2J2'\n"
     ]
    }
   ],
   "source": [
    "treatments = [\n",
    "    (\"sggpx4-1\", \"GPX4\", \"Glutathione Peroxidase 4\"),\n",
    "    (\"sggpx4-2\", \"GPX4\", \"Glutathione Peroxidase 4\"),\n",
    "    (\"sgor2j2\", \"or2j2\", \"Olfactory receptor family 2 subfamily J member 2\"),\n",
    "    (\"sglacz\", \"lacz\", \"beta-galactosidase control\"),  # Control from E. coli\n",
    "]\n",
    "organism = bt.Organism.lookup().human\n",
    "\n",
    "genetic_treatments = []\n",
    "for name, symbol, target_name in treatments:\n",
    "    treatment = wl.GeneticTreatment(system=\"CRISPR KO\", name=name).save()\n",
    "    if symbol != \"lacz\":\n",
    "        gene_result = bt.Gene.from_source(symbol=symbol, organism=organism)\n",
    "        gene = gene_result[0] if isinstance(gene_result, list) else gene_result\n",
    "        gene = gene.save()\n",
    "    else:\n",
    "        gene = bt.Gene(symbol=symbol, organism=organism).save()\n",
    "    target = wl.TreatmentTarget(name=target_name).save()\n",
    "    target.genes.add(gene)\n",
    "    treatment.targets.add(target)\n",
    "    genetic_treatments.append(treatment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'assay'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'development_stage'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'disease'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'donor_id'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'self_reported_ethnicity'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'sex'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'suspension_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'tissue_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'organism'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_line'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'genetic_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'compound_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'environmental_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'combination_treatments'\n",
      "\u001b[92m✓\u001b[0m var_index is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m assay is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m cell_type is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m development_stage is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m disease is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m donor_id is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m self_reported_ethnicity is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m sex is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m suspension_type is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m tissue_type is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m organism is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m cell_line is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m genetic_treatments is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[94m•\u001b[0m mapping \u001b[3mcompound_treatments\u001b[0m on \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m    \u001b[1;91m14 terms\u001b[0m are not validated: \u001b[1;91m'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'brd3379', 'JQ1', 'azd5591', 'control', ...\u001b[0m\n",
      "→ fix typos, remove non-existent values, or save terms via \u001b[1;91m.add_new_from('compound_treatments')\u001b[0m\n",
      "\u001b[92m✓\u001b[0m environmental_treatments is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m combination_treatments is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curator.validate()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m8 Compound records from Bionty\u001b[0m matching \u001b[3mname\u001b[0m: \u001b[1;95m'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'JQ1', 'everolimus'\u001b[0m\n",
      "\u001b[93m!\u001b[0m \u001b[1;91mdid not create\u001b[0m Compound records for \u001b[1;93m6 non-validated\u001b[0m \u001b[3mnames\u001b[0m: \u001b[1;93m'azd5591', 'brd3379', 'control', 'idasanutlin', 'prexasertib', 'taselisib'\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "compounds = wl.Compound.from_values(adata.obs[\"compound_treatments\"], field=\"name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The remaining compounds are not in chebi and we create records for them\n",
    "for missing in [\n",
    "    \"azd5591\",\n",
    "    \"brd3379\",\n",
    "    \"control\",\n",
    "    \"idasanutlin\",\n",
    "    \"prexasertib\",\n",
    "    \"taselisib\",\n",
    "]:\n",
    "    compounds.append(wl.Compound(name=missing))\n",
    "ln.save(compounds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "drug_metadata = adata.obs[adata.obs[\"compound_treatments\"].notna()]\n",
    "\n",
    "unique_treatments = drug_metadata[\n",
    "    [\"perturbation\", \"dose_unit\", \"dose_value\"]\n",
    "].drop_duplicates()\n",
    "\n",
    "compound_treatments = []\n",
    "for _, row in unique_treatments.iterrows():\n",
    "    compound = wl.Compound.get(name=row[\"perturbation\"])\n",
    "    treatment = wl.CompoundTreatment(\n",
    "        name=compound.name,\n",
    "        concentration=row[\"dose_value\"],\n",
    "        concentration_unit=row[\"dose_unit\"],\n",
    "    )\n",
    "    compound_treatments.append(treatment)\n",
    "\n",
    "ln.save(compound_treatments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P36507'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P00533'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q9UK79'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q15303'\u001b[0m\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'MAPK/ERK pathway'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P15056'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P23921'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P10415'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q07812'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CjakOwn0</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:36.548584+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid              name description run_id  \\\n",
       "id                                                  \n",
       "4   CjakOwn0  MAPK/ERK pathway        None   None   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "4  2024-10-29 14:07:36.548584+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P49721'\u001b[0m\n",
      "\u001b[93m!\u001b[0m record with similar name exists! did you mean to load it?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5EdhKMoj</td>\n",
       "      <td>Apoptosis regulation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:46.070632+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                  name description run_id  \\\n",
       "id                                                      \n",
       "7   5EdhKMoj  Apoptosis regulation        None   None   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "7  2024-10-29 14:07:46.070632+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O60885'\u001b[0m\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Transcription regulation (BET proteins)'\n",
      "\u001b[92m→\u001b[0m returning existing TreatmentTarget record with same name: 'Apoptosis regulation'\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q07820'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O14757'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CjakOwn0</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:36.548584+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>yBJ85bnI</td>\n",
       "      <td>Proteasome pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:49.114596+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                name description run_id  \\\n",
       "id                                                    \n",
       "4   CjakOwn0    MAPK/ERK pathway        None   None   \n",
       "8   yBJ85bnI  Proteasome pathway        None   None   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "4  2024-10-29 14:07:36.548584+00:00              1  \n",
       "8  2024-10-29 14:07:49.114596+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P42336'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'O00329'\u001b[0m\n",
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P48736'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5EdhKMoj</td>\n",
       "      <td>Apoptosis regulation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:46.070632+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ppl50wZN</td>\n",
       "      <td>Transcription regulation (BET proteins)</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:50.636207+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                                     name description run_id  \\\n",
       "id                                                                         \n",
       "7   5EdhKMoj                     Apoptosis regulation        None   None   \n",
       "9   Ppl50wZN  Transcription regulation (BET proteins)        None   None   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "7  2024-10-29 14:07:46.070632+00:00              1  \n",
       "9  2024-10-29 14:07:50.636207+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'Q00987'\u001b[0m\n",
      "\u001b[93m!\u001b[0m records with similar names exist! did you mean to load one of them?\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uid</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>run_id</th>\n",
       "      <th>created_at</th>\n",
       "      <th>created_by_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2HxVNgiL</td>\n",
       "      <td>PI3K/AKT/mTOR pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:54.820080+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CjakOwn0</td>\n",
       "      <td>MAPK/ERK pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:36.548584+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>yBJ85bnI</td>\n",
       "      <td>Proteasome pathway</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2024-10-29 14:07:49.114596+00:00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         uid                   name description run_id  \\\n",
       "id                                                       \n",
       "12  2HxVNgiL  PI3K/AKT/mTOR pathway        None   None   \n",
       "4   CjakOwn0       MAPK/ERK pathway        None   None   \n",
       "8   yBJ85bnI     Proteasome pathway        None   None   \n",
       "\n",
       "                         created_at  created_by_id  \n",
       "id                                                  \n",
       "12 2024-10-29 14:07:54.820080+00:00              1  \n",
       "4  2024-10-29 14:07:36.548584+00:00              1  \n",
       "8  2024-10-29 14:07:49.114596+00:00              1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m✓\u001b[0m created \u001b[1;95m1 Protein record from Bionty\u001b[0m matching \u001b[3muniprotkb_id\u001b[0m: \u001b[1;95m'P42345'\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "compounds_to_targets = {\n",
    "    \"trametinib\": (\"MAPK/ERK pathway\", [\"P36507\"]),\n",
    "    \"afatinib\": (\"EGFR, HER2, HER4 signaling\", [\"P00533\", \"Q9UK79\", \"Q15303\"]),\n",
    "    \"dabrafenib\": (\"MAPK/ERK pathway\", [\"P15056\"]),\n",
    "    \"gemcitabine\": (\"DNA synthesis inhibition\", [\"P23921\"]),  # No single protein target\n",
    "    \"navitoclax\": (\"Apoptosis regulation\", [\"P10415\", \"Q07812\"]),\n",
    "    \"bortezomib\": (\"Proteasome pathway\", [\"P49721\"]),\n",
    "    \"brd3379\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"JQ1\": (\"Transcription regulation (BET proteins)\", [\"O60885\"]),\n",
    "    \"azd5591\": (\"Apoptosis regulation\", [\"Q07820\"]),\n",
    "    \"control\": (\"Baseline\", [None]),  # No target for control\n",
    "    \"prexasertib\": (\"DNA damage response\", [\"O14757\"]),\n",
    "    \"taselisib\": (\"PI3K/AKT/mTOR pathway\", [\"P42336\", \"O00329\", \"P48736\"]),\n",
    "    \"idasanutlin\": (\"p53 regulation\", [\"Q00987\"]),\n",
    "    \"everolimus\": (\"mTOR pathway\", [\"P42345\"])\n",
    "}\n",
    "\n",
    "\n",
    "for compound_treatment_name, targets_tuple in compounds_to_targets.items():\n",
    "    compound_treatment = wl.CompoundTreatment.get(name=compound_treatment_name)\n",
    "    target = wl.TreatmentTarget(name=targets_tuple[0]).save()\n",
    "    proteins = []\n",
    "    for id in targets_tuple[1]:\n",
    "        if id is not None:\n",
    "            proteins.append(bt.Protein.from_source(uniprotkb_id=id).save())\n",
    "    target.proteins.set(proteins)\n",
    "    compound_treatment.targets.add(target)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = fr.Reference(\n",
    "    name=\"Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action\",\n",
    "    abbr=\"McFarland 2020\",\n",
    "    url=\"https://www.nature.com/articles/s41467-020-17440-w\",\n",
    "    doi=\"10.1038/s41467-020-17440-w\",\n",
    "    text=(\n",
    "        \"Assays to study cancer cell responses to pharmacologic or genetic perturbations are typically \"\n",
    "        \"restricted to using simple phenotypic readouts such as proliferation rate. Information-rich assays, \"\n",
    "        \"such as gene-expression profiling, have generally not permitted efficient profiling of a given \"\n",
    "        \"perturbation across multiple cellular contexts. Here, we develop MIX-Seq, a method for multiplexed \"\n",
    "        \"transcriptional profiling of post-perturbation responses across a mixture of samples with single-cell \"\n",
    "        \"resolution, using SNP-based computational demultiplexing of single-cell RNA-sequencing data. We show \"\n",
    "        \"that MIX-Seq can be used to profile responses to chemical or genetic perturbations across pools of 100 \"\n",
    "        \"or more cancer cell lines. We combine it with Cell Hashing to further multiplex additional experimental \"\n",
    "        \"conditions, such as post-treatment time points or drug doses. Analyzing the high-content readout of \"\n",
    "        \"scRNA-seq reveals both shared and context-specific transcriptional response components that can identify \"\n",
    "        \"drug mechanism of action and enable prediction of long-term cell viability from short-term transcriptional \"\n",
    "        \"responses to treatment.\"\n",
    "    ),\n",
    ").save()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove unused columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs = adata.obs.drop(\n",
    "    [\n",
    "        \"depmap_id\",\n",
    "        \"cancer\",\n",
    "        \"cell_quality\",\n",
    "        \"channel\",\n",
    "        \"perturbation\",\n",
    "        \"perturbation_type\",\n",
    "        \"singlet_dev\",\n",
    "        \"singlet_dev_z\",\n",
    "        \"singlet_margin\",\n",
    "        \"singlet_z_margin\",\n",
    "        \"nperts\",\n",
    "        \"ngenes\",\n",
    "        \"ncounts\",\n",
    "        \"cell_det_rate\",\n",
    "        \"doublet_GMM_prob\",\n",
    "        \"doublet_dev_imp\",\n",
    "        \"doublet_z_margin\",\n",
    "        'doublet_CL1',\n",
    "        'doublet_CL2'\n",
    "    ],\n",
    "    axis=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Register curated artifact"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[92m→\u001b[0m validating metadata using registries of instance \u001b[3mtest-perturbation\u001b[0m\n",
      "\u001b[94m•\u001b[0m saving validated records of 'var_index'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'assay'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'development_stage'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'disease'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'donor_id'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'self_reported_ethnicity'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'sex'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'suspension_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'tissue_type'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'organism'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'cell_line'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'genetic_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'compound_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'environmental_treatments'\n",
      "\u001b[94m•\u001b[0m saving validated terms of 'combination_treatments'\n",
      "\u001b[92m✓\u001b[0m var_index is validated against \u001b[3mGene.ensembl_gene_id\u001b[0m\n",
      "\u001b[92m✓\u001b[0m assay is validated against \u001b[3mExperimentalFactor.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m cell_type is validated against \u001b[3mCellType.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m development_stage is validated against \u001b[3mDevelopmentalStage.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m disease is validated against \u001b[3mDisease.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m donor_id is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m self_reported_ethnicity is validated against \u001b[3mEthnicity.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m sex is validated against \u001b[3mPhenotype.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m suspension_type is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m tissue_type is validated against \u001b[3mULabel.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m organism is validated against \u001b[3mOrganism.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m cell_line is validated against \u001b[3mCellLine.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m genetic_treatments is validated against \u001b[3mGeneticTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m compound_treatments is validated against \u001b[3mCompoundTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m environmental_treatments is validated against \u001b[3mEnvironmentalTreatment.name\u001b[0m\n",
      "\u001b[92m✓\u001b[0m combination_treatments is validated against \u001b[3mCombinationTreatment.name\u001b[0m\n",
      "\u001b[93m!\u001b[0m no run & transform got linked, call `ln.track()` & re-run\n",
      "\u001b[93m!\u001b[0m run input wasn't tracked, call `ln.track()` and re-run\n",
      "\u001b[93m!\u001b[0m    \u001b[1;93m11 unique terms\u001b[0m (42.30%) are not validated for \u001b[3mname\u001b[0m: \u001b[1;93mdose_unit, dose_value, hash_assignment, hash_tag, num_SNPs, singlet_ID, time, tot_reads, percent_mito, percent_ribo, chembl-ID\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "artifact = curator.save_artifact(description=\"McFarland AnnData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the perturbations and references\n",
    "artifact.genetic_treatments.set(genetic_treatments)\n",
    "artifact.compound_treatments.set(compound_treatments)\n",
    "artifact.references.add(reference)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;92mArtifact\u001b[0m(uid='y3B2kwItyAgpru4G0000', is_latest=True, description='McFarland AnnData', suffix='.h5ad', type='dataset', size=3345456, hash='dETFEMNWe6nH2enh1PNUTg', n_observations=1000, _hash_type='md5', _accessor='AnnData', visibility=1, _key_is_virtual=True, created_at=2024-10-29 14:08:04 UTC)\n",
      "  \u001b[3mProvenance\u001b[0m\n",
      "    .storage = '/home/zeth/PycharmProjects/pertpy-datasets/lamin/test-perturbation'\n",
      "    .created_by = 'zethson'\n",
      "  \u001b[3mLabels\u001b[0m\n",
      "    .references = 'Multiplexed single-cell transcriptional response profiling to define cancer vulnerabilities and therapeutic mechanism of action'\n",
      "    .genetic_treatments = 'sggpx4-1', 'sggpx4-2', 'sgor2j2', 'sglacz'\n",
      "    .compound_treatments = 'trametinib', 'afatinib', 'dabrafenib', 'gemcitabine', 'navitoclax', 'bortezomib', 'brd3379', 'JQ1', 'azd5591', 'control', ...\n",
      "    .organisms = 'human'\n",
      "    .cell_types = 'unknown'\n",
      "    .diseases = 'lung cancer', 'urinary bladder carcinoma', 'colorectal cancer', 'head and neck cancer', 'brain cancer', 'skin cancer', 'gastric cancer', 'esophageal cancer', 'ovarian cancer', 'malignant pancreatic neoplasm', ...\n",
      "    .cell_lines = 'LUDLU-1', 'LU99', 'J82', 'HCT-15', 'YD-10B', 'SNB75', 'SK-MEL-2', 'UM-UC-1', 'AGS', 'COLO-680N', ...\n",
      "    .phenotypes = 'male', 'female', 'unknown'\n",
      "    .experimental_factors = 'single-cell RNA sequencing'\n",
      "    .developmental_stages = 'unknown'\n",
      "    .ethnicities = 'unknown'\n",
      "    .ulabels = 'unknown', 'cell', 'cell_line'\n",
      "  \u001b[3mFeatures\u001b[0m\n",
      "    'assay' = 'single-cell RNA sequencing'\n",
      "    'cell_line' = '22Rv1', '253J-BV', '42-MG-BA', '639-V', '647-V', '769-P', '786-O', '8505C', 'A-375', 'A2780', ...\n",
      "    'cell_type' = 'unknown'\n",
      "    'development_stage' = 'unknown'\n",
      "    'disease' = 'bile duct cancer', 'bone cancer', 'brain cancer', 'breast cancer', 'colorectal cancer', 'esophageal cancer', 'gastric cancer', 'head and neck cancer', 'kidney cancer', 'liver cancer', ...\n",
      "    'donor_id' = 'unknown'\n",
      "    'organism' = 'human'\n",
      "    'self_reported_ethnicity' = 'unknown'\n",
      "    'sex' = 'female', 'male', 'unknown'\n",
      "    'suspension_type' = 'cell'\n",
      "    'tissue_type' = 'cell_line'\n",
      "  \u001b[3mFeature sets\u001b[0m\n",
      "    'var' = 'MAGED2', 'CPE', 'DDX43', 'SPANXA2', 'LNCPRESS1', 'TMEM175', 'None', 'MMD', 'LINC01834', 'DYNC2I1', 'CCRL2', 'MRPS6', 'KCNH1', 'TBILA', 'SLC5A7', 'GSX2', 'ATP5MG', 'TMEM190', 'GTF2H1'\n",
      "    'obs' = 'assay', 'cell_type', 'development_stage', 'disease', 'donor_id', 'self_reported_ethnicity', 'sex', 'suspension_type', 'tissue_type', 'organism', 'cell_line', 'genetic_treatments', 'compound_treatments', 'environmental_treatments', 'combination_treatments'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "artifact.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line</th>\n",
       "      <th>disease</th>\n",
       "      <th>dose_unit</th>\n",
       "      <th>dose_value</th>\n",
       "      <th>hash_assignment</th>\n",
       "      <th>hash_tag</th>\n",
       "      <th>num_SNPs</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>singlet_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>assay</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>genetic_treatments</th>\n",
       "      <th>compound_treatments</th>\n",
       "      <th>environmental_treatments</th>\n",
       "      <th>combination_treatments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AACTGGTGTCTCTCTG</th>\n",
       "      <td>LUDLU-1</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>481</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>LUDLU1_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trametinib</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATAGGCTCAGATTTCG</th>\n",
       "      <td>LU99</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.5</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>1003</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>LU99_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>afatinib</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCCAAATCAAGCCGTC</th>\n",
       "      <td>J82</td>\n",
       "      <td>urinary bladder carcinoma</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>647</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>J82_URINARY_TRACT</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dabrafenib</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGGAGAAGTCGCGTCA</th>\n",
       "      <td>HCT-15</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>30</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>HCT15_LARGE_INTESTINE</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gemcitabine</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAGTTGGAGATCGATA</th>\n",
       "      <td>YD-10B</td>\n",
       "      <td>head and neck cancer</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>874</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>YD10B_UPPER_AERODIGESTIVE_TRACT</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>sggpx4-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATCTAGTCATCGGAT</th>\n",
       "      <td>NCI-H1581</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>480</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>NCIH1581_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dabrafenib</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACTCTTAGTCCCACG</th>\n",
       "      <td>LS1034</td>\n",
       "      <td>colorectal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>366</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>LS1034_LARGE_INTESTINE</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>control</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACTGTCCAGTCACGC</th>\n",
       "      <td>A549</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>2.5</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>528</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>A549_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JQ1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACCTTGGTCGACTAT</th>\n",
       "      <td>NCI-H2347</td>\n",
       "      <td>lung cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>1230</td>\n",
       "      <td>human</td>\n",
       "      <td>Female</td>\n",
       "      <td>NCIH2347_LUNG</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>control</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGTAAGG</th>\n",
       "      <td>TE-6</td>\n",
       "      <td>esophageal cancer</td>\n",
       "      <td>µM</td>\n",
       "      <td>0.0</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>782</td>\n",
       "      <td>human</td>\n",
       "      <td>Male</td>\n",
       "      <td>TE6_OESOPHAGUS</td>\n",
       "      <td>...</td>\n",
       "      <td>single-cell RNA sequencing</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>brd3379</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  cell_line                    disease dose_unit  dose_value  \\\n",
       "AACTGGTGTCTCTCTG    LUDLU-1                lung cancer        µM         0.1   \n",
       "ATAGGCTCAGATTTCG       LU99                lung cancer        µM         0.5   \n",
       "GCCAAATCAAGCCGTC        J82  urinary bladder carcinoma        µM         0.1   \n",
       "CGGAGAAGTCGCGTCA     HCT-15          colorectal cancer        µM         0.1   \n",
       "TAGTTGGAGATCGATA     YD-10B       head and neck cancer       nan         NaN   \n",
       "...                     ...                        ...       ...         ...   \n",
       "GATCTAGTCATCGGAT  NCI-H1581                lung cancer        µM         0.1   \n",
       "AACTCTTAGTCCCACG     LS1034          colorectal cancer        µM         0.0   \n",
       "CACTGTCCAGTCACGC       A549                lung cancer        µM         2.5   \n",
       "CACCTTGGTCGACTAT  NCI-H2347                lung cancer        µM         0.0   \n",
       "TTTGTCACACGTAAGG       TE-6          esophageal cancer        µM         0.0   \n",
       "\n",
       "                 hash_assignment hash_tag  num_SNPs organism     sex  \\\n",
       "AACTGGTGTCTCTCTG             nan      nan       481    human    Male   \n",
       "ATAGGCTCAGATTTCG             nan      nan      1003    human    Male   \n",
       "GCCAAATCAAGCCGTC             nan      nan       647    human    Male   \n",
       "CGGAGAAGTCGCGTCA             nan      nan        30    human    Male   \n",
       "TAGTTGGAGATCGATA             nan      nan       874    human    Male   \n",
       "...                          ...      ...       ...      ...     ...   \n",
       "GATCTAGTCATCGGAT             nan      nan       480    human    Male   \n",
       "AACTCTTAGTCCCACG             nan      nan       366    human    Male   \n",
       "CACTGTCCAGTCACGC             nan      nan       528    human    Male   \n",
       "CACCTTGGTCGACTAT             nan      nan      1230    human  Female   \n",
       "TTTGTCACACGTAAGG             nan      nan       782    human    Male   \n",
       "\n",
       "                                       singlet_ID  ...  \\\n",
       "AACTGGTGTCTCTCTG                      LUDLU1_LUNG  ...   \n",
       "ATAGGCTCAGATTTCG                        LU99_LUNG  ...   \n",
       "GCCAAATCAAGCCGTC                J82_URINARY_TRACT  ...   \n",
       "CGGAGAAGTCGCGTCA            HCT15_LARGE_INTESTINE  ...   \n",
       "TAGTTGGAGATCGATA  YD10B_UPPER_AERODIGESTIVE_TRACT  ...   \n",
       "...                                           ...  ...   \n",
       "GATCTAGTCATCGGAT                    NCIH1581_LUNG  ...   \n",
       "AACTCTTAGTCCCACG           LS1034_LARGE_INTESTINE  ...   \n",
       "CACTGTCCAGTCACGC                        A549_LUNG  ...   \n",
       "CACCTTGGTCGACTAT                    NCIH2347_LUNG  ...   \n",
       "TTTGTCACACGTAAGG                   TE6_OESOPHAGUS  ...   \n",
       "\n",
       "                                       assay cell_type  development_stage  \\\n",
       "AACTGGTGTCTCTCTG  single-cell RNA sequencing   unknown            unknown   \n",
       "ATAGGCTCAGATTTCG  single-cell RNA sequencing   unknown            unknown   \n",
       "GCCAAATCAAGCCGTC  single-cell RNA sequencing   unknown            unknown   \n",
       "CGGAGAAGTCGCGTCA  single-cell RNA sequencing   unknown            unknown   \n",
       "TAGTTGGAGATCGATA  single-cell RNA sequencing   unknown            unknown   \n",
       "...                                      ...       ...                ...   \n",
       "GATCTAGTCATCGGAT  single-cell RNA sequencing   unknown            unknown   \n",
       "AACTCTTAGTCCCACG  single-cell RNA sequencing   unknown            unknown   \n",
       "CACTGTCCAGTCACGC  single-cell RNA sequencing   unknown            unknown   \n",
       "CACCTTGGTCGACTAT  single-cell RNA sequencing   unknown            unknown   \n",
       "TTTGTCACACGTAAGG  single-cell RNA sequencing   unknown            unknown   \n",
       "\n",
       "                  donor_id  self_reported_ethnicity suspension_type  \\\n",
       "AACTGGTGTCTCTCTG   unknown                  unknown            cell   \n",
       "ATAGGCTCAGATTTCG   unknown                  unknown            cell   \n",
       "GCCAAATCAAGCCGTC   unknown                  unknown            cell   \n",
       "CGGAGAAGTCGCGTCA   unknown                  unknown            cell   \n",
       "TAGTTGGAGATCGATA   unknown                  unknown            cell   \n",
       "...                    ...                      ...             ...   \n",
       "GATCTAGTCATCGGAT   unknown                  unknown            cell   \n",
       "AACTCTTAGTCCCACG   unknown                  unknown            cell   \n",
       "CACTGTCCAGTCACGC   unknown                  unknown            cell   \n",
       "CACCTTGGTCGACTAT   unknown                  unknown            cell   \n",
       "TTTGTCACACGTAAGG   unknown                  unknown            cell   \n",
       "\n",
       "                 genetic_treatments compound_treatments  \\\n",
       "AACTGGTGTCTCTCTG                NaN          trametinib   \n",
       "ATAGGCTCAGATTTCG                NaN            afatinib   \n",
       "GCCAAATCAAGCCGTC                NaN          dabrafenib   \n",
       "CGGAGAAGTCGCGTCA                NaN         gemcitabine   \n",
       "TAGTTGGAGATCGATA           sggpx4-2                 NaN   \n",
       "...                             ...                 ...   \n",
       "GATCTAGTCATCGGAT                NaN          dabrafenib   \n",
       "AACTCTTAGTCCCACG                NaN             control   \n",
       "CACTGTCCAGTCACGC                NaN                 JQ1   \n",
       "CACCTTGGTCGACTAT                NaN             control   \n",
       "TTTGTCACACGTAAGG                NaN             brd3379   \n",
       "\n",
       "                 environmental_treatments combination_treatments  \n",
       "AACTGGTGTCTCTCTG                                                  \n",
       "ATAGGCTCAGATTTCG                                                  \n",
       "GCCAAATCAAGCCGTC                                                  \n",
       "CGGAGAAGTCGCGTCA                                                  \n",
       "TAGTTGGAGATCGATA                                                  \n",
       "...                                   ...                    ...  \n",
       "GATCTAGTCATCGGAT                                                  \n",
       "AACTCTTAGTCCCACG                                                  \n",
       "CACTGTCCAGTCACGC                                                  \n",
       "CACCTTGGTCGACTAT                                                  \n",
       "TTTGTCACACGTAAGG                                                  \n",
       "\n",
       "[1000 rows x 26 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pertpy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
